{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Pawel_workflow-Low_Bacteria_Riken"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This document is an example of how to process a C1-CAGE library with a Jupyter notebook from raw reads to single molecule count. In the following section we assume that:\n",
    "- The softwares used in this workflow are mentioned in the [prerequesite](https://github.com/Population-Transcriptomics/C1-CAGE-preview/blob/master/prerequisite.md) section.\n",
    "- The reference genome has to be already indexed with bwa\n",
    "- The tutorials are introduced with the example file mentioned in the tutorial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In our hands this notebook worked without trouble on a machine running Debian GNU/Linux 8. We noticed that the behavior of tagdust2 in single-end mode was different on Mac OSX. In short, the order of the reads1 is changed after extraction on Mac OSX which is a problem because syncpairs expect the order of reads1 and reads2 to be the same. One way to overcome this issueis sort reads1 and reads2 separately after the exctraction then syncpairs will work properly."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setting the Input and Output Folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Setting the output folder where all the results would be stored\n",
    "outputFolder = '/home/baker/my-scratch-share/James_Bagnall_LowMOI_Unique_Bacteria_Riken_Misesq_length/output/'\n",
    "\n",
    "# inputFolder : Folder name with all the fastq files\n",
    "intputFolder = '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import subprocess, os, csv, signal, pysam, collections, optparse, sys, tempfile, re, optparse\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import rpy2.interactive as r\n",
    "import rpy2.interactive.packages # this can take few seconds\n",
    "import time\n",
    "#from ZIFA import ZIFA\n",
    "#from ZIFA import block_ZIFA\n",
    "import pylab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: \n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Loading required package: limma\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: \n",
      "Attaching package: ‘cowplot’\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: The following object is masked from ‘package:ggplot2’:\n",
      "\n",
      "    ggsave\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Loading required package: flexmix\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Loading required package: lattice\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Loading required package: AnnotationDbi\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Loading required package: stats4\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Loading required package: BiocGenerics\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Loading required package: parallel\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: The following object is masked from ‘package:flexmix’:\n",
      "\n",
      "    unique\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, xtabs\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, as.vector, cbind, colnames,\n",
      "    do.call, duplicated, eval, evalq, Filter, Find, get, grep, grepl,\n",
      "    intersect, is.unsorted, lapply, lengths, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unlist, unsplit\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Loading required package: Biobase\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Loading required package: IRanges\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Loading required package: S4Vectors\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: \n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: \n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Loading required package: DBI\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: \n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Loading required package: Lmoments\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Loading required package: distillery\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: Loading required package: car\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: \n",
      "Attaching package: ‘extRemes’\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n",
      "/usr/local/lib/python2.7/dist-packages/rpy2/robjects/functions.py:106: UserWarning: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    qqnorm, qqplot\n",
      "\n",
      "\n",
      "  res = super(Function, self).__call__(*new_args, **new_kwargs)\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "library(\"dplyr\")\n",
    "library(\"edgeR\")\n",
    "library(\"ggplot2\")\n",
    "library(\"cowplot\")\n",
    "library(\"BASiCS\")\n",
    "library(\"scde\")\n",
    "library(repr)\n",
    "library(biomaRt)\n",
    "library(GO.db)\n",
    "library(extRemes)\n",
    "library(Lmoments)\n",
    "library(distillery)\n",
    "library(car)\n",
    "theme_set(theme_bw(base_size = 16))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Custom functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "remove_extension = lambda x: x.split('.')[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Declare the function that deals with inputs and outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_args(read1, read2, ref_genome, bacterial_ref_genome, Spike_in_genome, output_folders):\n",
    "    '''Set the input and output path for a given pair of reads'''\n",
    "    r1_shortname = remove_extension(os.path.basename(read1))\n",
    "\n",
    "    args = {  \n",
    "        'r1_input': read1,\n",
    "        'r2_input': read2,\n",
    "        'ref_genome': ref_genome,\n",
    "        'bacterial_ref_genome': bacterial_ref_genome,\n",
    "        'Spike_in_genome': Spike_in_genome\n",
    "    }\n",
    "    \n",
    "    output_paths = {folder: os.path.join(outputFolder, folder, r1_shortname) for folder in output_folders}\n",
    "    \n",
    "    return dict(args, **output_paths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def ERCC_finder(f):\n",
    "    full_path = cleanedReadsFolder + f\n",
    "    file_ERCC = open(full_path,'rb')\n",
    "    ERCCTempFilePath = outputFolder + 'ERCC_count/ERCC_count_'\n",
    "    for line in file_ERCC:\n",
    "        if 'ERCC' in line:\n",
    "            temp = str.split(line)            \n",
    "            ERCC_file_path = ERCCTempFilePath + f.split('_', 1)[0]+\".txt\"            \n",
    "            ERCC_command_file = open(ERCC_file_path, 'a')\n",
    "            ERCC_line_to_write = temp[3] + \"\\t\" + temp[2]+ \"\\n\"\n",
    "            ERCC_command_file.writelines(ERCC_line_to_write)\n",
    "            ERCC_command_file.close"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def expanderuser(path):\n",
    "    \"\"\"Replaces the ~ with the users home directory\"\"\"\n",
    "    if path.startswith(\"~\"):\n",
    "        return os.path.expanduser(\"~\") + path[1: ]\n",
    "    return path\n",
    "\n",
    "def multiple_match(regex, old_name, name):\n",
    "    report_error(regex+\" matched both \" + old_name + \" and \" + name + \"strickter regex required\")\n",
    "\n",
    "def do_walk(source, regex, target_path, onerror=None, followlinks=False, verbose=True, divider=None, sort=None):\n",
    "    \"\"\"\n",
    "    Walker method\n",
    "    Inputs are:\n",
    "    source Parent directory to be walked through\n",
    "    onerror is a function on underlying os.listdir fails\n",
    "    followlinks would allow following symbolic links\n",
    "        WARNING Be aware that setting followlinks to True can lead to infinite\n",
    "           recursion if a link points to a parent directory of itself.\n",
    "           walk() does not keep track of the directories it visited already.\n",
    "    verbose: Flag passed to action methods to provide more output\n",
    "    \"\"\"\n",
    "    source = expanderuser(source)\n",
    "    (names_number, names_path) = tempfile.mkstemp()\n",
    "    re_object = re.compile(regex)\n",
    "    file_paths = []\n",
    "    with open(names_path, 'w') as f:\n",
    "        # Must be topdown=True otherwise walk will process subdirectories before checking them\n",
    "        for root, dirs, files in os.walk(source, topdown=True, onerror=onerror, followlinks=followlinks):\n",
    "            for name in files:\n",
    "                if re_object.search(name):\n",
    "                    file_path = os.path.join(root, name)\n",
    "                    f.write(file_path)\n",
    "                    f.write(\"\\n\")\n",
    "                    file_paths.append(file_path)\n",
    "                    if verbose:\n",
    "                        print \"Merging\",file_path\n",
    "                #else:\n",
    "                #    print \"NO\",name, root\n",
    "    if len(file_paths) == 0:\n",
    "        report_error(\"NO files found to match \"+ regex)\n",
    "    if divider:\n",
    "        if sort:\n",
    "            merge_files(file_paths, names_path, target_path, verbose=verbose, divider = divider, sort = sort)\n",
    "        else:\n",
    "            merge_files(file_paths, names_path, target_path, verbose=verbose, divider = divider)\n",
    "    else:\n",
    "        if sort:\n",
    "            merge_files(file_paths, names_path, target_path, verbose, sort=sort)\n",
    "        else:\n",
    "            merge_files(file_paths, names_path, target_path, verbose)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "def report_error(error):\n",
    "    \"\"\"Prints the error, and exits -1\"\"\"\n",
    "    print error\n",
    "    sys.stderr.write(error)\n",
    "    sys.stderr.write(\"\\n\")\n",
    "    sys.stderr.flush()\n",
    "    sys.exit(1)\n",
    "\n",
    "\n",
    "def remove_common(names):\n",
    "    start = names[0]\n",
    "    end = names[0]\n",
    "    for name in names:\n",
    "        while len(start) > 0 and not(name.startswith(start)):\n",
    "            start = start[: -1]\n",
    "        while len(end) > 0 and not(name.endswith(end)):\n",
    "            end = end[1:]\n",
    "    new_names = []\n",
    "    for name in names:\n",
    "        new_name = name[len(start): -len(end)]\n",
    "        new_names.append(\"C\"+new_name)\n",
    "    return new_names\n",
    "\n",
    "\n",
    "def remove_symbols(s):\n",
    "    if s.find(\"__\") == -1:\n",
    "        return s\n",
    "    # Patterns used by Galaxy\n",
    "    s = s.replace(\"__cb__\", ']')\n",
    "    s = s.replace(\"__cc__\", '}')\n",
    "    s = s.replace(\"__dq__\", '\"')\n",
    "    s = s.replace(\"__lt__\", '<')\n",
    "    s = s.replace(\"__gt__\", '>')\n",
    "    s = s.replace(\"__ob__\", '[')\n",
    "    s = s.replace(\"__oc__\", '{')\n",
    "    s = s.replace(\"__sq__\", \"'\")\n",
    "    # Patterns added by Christian\n",
    "    s = s.replace(\"__in__\", '%in%')\n",
    "    s = s.replace(\"__not__\", '!')\n",
    "    end = -2\n",
    "    # tab = 9\n",
    "    # | = 124\n",
    "    while True:\n",
    "        start = s.find(\"__\", end + 2) + 2\n",
    "        if start == 1:\n",
    "            return s\n",
    "        end = s.find(\"__\", start)\n",
    "        if end == -1:\n",
    "            return s\n",
    "        part = s[start: end]\n",
    "        try:\n",
    "            ascii = int(part)\n",
    "            s = s.replace(\"__\" + part + \"__\", chr(ascii))\n",
    "            end = -2\n",
    "        except ValueError:\n",
    "            pass\n",
    "    return s\n",
    "\n",
    "\n",
    "def clean_part(part):\n",
    "    part = part.strip()\n",
    "    part = part.replace(\"\\t\", \"__9__\")\n",
    "    return part\n",
    "\n",
    "\n",
    "def return_blank():\n",
    "    return \"0\"\n",
    "\n",
    "\n",
    "def black_dict():\n",
    "    return collections.defaultdict(return_blank)\n",
    "\n",
    "\n",
    "def merge_files(file_paths, names_path, target_path, verbose=False, divider=\"\\t\", sort = None):\n",
    "    if sort:\n",
    "        if sort == \"column_names\":\n",
    "            column_sort = True\n",
    "            row_sort = False\n",
    "        elif sort == \"row_names\":\n",
    "            column_sort = False\n",
    "            row_sort = true\n",
    "        elif sort == \"both\":\n",
    "            column_sort = True\n",
    "            row_sort = True\n",
    "        elif sort == \"none\":\n",
    "            column_sort = False\n",
    "            row_sort = False\n",
    "        else:\n",
    "            report_error(\"Unexpected value: \" + sort + \" for sort parameter. Legeal values are: column_names, row_names, both or none\")\n",
    "    else:\n",
    "        column_sort = False\n",
    "        row_sort = False\n",
    "\n",
    "    names = []\n",
    "    with open(names_path, 'r') as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if len(line) > 0:\n",
    "                names.append(line)\n",
    "    if len(names) != len(file_paths):\n",
    "        report_error(\"Found \" + str(len(file_paths)) + \" file_paths but \" + names_path + \" contains \" + str(len(names)) + \" lines.\")\n",
    "    new_names = remove_common(names)\n",
    "    clean_divider = remove_symbols(divider)\n",
    "    all_values = collections.defaultdict(black_dict)\n",
    "    for count, file_path in enumerate(file_paths):\n",
    "        mis_match = 0\n",
    "        with open(file_path, 'r') as f:\n",
    "            for line in f:\n",
    "                parts = line.strip().split(clean_divider)\n",
    "                if len(parts) == 2:\n",
    "                    key = clean_part(parts[0])\n",
    "                    value = clean_part(parts[1])\n",
    "                    all_values[key][new_names[count]] = value\n",
    "                else:\n",
    "                    mis_match+= 1\n",
    "                    if verbose:\n",
    "                        if mis_match < 5:\n",
    "                            print \"ignoring following line from\", file_path\n",
    "                            print line\n",
    "        if mis_match > 0:\n",
    "            print \"In file \" + file_path + \" \" + str(mis_match) + \" lines did not have 1 divider (\" + clean_divider + \") \" + divider\n",
    "\n",
    "    if column_sort:\n",
    "        new_names = sorted(new_names)\n",
    "    if row_sort:\n",
    "        row_names = sorted(all_values.keys())\n",
    "    else:\n",
    "        row_names = all_values.keys()\n",
    "\n",
    "    with open(target_path, 'w') as f:\n",
    "        for name in new_names:\n",
    "            f.write(\"\\t\")\n",
    "            f.write(name)\n",
    "        f.write(\"\\n\")\n",
    "        for key in row_names:\n",
    "            f.write(key)\n",
    "            for name in new_names:\n",
    "                f.write(\"\\t\")\n",
    "                f.write(all_values[key][name])\n",
    "            f.write(\"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If the required softwares are not in the PATH you can manually set their location here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tagdust2_path = 'tagdust'\n",
    "bwa_path = '/home/baker/bwa/bwa'\n",
    "samtools_path = 'samtools'\n",
    "paired_bam_to_bed12_path = '/home/baker/pairedBamToBed12/bin/pairedBamToBed12'\n",
    "umicountFP_path = 'umicountFP'\n",
    "syncpairs_path = 'syncpairs'\n",
    "bedToBam_path = 'bedToBam'\n",
    "featureCount_path = '/home/baker/Downloads/subread-1.5.0-p1-source/bin/featureCounts'\n",
    "Trimmo_path = '/home/baker/my-hydra-share/Packages/Trimmomatic-0.35/trimmomatic-0.35.jar'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Path to the reference genome you want to align your reads against"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#ref_genome = '/home/baker/my-mm10-index-share/Mus_musculus.GRCm38.71.fa'\n",
    "ref_genome = '/home/baker/my-mm10-index-share/bwa/bwa_mm10_random_chrM_chrUn'\n",
    "bacterial_ref_genome = '/home/baker/my-genomeIndex-share/L.monocytogenes/bwa/genome'\n",
    "Spike_in_genome = '/home/baker/my-hydra-share/Packages/Spike-In_Index/ERCC92'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "softwares = {    \n",
    "    'bwa': bwa_path,\n",
    "    'tagdust': tagdust2_path,\n",
    "    'syncpairs': syncpairs_path,\n",
    "    'samtools': samtools_path,\n",
    "    'pairedBamToBed12': paired_bam_to_bed12_path,\n",
    "    'umicountFP': umicountFP_path,\n",
    "    'bedToBam': bedToBam_path,\n",
    "    'featureCount':featureCount_path,\n",
    "    'trimmomatic':Trimmo_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The name of the output folders for each command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "output_folders = [ 'tagdust_r1', 'unzip_r2'                    # Demultiplexed R1, unziped R2\n",
    "                 , 'extracted_r1', 'extracted_r2'              # Synced R1 and R2\n",
    "                 , 'cleaned_reads', 'cleaned_r1', 'cleaned_r2' # rRNA reads removed\n",
    "                 , 'r1_sai', 'r2_sai', 'sampe'                 # Intermediate files from BWA\n",
    "                 , 'r1_bac_sai', 'r2_bac_sai', 'sampe_bac'     # Intermediate files from BWA for Bacterial\n",
    "                 , 'genome_mapped', 'properly_paired'          # Final output in BAM format\n",
    "                 , 'cagescan_pairs', 'cagescan_fragments'      # Final output in BED12 format\n",
    "                 , 'htseq_output', 'ERCC_count'                # Htseq-count and ERCC-Spike in counts\n",
    "                 , 'bedToBam_output','featureCount_output'\n",
    "                 , 'sampeBacteria', 'bacterial_genome_mapped'\n",
    "                 , 'Spikesampe', 'Spike_genome_mapped'\n",
    "                 , 'Spike_properly_paired', 'Spike_cagescan_pairs'\n",
    "                 , 'Spike_cagescan_fragments', 'Fortrimmo_r1', 'Fortrimmo_r2'\n",
    "                 , 'bedToBam_output_Spike'\n",
    "                 ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create the folders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for folder in output_folders:\n",
    "    os.makedirs(os.path.join(outputFolder, folder))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The actual command to run. See the [tutorial](https://github.com/Population-Transcriptomics/C1-CAGE-preview/blob/master/tutorial.md) section for more details about each command"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Samtools Flags\n",
    "\n",
    "Flag Description (The way I used):\n",
    "- 0x0001 Paired in sequencing\n",
    "- __0x0002 Mapped in proper read pair__\n",
    "- 0x0004 Query unmapped\n",
    "- 0x0008 Mate unmapped\n",
    "- 0x0010 Strand of the query (1 for reverse)\n",
    "- 0x0020 Strand of the mate (1 for reverse)\n",
    "- 0x0040 First read in the pair\n",
    "- 0x0080 Second read in the pair\n",
    "- __0x0100 Secondary alignment__\n",
    "- 0x0200 QC failure\n",
    "- 0x0400 Optical or PCR duplicates\n",
    "\n",
    "Flag Description (The way mentioned in samtools help):\n",
    "- 0x1\tPAIRED\tpaired-end (or multiple-segment) sequencing technology\n",
    "- 0x2\tPROPER_PAIR\teach segment properly aligned according to the aligner\n",
    "- 0x4\tUNMAP\tsegment unmapped\n",
    "- 0x8\tMUNMAP\tnext segment in the template unmapped\n",
    "- 0x10\tREVERSE\tSEQ is reverse complemented\n",
    "- 0x20\tMREVERSE\tSEQ of the next segment in the template is reverse complemented\n",
    "- 0x40\tREAD1\tthe first segment in the template\n",
    "- 0x80\tREAD2\tthe last segment in the template\n",
    "- 0x100\tSECONDARY\tsecondary alignment\n",
    "- 0x200\tQCFAIL\tnot passing quality controls\n",
    "- 0x400\tDUP\tPCR or optical duplicate\n",
    "- 0x800\tSUPPLEMENTARY\tsupplementary alignment\n",
    "\n",
    "If you want to extract reads that mapped only to a certain strand of the reference, you can use the -F and -f flags in samtools:\n",
    "\n",
    "__for positive strand__<br>\n",
    "samtools view -F 0x10 -b input.bam > postiveStrand.sam\n",
    "\n",
    "__for negative strand __<br>\n",
    "samtools view -f 0x10 -b input.bam > negativeStrand.sam \n",
    "\n",
    "The 0x10 flag of the bitwise flags indicates whether the read is reverse complemented or not. If 0x10 is set, then the read mapped to the reverse strand. If it is not set, then the read mapped to the positive strand.\n",
    "\n",
    "The -F parameter with 0x10 will skip anything that has 0x10 set, meaning it will skip anything that mapped to the reverse strand. The -f parameter will only keep anything that has 0x10 set.\n",
    "\n",
    "__Samtools view options__\n",
    "\n",
    " OPTIONS:\n",
    "\n",
    "-b\n",
    "\n",
    "    Output in the BAM format. \n",
    "-C\n",
    "\n",
    "    Output in the CRAM format (requires -T). \n",
    "-1\n",
    "\n",
    "    Enable fast BAM compression (implies -b). \n",
    "-u\n",
    "\n",
    "    Output uncompressed BAM. This option saves time spent on compression/decompression and is thus preferred when the output is piped to another samtools command. \n",
    "-h\n",
    "\n",
    "    Include the header in the output. \n",
    "-H\n",
    "\n",
    "    Output the header only. \n",
    "-c\n",
    "\n",
    "    Instead of printing the alignments, only count them and print the total number. All filter options, such as -f, -F, and -q, are taken into account. \n",
    "-?\n",
    "\n",
    "    Output long help and exit immediately. \n",
    "-o FILE\n",
    "\n",
    "    Output to FILE [stdout]. \n",
    "-U FILE\n",
    "\n",
    "    Write alignments that are not selected by the various filter options to FILE. When this option is used, all alignments (or all alignments intersecting the regions specified) are written to either the output file or this file, but never both. \n",
    "-t FILE\n",
    "\n",
    "    A tab-delimited FILE. Each line must contain the reference name in the first column and the length of the reference in the second column, with one line for each distinct reference. Any additional fields beyond the second column are ignored. This file also defines the order of the reference sequences in sorting. If you run: `samtools faidx <ref.fa>', the resulting index file <ref.fa>.fai can be used as this FILE. \n",
    "-T FILE\n",
    "\n",
    "    A FASTA format reference FILE, optionally compressed by bgzip and ideally indexed by samtools faidx. If an index is not present, one will be generated for you. \n",
    "-L FILE\n",
    "\n",
    "    Only output alignments overlapping the input BED FILE [null]. \n",
    "-r STR\n",
    "\n",
    "    Only output alignments in read group STR [null]. \n",
    "-R FILE\n",
    "\n",
    "    Output alignments in read groups listed in FILE [null]. \n",
    "-q INT\n",
    "\n",
    "    Skip alignments with MAPQ smaller than INT [0]. \n",
    "-l STR\n",
    "\n",
    "    Only output alignments in library STR [null]. \n",
    "-m INT\n",
    "\n",
    "    Only output alignments with number of CIGAR bases consuming query sequence ≥ INT [0] \n",
    "-f INT\n",
    "\n",
    "    Only output alignments with all bits set in INT present in the FLAG field. INT can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/) or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0]. \n",
    "-F INT\n",
    "\n",
    "    Do not output alignments with any bits set in INT present in the FLAG field. INT can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/) or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0]. \n",
    "-x STR\n",
    "\n",
    "    Read tag to exclude from output (repeatable) [null] \n",
    "-B\n",
    "\n",
    "    Collapse the backward CIGAR operation. \n",
    "-s FLOAT\n",
    "\n",
    "    Integer part is used to seed the random number generator [0]. Part after the decimal point sets the fraction of templates/pairs to subsample [no subsampling]. \n",
    "-@ INT\n",
    "\n",
    "    Number of BAM compression threads to use in addition to main thread [0]. \n",
    "-S\n",
    "\n",
    "    Ignored for compatibility with previous samtools versions. Previously this option was required if input was in SAM format, but now the correct format is automatically detected by examining the first few characters of input. \n",
    "\n",
    "### bwa mem -M option\n",
    "\"Changed the way a chimeric alignment is reported (conforming to the upcoming\n",
    "SAM spec v1.5). With 0.7.5, if the read has a chimeric alignment, the paired\n",
    "or the top hit uses soft clipping and is marked with neither 0x800 nor 0x100\n",
    "bits. All the other hits part of the chimeric alignment will use hard\n",
    "clipping and be marked with 0x800 if option \"-M\" is not in use, or marked\n",
    "with 0x100 otherwise __(Seondary Alignment)__.\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Formatting the fasta file for BWA index generation__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "for seq_record in SeqIO.parse(\"/home/baker/my-hydra-share/Packages/Spike-In_Index/ERCC92.fa\",\"fasta\"):\n",
    "    f = open(\"/home/baker/my-hydra-share/Packages/Spike-In_Index/ERCC92.fa.modified\", 'a')\n",
    "    f.write('>' + str(seq_record.id) + '\\n')\n",
    "    f.write(str(seq_record.seq) + '\\n')\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sequences to generate the BWA index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "bwa index -a bwtsw reference.fa\n",
    "samtools faidx reference.fa \n",
    "\n",
    "\n",
    "java -jar picard.jar CreateSequenceDictionary \\\n",
    "    REFERENCE=reference.fa \\ \n",
    "    OUTPUT=reference.dict "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cmds = [\n",
    "    \n",
    "    '{tagdust} -t8 -o {tagdust_r1} -1 F:NNNNNNNN -2 S:TATAGGG -3 R:N {r1_input}',\n",
    "    \n",
    "    'gunzip -c {r2_input} > {unzip_r2}.fq',\n",
    "    \n",
    "    '{syncpairs} {tagdust_r1}.fq {unzip_r2}.fq {extracted_r1}.fq {extracted_r2}.fq',\n",
    "    \n",
    "    '{tagdust} -arch SimpleArchitecture.txt -ref ercc_and_mouse_rRNA_and_tagdust.fa -o {cleaned_reads} {extracted_r1}.fq {extracted_r2}.fq',\n",
    "    \n",
    "    'cp {cleaned_reads}_READ1.fq {cleaned_r1}.fq',\n",
    "    \n",
    "    'cp {cleaned_reads}_READ2.fq {cleaned_r2}.fq',\n",
    "    \n",
    "    '{bwa} aln -t8 -n 0 {ref_genome} {cleaned_r1}.fq > {r1_sai}.sai',\n",
    "    \n",
    "    '{bwa} aln -t8 -n 0 {ref_genome} {cleaned_r2}.fq > {r2_sai}.sai',\n",
    "    \n",
    "    '{bwa} sampe -a 2000000 {ref_genome} {r1_sai}.sai {r2_sai}.sai {cleaned_r1}.fq {cleaned_r2}.fq > {sampe}.sam',\n",
    "    \n",
    "    '{bwa} aln -t8 -n 0 {bacterial_ref_genome} {cleaned_r1}.fq > {r1_bac_sai}.sai',\n",
    "    \n",
    "    '{bwa} aln -t8 -n 0 {bacterial_ref_genome} {cleaned_r2}.fq > {r2_bac_sai}.sai',\n",
    "    \n",
    "    '{bwa} sampe -a 2000000 {bacterial_ref_genome} {r1_bac_sai}.sai {r2_bac_sai}.sai {cleaned_r1}.fq {cleaned_r2}.fq > {sampeBacteria}.sam',\n",
    "    \n",
    "    '{samtools} view -uSo - {sampe}.sam | {samtools} sort - {genome_mapped}',\n",
    "    \n",
    "    '{samtools} flagstat {genome_mapped}.bam >{genome_mapped}.Logfile',\n",
    "    \n",
    "    '{samtools} view -uSo - {sampeBacteria}.sam | {samtools} sort - {bacterial_genome_mapped}',\n",
    "    \n",
    "    '{samtools} flagstat {bacterial_genome_mapped}.bam >{bacterial_genome_mapped}.Logfile',\n",
    "    \n",
    "    '{samtools} view -f 0x0002 -F 0x0100 -uo - {genome_mapped}.bam | {samtools} sort -n - {properly_paired}',\n",
    "    \n",
    "    '{pairedBamToBed12} -i {properly_paired}.bam > {cagescan_pairs}.bed',\n",
    "    \n",
    "    '{umicountFP} -f {cagescan_pairs}.bed > {cagescan_fragments}.bed',\n",
    "    \n",
    "    '{bedToBam} -i {cagescan_fragments}.bed -g /home/baker/my-hydra-share/Packages/MouseChromosome/chrom.sizes > {bedToBam_output}.bam',\n",
    "    \n",
    "    '{samtools} sort {bedToBam_output}.bam {bedToBam_output}_sorted',\n",
    "    \n",
    "    '{samtools} view -bh {bedToBam_output}_sorted.bam > {bedToBam_output}_sorted_header.bam'    \n",
    "    \n",
    "    \n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the reads. Here we assume that the reads are in the current directory, in a folder named following the MiSeq run id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "root, folders, files = os.walk(intputFolder).next()\n",
    "\n",
    "files = [f for f in files if not f.startswith('.')] #remove hidden files if there exist\n",
    "reads1 = sorted([os.path.join(root, f) for f in files if 'R1' in f])\n",
    "reads2 = sorted([os.path.join(root, f) for f in files if 'R2' in f])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C01_S17_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C02_S9_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C03_S1_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C04_S65_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C05_S57_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C06_S49_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C07_S18_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C08_S10_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C09_S2_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C10_S66_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C11_S58_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C12_S50_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C13_S19_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C14_S11_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C15_S3_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C16_S67_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C17_S59_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C18_S51_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C19_S20_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C20_S12_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C21_S4_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C22_S68_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C23_S60_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C24_S52_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C25_S5_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C26_S13_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C27_S21_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C28_S53_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C29_S61_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C30_S69_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C31_S6_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C32_S14_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C33_S22_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C34_S54_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C35_S62_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C36_S70_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C37_S7_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C38_S15_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C39_S23_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C40_S55_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C41_S63_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C42_S71_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C43_S8_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C44_S16_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C45_S24_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C46_S56_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C47_S64_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C48_S72_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C49_S25_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C50_S33_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C51_S41_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C52_S73_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C53_S81_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C54_S89_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C55_S26_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C56_S34_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C57_S42_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C58_S74_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C59_S82_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C60_S90_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C61_S27_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C62_S35_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C63_S43_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C64_S75_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C65_S83_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C66_S91_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C67_S28_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C68_S36_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C69_S44_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C70_S76_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C71_S84_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C72_S92_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C73_S45_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C74_S37_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C75_S29_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C76_S93_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C77_S85_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C78_S77_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C79_S46_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C80_S38_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C81_S30_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C82_S94_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C83_S86_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C84_S78_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C85_S47_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C86_S39_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C87_S31_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C88_S95_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C89_S87_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C90_S79_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C91_S48_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C92_S40_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C93_S32_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C94_S96_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C95_S88_R1_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C96_S80_R1_001.fastq.gz'] ['/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C01_S17_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C02_S9_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C03_S1_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C04_S65_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C05_S57_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C06_S49_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C07_S18_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C08_S10_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C09_S2_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C10_S66_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C11_S58_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C12_S50_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C13_S19_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C14_S11_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C15_S3_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C16_S67_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C17_S59_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C18_S51_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C19_S20_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C20_S12_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C21_S4_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C22_S68_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C23_S60_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C24_S52_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C25_S5_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C26_S13_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C27_S21_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C28_S53_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C29_S61_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C30_S69_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C31_S6_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C32_S14_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C33_S22_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C34_S54_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C35_S62_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C36_S70_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C37_S7_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C38_S15_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C39_S23_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C40_S55_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C41_S63_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C42_S71_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C43_S8_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C44_S16_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C45_S24_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C46_S56_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C47_S64_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C48_S72_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C49_S25_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C50_S33_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C51_S41_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C52_S73_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C53_S81_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C54_S89_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C55_S26_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C56_S34_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C57_S42_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C58_S74_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C59_S82_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C60_S90_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C61_S27_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C62_S35_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C63_S43_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C64_S75_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C65_S83_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C66_S91_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C67_S28_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C68_S36_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C69_S44_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C70_S76_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C71_S84_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C72_S92_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C73_S45_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C74_S37_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C75_S29_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C76_S93_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C77_S85_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C78_S77_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C79_S46_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C80_S38_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C81_S30_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C82_S94_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C83_S86_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C84_S78_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C85_S47_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C86_S39_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C87_S31_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C88_S95_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C89_S87_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C90_S79_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C91_S48_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C92_S40_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C93_S32_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C94_S96_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C95_S88_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/C96_S80_R2_001.fastq.gz', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/miseq_size_R2', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/miseq_size_R2.e266525', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/miseq_size_R2.o266525', '/home/baker/my-single-cell-share/James_Low_Bacteria_Miseq_Size_Trimmed/miseq_size_R2~']\n"
     ]
    }
   ],
   "source": [
    "print reads1, reads2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run the commands for all the pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-19-8b76c8853df6>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      4\u001b[0m     \u001b[1;32mfor\u001b[0m \u001b[0mcmd\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mcmds\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m         \u001b[1;31m#print cmd.format(**args)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 6\u001b[1;33m         \u001b[0msubprocess\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m**\u001b[0m\u001b[0margs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mpreexec_fn\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mlambda\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0msignal\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msignal\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msignal\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mSIGPIPE\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0msignal\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mSIG_DFL\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mshell\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m/usr/lib/python2.7/subprocess.pyc\u001b[0m in \u001b[0;36mcall\u001b[1;34m(*popenargs, **kwargs)\u001b[0m\n\u001b[0;32m    520\u001b[0m     \u001b[0mretcode\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"ls\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"-l\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    521\u001b[0m     \"\"\"\n\u001b[1;32m--> 522\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0mPopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0mpopenargs\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    523\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    524\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/usr/lib/python2.7/subprocess.pyc\u001b[0m in \u001b[0;36mwait\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m   1374\u001b[0m             \u001b[1;32mwhile\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreturncode\u001b[0m \u001b[1;32mis\u001b[0m \u001b[0mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1375\u001b[0m                 \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1376\u001b[1;33m                     \u001b[0mpid\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0msts\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0m_eintr_retry_call\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwaitpid\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpid\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1377\u001b[0m                 \u001b[1;32mexcept\u001b[0m \u001b[0mOSError\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1378\u001b[0m                     \u001b[1;32mif\u001b[0m \u001b[0me\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0merrno\u001b[0m \u001b[1;33m!=\u001b[0m \u001b[0merrno\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mECHILD\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/usr/lib/python2.7/subprocess.pyc\u001b[0m in \u001b[0;36m_eintr_retry_call\u001b[1;34m(func, *args)\u001b[0m\n\u001b[0;32m    474\u001b[0m     \u001b[1;32mwhile\u001b[0m \u001b[0mTrue\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    475\u001b[0m         \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 476\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0margs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    477\u001b[0m         \u001b[1;32mexcept\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mOSError\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mIOError\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    478\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0me\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0merrno\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0merrno\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mEINTR\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "for read1, read2 in zip(reads1, reads2):\n",
    "    args = get_args(read1, read2, ref_genome, bacterial_ref_genome, Spike_in_genome, output_folders)\n",
    "    args = dict(args, **softwares)    \n",
    "    for cmd in cmds:        \n",
    "        #print cmd.format(**args)\n",
    "        subprocess.call(cmd.format(**args), preexec_fn=lambda: signal.signal(signal.SIGPIPE, signal.SIG_DFL), shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FeatureCountFolder = outputFolder + 'featureCount_output/'\n",
    "bedToBamOutputFolder = outputFolder + 'bedToBam_output/'\n",
    "root, folders, files = os.walk(bedToBamOutputFolder).next()\n",
    "files = [os.path.join(root, f) for f in files if f.endswith('_sorted_header.bam')]\n",
    "files = sorted(files)\n",
    "featureCountCmd = '/home/baker/Downloads/subread-1.5.0-p1-source/bin/./featureCounts -p -t exon -g gene_id -a /home/baker/my-mm10-index-share/annotation/Gencode-M5/gencode.vM5.annotation.gtf -o '+ FeatureCountFolder+'AllFeatureCounts.txt '\n",
    "for f in files:   \n",
    "    featureCountCmd = featureCountCmd + f + ' '    \n",
    "subprocess.call(featureCountCmd, shell=True)\n",
    "#print featureCountCmd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bacterial genome count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/baker/Downloads/subread-1.5.0-p1-source/bin/./featureCounts -p -F gff -g Name -a /home/baker/l.monocytogenes_genome/AllGtf/All.gff -o /home/baker/my-scratch-share/James_Bagnall_Low_Bacteria_Workflow_Riken_Misesq_length/output/featureCount_output/BacterialFeatureCounts.txt \n"
     ]
    }
   ],
   "source": [
    "FeatureCountFolder = outputFolder + 'featureCount_output/'\n",
    "bedToBamOutputFolder = outputFolder + 'bacterial_genome_mapped/'\n",
    "root, folders, files = os.walk(bedToBamOutputFolder).next()\n",
    "files = [os.path.join(root, f) for f in files if f.endswith('_sorted_header.bam')]\n",
    "files = sorted(files)\n",
    "featureCountCmd = '/home/baker/Downloads/subread-1.5.0-p1-source/bin/./featureCounts -p -F gff -g Name -a /home/baker/l.monocytogenes_genome/AllGtf/All.gff -o '+ FeatureCountFolder+'BacterialFeatureCounts.txt '\n",
    "for f in files:   \n",
    "    featureCountCmd = featureCountCmd + f + ' '    \n",
    "#subprocess.call(featureCountCmd, shell=True)\n",
    "print featureCountCmd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Extracting the counts mapped with Bacterial Genome__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "BacterialGenomeFolder = outputFolder + 'bacterial_genome_mapped/'\n",
    "root, folders, files = os.walk(BacterialGenomeFolder).next()\n",
    "files = [os.path.join(root, f) for f in files if f.endswith('.Logfile')]\n",
    "files = sorted(files)\n",
    "\n",
    "BacterialCountFile = BacterialGenomeFolder + 'BacteriaCount.txt'\n",
    "BacterialCount = \"CellName\"+ \"\\t\" + \"Counts\" + \"\\t\" + \"PercentMapped \\n\"\n",
    "\n",
    "for f in files:\n",
    "    file_Bacteria = open(f,'rb')\n",
    "    for line in file_Bacteria:\n",
    "        if \"mapped (\" in line :\n",
    "            BacterialCount = BacterialCount + f.rsplit('/', 1)[1].rsplit('_',3)[0] + '\\t' + line.rsplit('+', 1)[0] + '\\t' + line.rsplit('(', 1)[1].split('%', 1)[0]+'\\n' \n",
    "        \n",
    "\n",
    "target = open(BacterialCountFile, 'w')\n",
    "target.write(BacterialCount)\n",
    "target.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Extracting the counts mapped with Mouse Genome__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "MouseGenomeFolder = outputFolder + 'genome_mapped/'\n",
    "root, folders, files = os.walk(MouseGenomeFolder).next()\n",
    "files = [os.path.join(root, f) for f in files if f.endswith('.Logfile')]\n",
    "files = sorted(files)\n",
    "\n",
    "MouseCountFile = MouseGenomeFolder + 'MouseCount.txt'\n",
    "MouseCount = \"CellName\"+ \"\\t\" + \"TotalReads\" + \"\\t\" + \"Mapped\" + \"\\t\" + \"PercentMapped \\n\"\n",
    "\n",
    "for f in files:\n",
    "    file_Mouse = open(f,'rb')\n",
    "    line = file_Mouse.readline().rstrip()\n",
    "    MouseCount = MouseCount + f.rsplit('/', 1)[1].rsplit('_',3)[0] + '\\t' + line.split(' + ', 1)[0]\n",
    "    for line in file_Mouse:        \n",
    "        if \"mapped (\" in line :\n",
    "            MouseCount = MouseCount + '\\t' + line.rsplit('+', 1)[0] + '\\t' + line.rsplit('(', 1)[1].split('%', 1)[0]+'\\n' \n",
    "        \n",
    "\n",
    "target = open(MouseCountFile, 'w')\n",
    "target.write(MouseCount)\n",
    "target.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generate the level1 file (__We are not using it at the moment__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genomeMapped = outputFolder + 'genome_mapped/'\n",
    "root, folders, files = os.walk(genomeMapped).next()\n",
    "files = [os.path.join(root, f) for f in files if f.endswith('bam')]\n",
    "level1 = 'python /home/baker/PromoterPipeline_20150516/level1.py -o /home/baker/my-scratch-share/James_Bagnall_Single_cell_rna_seq_Feature_Count/output/mylevel1file.l1.osc.gz -f 0x0042 -F 0x0104 --fingerprint {files}'.format(files=' '.join(files))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "subprocess.call(level1, shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate logs (triplet)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we generate four summary files that will be used for [QC](https://github.com/Population-Transcriptomics/C1-CAGE-preview/blob/master/QC.md) and place them in the 'output' directory. \n",
    "\n",
    "1.  mapped.log: The number of mapped reads per cell\n",
    "2.  extracted.log: The number of remaining reads after filtering for ribosomal DNA and unreadable UMIs\n",
    "3.  filtered.log: The detailed number of ribosomal DNA extracted per cell\n",
    "4.  transcript_count.log: The exact number of unique transcprit per cell\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "total_cmd       = \"grep 'total input reads'      {tagdust_r1}_logfile.txt    | cut -f 2\"\n",
    "extracted_cmd   = \"grep 'successfully extracted' {tagdust_r1}_logfile.txt    | cut -f 2\"\n",
    "cleaned_cmd     = \"grep 'successfully extracted' {cleaned_reads}_logfile.txt | cut -f 2\"\n",
    "rdna_cmd        = \"grep 'ribosomal'              {cleaned_reads}_logfile.txt | cut -f 2\"\n",
    "spikes_cmd      = \"grep 'ERCC'                   {cleaned_reads}_logfile.txt | cut -f 2 | paste -s -d+ - | bc\"\n",
    "mapped_cmd      = \"{samtools} view -u -f 0x40 {genome_mapped}.bam | {samtools} flagstat - | grep 'mapped (' | cut -f 1 -d ' '\"\n",
    "properpairs_cmd = \"{samtools} view -u -f 0x42 {genome_mapped}.bam | {samtools} flagstat - | grep 'mapped (' | cut -f 1 -d ' '\"\n",
    "counts_cmd      = \"wc -l {cagescan_fragments}.bed | cut -f 1 -d ' '\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#remove _R1 from the file's name\n",
    "custom_rename = lambda x: x.replace('_R1', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "total, cleaned, extracted, rdna, spikes, mapped, properpairs, counts = ([], [], [], [], [], [], [], [])\n",
    "\n",
    "def run_qc(command, dest, keyword):\n",
    "    output = subprocess.check_output(command.format(**args), shell=True).strip().decode()\n",
    "    dest.append([keyword, custom_rename(r1_shortname), output])\n",
    "\n",
    "for read1 in reads1:\n",
    "    r1_shortname = remove_extension(os.path.basename(read1))\n",
    "    \n",
    "    args = {'tagdust_r1':         os.path.join(outputFolder, 'tagdust_r1',         r1_shortname),\n",
    "            'genome_mapped':      os.path.join(outputFolder, 'genome_mapped',      r1_shortname),\n",
    "            'cagescan_fragments': os.path.join(outputFolder, 'cagescan_fragments', r1_shortname),\n",
    "            'cleaned_reads':      os.path.join(outputFolder, 'cleaned_reads',      r1_shortname),\n",
    "            'samtools': samtools_path}\n",
    "    \n",
    "    run_qc(total_cmd,       total,       'total')\n",
    "    run_qc(extracted_cmd,   extracted,   'extracted')\n",
    "    run_qc(cleaned_cmd,     cleaned,     'cleaned')\n",
    "    run_qc(rdna_cmd,        rdna,        'rdna')\n",
    "    run_qc(spikes_cmd,      spikes,      'spikes')\n",
    "    run_qc(mapped_cmd,      mapped,      'mapped')\n",
    "    run_qc(properpairs_cmd, properpairs, 'properpairs')\n",
    "    run_qc(counts_cmd,      counts,      'counts')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def write_logs(file_dest, results):\n",
    "    with open(file_dest, 'w') as handler:\n",
    "        writer = csv.writer(handler, delimiter='\\t')\n",
    "        writer.writerows(results)\n",
    "\n",
    "write_logs(outputFolder + 'total.log',       total)\n",
    "write_logs(outputFolder + 'cleaned.log',     cleaned)\n",
    "write_logs(outputFolder + 'extracted.log',   extracted)\n",
    "write_logs(outputFolder + 'rdna.log',        rdna)\n",
    "write_logs(outputFolder + 'spikes.log',      spikes)\n",
    "write_logs(outputFolder + 'mapped.log',      mapped)\n",
    "write_logs(outputFolder + 'properpairs.log', properpairs)\n",
    "write_logs(outputFolder + 'counts.log',      counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Generating commands for htseq-count runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "log_files = \"\"\n",
    "properlyPairedFolder = outputFolder + 'properly_paired/'\n",
    "root, folders, files = os.walk(properlyPairedFolder).next()\n",
    "htseq_command=\"\"\n",
    "htseqFile = outputFolder + 'htseq_command.txt'\n",
    "for f in files:   \n",
    "    htseq_command_file = open(htseqFile, 'a')\n",
    "    #htseq_command = \"qsub run_htseq-count output/properly_paired/\" + f + \" output/htseq_output/\"+f.split('_', 1)[0]+\"_htseq_count.txt\\n\"\n",
    "    htseq_command =\"sh run_htseq-count_desktop output/properly_paired/\" + f + \" output/htseq_output/\"+f.split('_', 1)[0]+\"_htseq_count.txt\\n\"\n",
    "    htseq_command_file.writelines(htseq_command)\n",
    "    htseq_command_file.close\n",
    "        #htseq_command.append([\"qsub run_htseq-count\", f])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Generating command for ERCC molecule count with -u option__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "properlyPairedFolder = outputFolder + 'properly_paired/'\n",
    "root, folders, files = os.walk(properlyPairedFolder).next()\n",
    "htseq_command=\"\"\n",
    "htseqFile = outputFolder + 'ERCC_command.txt'\n",
    "for f in files:   \n",
    "    htseq_command_file = open(htseqFile, 'a')\n",
    "    htseq_command = \"qsub -pe smp.pe 2 run_htseq_ERCC-count output/properly_paired/\" + f + \" output/ERCC_Reads/\"+f.split('_', 1)[0]+\"_ERCC_count.txt\\n\"\n",
    "    htseq_command_file.writelines(htseq_command)\n",
    "    htseq_command_file.close"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Combinging htseq count files and ERRC count files__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following command counts ERCC reads from the tagdust output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cleanedReadsFolder = outputFolder + 'cleaned_reads/'\n",
    "root, folders, files = os.walk(cleanedReadsFolder).next()\n",
    "files = [os.path.join(\"\", f) for f in files if f.endswith('txt')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for f in files:    \n",
    "    ERCC_val = ERCC_finder(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ERCCFolder = outputFolder + 'ERCC_count/'\n",
    "ERCCFile = outputFolder + 'AllERCC_Counts.tsv'\n",
    "do_walk(ERCCFolder, 'ERCC_count_', ERCCFile, onerror=None, followlinks=False, verbose=True, divider=None, sort=\"both\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This command combines the htseq counts from hydra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "HTSEQFolder = outputFolder + 'htseq_output/'\n",
    "HTSEqFile = outputFolder + 'AllHtseqCounts.tsv'\n",
    "do_walk(HTSEQFolder, 'htseq_count.txt', HTSEqFile, onerror=None, followlinks=False, verbose=True, divider=None, sort=\"both\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "CountData <- read.table('/home/baker/my-scratch-share/James_Bagnall_Single_cell_rna_seq_BWA_mem_Feature_Count/output/AllHtseqCounts.tsv',header=TRUE, sep='\\t',row.names=1)\n",
    "CellNumber <- dim(CountData)[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "head(CountData)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "# summary_data <- data.frame(CellNumber= character(CellNumber), Assigned_Reads=numeric(CellNumber),alignment_not_unique=numeric(CellNumber),ambiguous=numeric(CellNumber),no_feature=numeric(CellNumber),not_aligned=numeric(CellNumber),too_low_aQual=numeric(CellNumber ) )\n",
    "summary_data <- data.frame(CellNumber= character(CellNumber), Assigned_Reads=numeric(CellNumber))\n",
    "summary_data$CellNumber <- colnames(CountData)\n",
    "summary_data$Assigned_Reads <- apply(CountData[grep(\"ENSMUSG\", rownames(CountData)), ],2,sum) \n",
    "#summary_data$alignment_not_unique <- t(CountData[grep(\"alignment_not_unique\", rownames(CountData)), ])\n",
    "#summary_data$ambiguous <- t(CountData[grep(\"ambiguous\", rownames(CountData)), ])\n",
    "#summary_data$no_feature <- t(CountData[grep(\"no_feature\", rownames(CountData)), ])\n",
    "#summary_data$not_aligned <- t(CountData[grep(\"not_aligned\", rownames(CountData)), ])\n",
    "#summary_data$too_low_aQual <- t(CountData[grep(\"too_low_aQual\", rownames(CountData)), ])\n",
    "\n",
    "#apply(CountData[grep(\"ENSMUSG\", rownames(CountData)), ],2,sum) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "summary_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "ERCCData <- read.table('/home/baker/my-scratch-share/James_Bagnall_Single_cell_rna_seq_BWA_mem_Feature_Count/output/AllERCC_Counts.tsv',header=TRUE, sep='\\t',row.names=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "head(ERCCData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "ERCC_reads <- apply(ERCCData[grep(\"ERCC\", rownames(ERCCData)), ],2,sum)\n",
    "ercc_reads_plot <- ggplot(ERCC_reads,\n",
    "                   aes(x = factor(colnames(ERCCData)), y = ERCC_reads,\n",
    "                   fill = factor(colnames(ERCCData))), height = 600, width = 2000) +\n",
    "                   geom_violin(alpha = .5) + \n",
    "                   geom_boxplot(alpha = .01, width = .2, position = position_dodge(width = .9)) +\n",
    "                   labs(x = \"\", y = \"ERCC molecules\", title = \"Total ERCC molecule counts per cell\") +\n",
    "                   theme(axis.text.x = element_text(hjust=1, angle = 45))\n",
    "\n",
    "#theme_set(theme_bw(base_size = 16))\n",
    "print(ercc_reads_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "molecules_filter <- ERCCData\n",
    "molecules_filter <- molecules_filter[rowSums(molecules_filter)>0,]\n",
    "dim(molecules_filter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filtering the Cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R -o cdNonFilter\n",
    "# omit the non-mapped gene information\n",
    "cdNonFilter = CountData[1:(dim(CountData)[1]-5),]\n",
    "# omit genes that are never detected\n",
    "cdNonFilter <- cdNonFilter[rowSums(cdNonFilter)>0, ]\n",
    "# omit cells with very poor coverage\n",
    "cdNonFilter <- cdNonFilter[, colSums(cdNonFilter)>1e4]\n",
    "CellNumber <- dim(cdNonFilter)[2]\n",
    "print(dim(cdNonFilter))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "GeneAnnotation <- read.table('/home/baker/my-scratch-share/James_Bagnall_Single_cell_rna_seq_V3/output/GeneAnnotation.csv', header=TRUE, sep=\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "FilteredAnnotatedData <- AnnotatedCountData\n",
    "FilteredAnnotatedData <- FilteredAnnotatedData[rowSums(FilteredAnnotatedData[,12:dim(FilteredAnnotatedData)[2]]) > 0, ]\n",
    "write.csv(FilteredAnnotatedData, \"/home/baker/my-scratch-share/James_Bagnall_Single_cell_rna_seq_V3/output/Filtered_Annotated_read_count_data.csv\", row.names=T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "spike <- read.table(\"/home/baker/my-scratch-share/James_Bagnall_Single_cell_rna_seq_V3/output/expected-ercc-molecules.txt\", header=TRUE, sep = \"\\t\", stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "spike_input <- data.frame(Spike=spike$id[spike$id %in% rownames(molecules_filter)], Read=spike$ercc_molecules_well[spike$id %in% rownames(molecules_filter)])\n",
    "#names(spike_input) <- spike$id[spike$id %in% rownames(molecules_filter)]\n",
    "#spike_input <- data.frame(spike_input[order(names(spike_input))])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Short Data Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "  X        V1            V2                                       V3    V4\n",
       "1 1 mMA031304         Dscr1 Mus musculus Down syndrome critical regi 2.493\n",
       "2 2 mMA032268          Irf1 Mus musculus interferon regulatory facto 2.348\n",
       "3 3 mMA032453          Irf1 Mus musculus interferon regulatory facto 3.365\n",
       "4 4 mMA033839 B430217B02Rik Mus musculus RIKEN cDNA B430217B02 gene  2.929\n",
       "5 5 mMA033942         Peli1    Mus musculus pellino 1 (Peli1), mRNA. 2.289\n",
       "6 6 mMA034042         Oasl1 Mus musculus 2'-5' oligoadenylate synthe 3.071\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "ListeriaAllAnnotated <- read.csv('ListeriaAllAnnotated.csv')\n",
    "head(ListeriaAllAnnotated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                 Geneid GeneName C01 C02 C03 C04 C05 C06 C07 C08 C09 C10 C11\n",
       "1  ENSMUSG00000000001.4    Gnai3   0   0   0   0   0   0   0   0   0   0   0\n",
       "2 ENSMUSG00000000003.12     Pbsn   0   0   0   0   0   0   0   0   0   0   0\n",
       "3 ENSMUSG00000000028.11    Cdc45   0   0   0   0   0   0   0   0   0   0   0\n",
       "4 ENSMUSG00000000031.12      H19   0   0   0   0   0   0   0   0   0   0   0\n",
       "5 ENSMUSG00000000037.13    Scml2   0   0   0   0   0   0   0   0   0   0   0\n",
       "6  ENSMUSG00000000049.8     Apoh   0   0   0   0   0   0   0   0   0   0   0\n",
       "  C12 C13 C14 C15 C16 C17 C18 C19 C20 C21 C22 C23 C24 C25 C26 C27 C28 C29 C30\n",
       "1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "2   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "3   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "4   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "5   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "6   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "  C31 C32 C33 C34 C35 C36 C37 C38 C39 C40 C41 C42 C43 C44 C45 C46 C47 C48 C49\n",
       "1   0   0   0   0   0   0   2   0   0   0   0   0   0   0   0   0   7   0   0\n",
       "2   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "3   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "4   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "5   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "6   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "  C50 C51 C52 C53 C54 C55 C56 C57 C58 C59 C60 C61 C62 C63 C64 C65 C66 C67 C68\n",
       "1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "2   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "3   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "4   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "5   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "6   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "  C69 C70 C71 C72 C73 C74 C75 C76 C77 C78 C79 C80 C81 C82 C83 C84 C85 C86 C87\n",
       "1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "2   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "3   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "4   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "5   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "6   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "  C88 C89 C90 C91 C92 C93 C94 C95 C96\n",
       "1   0   0   0   0   0   0   0   0   0\n",
       "2   0   0   0   0   0   0   0   0   0\n",
       "3   0   0   0   0   0   0   0   0   0\n",
       "4   0   0   0   0   0   0   0   0   0\n",
       "5   0   0   0   0   0   0   0   0   0\n",
       "6   0   0   0   0   0   0   0   0   0\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "cd <- read.csv('/home/baker/my-scratch-share/James_Bagnall_Bacterial_Workflow_Riken_Misesq_length/output/featureCount_output/AllFeatureCountsModified.txt')\n",
    "head(cd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      C01 C02 C03 C04 C05 C06 C07 C08 C09 C10 C11 C12 C13 C14 C15 C16 C17 C18\n",
       "Mx1     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Pdgfb   0   0   0   4   0   0   0   0   0   0   0   0   0   0  24   0   0   0\n",
       "Mmp14   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1\n",
       "Ccl3    0   0   0   1   5  36   0  42  22   7  60  64   1   0  13   0   2 159\n",
       "Ckb     0   0   0   0   0   1   0   1   0   1   0   0   0   0  38   0   0   1\n",
       "Relb    0   0   1   0   0   0   0   0  11  91   0   0   7   0   0   0   0   0\n",
       "      C19 C20 C21 C22 C23 C24 C25 C26 C27 C28 C29 C30 C31 C32 C33 C34 C35 C36\n",
       "Mx1     0   0   0   0   0   0   0   0   0   0   4   0   0   0   0   0   1   0\n",
       "Pdgfb   0   0   0   6   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Mmp14   0   0   0   0   0   0   0   0   3   0   0   0   0   0   0   0   0   0\n",
       "Ccl3    3 256   5  61 103   2  21  44   6  61 183  15   0  93   9   1  97   3\n",
       "Ckb     2   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Relb    0   0   0   0   0   0   2   0   0   0   0   0   0   0   0   3   1  10\n",
       "      C37 C38 C39 C40 C41 C42 C43 C44 C45 C46 C47 C48 C49 C50 C51 C52 C53 C54\n",
       "Mx1     0   1   0   0   0   8   0   0  50   0   0   0   0   0   0   0   0   0\n",
       "Pdgfb   0   0   0   0   0   0   0   0   0   0   0   0 195   0   0   0   0   0\n",
       "Mmp14   0   0   0   0  11   1   0   0  24   0   0   0   0   0   0   1   0   0\n",
       "Ccl3    2 289 923 244 100   4   7   0 560   0  67   9   2   0  21   0  15   0\n",
       "Ckb     0 168   0   0   2   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Relb    0  88   0   0   2  54   0   0   0   0   0   0   0   0   0  10   0   0\n",
       "      C55 C56 C57 C58 C59 C60 C61 C62 C63 C64 C65 C66 C67 C68 C69 C70 C71 C72\n",
       "Mx1     0   0   0   0   0   0   0   0   1   3   0   0   1   0   0   0   0   0\n",
       "Pdgfb   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Mmp14   0   0   0   0   0   0   0   0   0   3   0   0   4   0   0   0   0   0\n",
       "Ccl3    4 264  51   4 101   1   0  16   6 124 135   1  78   1   0   0   2   0\n",
       "Ckb     1   0   0   1   0   0   0   0   0   0   1   0   0   0   0   0   0   0\n",
       "Relb    0   0   0   2  14   1   0   0   0  13  48   0 129   0   0   3   0   0\n",
       "      C73 C74 C75 C76 C77 C78 C79 C80 C81 C82 C83 C84 C85 C86 C87 C88 C89 C90\n",
       "Mx1     0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Pdgfb   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Mmp14   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   8   0\n",
       "Ccl3    0   0  20 948   3   0   0  17   8   0 795  10   5   7   5 745 483  18\n",
       "Ckb     0   0   0   0   0   0  46   0   0   0   0  39   0   0   0   0  92   1\n",
       "Relb    1   0   0   0   0   1   0   0   0   0   0   0   0   1  40   0   2   0\n",
       "      C91 C92 C93 C94 C95 C96\n",
       "Mx1     0   0   0   0   0   0\n",
       "Pdgfb   0   0   0   0   0   0\n",
       "Mmp14   0   0   0   0   0   0\n",
       "Ccl3    0 113 194   1   0  65\n",
       "Ckb     0   0   0   0   0   0\n",
       "Relb    0   0   0  78   0  10\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "cdBactAnnot <- cd[cd$GeneName %in% ListeriaAllAnnotated$V2,3:98]\n",
    "rownames(cdBactAnnot) <- cd[cd$GeneName %in% ListeriaAllAnnotated$V2,2]\n",
    "head(cdBactAnnot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlgAAAJYCAIAAAAxBA+LAAAgAElEQVR4nOzdeUAU5/0/8GdmT9hd\nluW+QVgOuQQB8QAEoxVF8UqiRg410abfNmnTX5v2F5vj+0uatOmV1KRJm8MYE6NRwQtRxAsQ5PDg\nvg+5r4W97535/TF1Q1ABURl1P6+/dmeemXlmduTtzDzzPBhJkggAAACwVDjdFQAAAADoBEEIAADA\nokEQAgAAsGgQhAAAACwaBCEAAACLBkEIAADAokEQAgAAsGgQhBbh4sWL2I8xmczAwMCf/vSnQ0ND\n4wqbTKY9e/YsWbLEycnJ2to6JCRk+/btbW1tt6+WJElfX18Mw6Kioh5gbYOCgjAMa2houH3Wpk2b\nbG1tp76qK1euYLextrYODQ19/fXX5XL5g6v1XeXn599eB7OOjg70411eu3ati4vLDFTsbvr7+zEM\ne+utt+449+TJk7cfz/Dw8HfffVetVs9sTQF4MJh0VwDMnAULFixZsoT6rFQqCwsL//Of/+Tk5NTV\n1dnY2FDTpVLp6tWri4qK3NzcFi9e7Ojo2NDQsGfPnm+++eb06dPmxSklJSXt7e0IoWvXrjU0NAQF\nBT2QegoEAqFQiOM4QujGjRtHjx594YUXPDw8pr3CqKio5ORk6jNJksPDw+fPn3/nnXcKCwvPnz9P\nbehhW7RoUWJi4u3ThUIh+vEuj/VAdv9hiI+PT0hIQAgRBNHW1nblypVdu3YdOHAgPz/fyclpiiuZ\n4b17ZA8moB8JLMCFCxcQQr/73e/GTjSZTM8//zxC6M0336SmEATxk5/8BCH0y1/+0mAwmEsWFxfz\neDwHBweJRDJ2Df/zP/+DEEpJSUEIvfHGGw+j5nv27EEIlZSUUF83btwoFAqnvnhJSQlC6Oc///m4\n6TqdbsGCBQihvLy8B1bXuzh79ixCaNeuXVMsv2bNGmdnZ+rzuN2fGX19fWPPinFOnDhx+1y9Xv/S\nSy8hhFavXj31Dc3w3tFyMMFjAW6NWi4cx1999VWEUFlZGTXl+PHjeXl5GzZs+OCDD5jMH+4WLFiw\n4A9/+MPw8HBWVpZ5osFg+P7774VC4UcffYQQ2r9/P/n4dNfHZrM3btyIEKqurqa7Lk8CFov14Ycf\nxsTEnDhx4saNG3RXB4B7A0Fo0RwdHRFCAwMD1Nevv/4aIfTGG2/cXnLr1q3vvfees7OzeUpeXt7w\n8PC6det8fHzmz5/f0tJSUVFxtw0lJSVxOBzzMyTqsZlIJDKZTNSUQ4cOYRj297//HY15ELh06dJt\n27YhhBYsWODj42Nem0wm27Ztm6+vr0AgiI+Pv3z58rSPAIfDMX/Ozc1NSkpydHTkcDg+Pj4/+9nP\nzEfmnuo/DXd89nnH3R8eHv7pT38aHBxsZWUVGBj4hz/8QaVSjV1PdHS0VCpduXIlm82uqamZdBGE\nUF5eXlJSkkgk8vf3/+lPf6pQKKaxCxiGvfLKKwihU6dOmSdOcEjvuHcTlEcIabXad999Nzg42Nra\n2sXFJTU1tbKycmwdJtjTu51LACAIQgtXVFSEEDI/26utrXVwcAgPD7+9pIuLy+9///vVq1ebp3z7\n7bcIIeq6at26dQih/fv3321DK1eu1Ov15sS6ePEiQkgqlZr/kJ0/fx4htGrVqrFLvf/++7/5zW8Q\nQh9++OE333xDTdTpdIsWLZLL5b/73e9+/vOfV1RUrFy5cnh4+J523GAwHDx4kMVimbd44cKFVatW\ndXZ2vvjii7t27QoODv7000+fffbZ+6n/fbp997u7u8PCwr799tukpKQ33ngjKCjoj3/8Y3x8vFar\nHbtrTz/9dEVFxeLFi4VC4aSLfPHFF8uXL6+vr8/IyNiwYcO5c+fMz1PvVUREBELI3K5q4kN6+95N\nXB4htGPHjl27drm6uv7+979/7rnnLl68uGzZMvNPP/Ge3vFcAuC/6L43C2bC7c8IFQrFqVOnqFYD\nZ86cIUnSYDCwWKx58+ZNZYUKhcLa2trOzk6v15Mk2dTUhBBycXExGo13LF9bW4sQevXVV6mvcXFx\noaGhCKG//vWv1JSAgACxWEx9Hvsg8PZnhAihl156ybzm9957DyGUk5Nzx+1SzwgjIyN/d8urr766\nY8cOX19fW1vb7Oxsc8lt27Yxmcz+/n7zlC1btmAYplKp7rX+41DPCBcsWPC723R1dd2+yxM8I9y8\nebNAIGhoaDCv/JNPPkEI/eUvfxl7fNatW6dUKqeyyOjoqJ2dnbu7e19fHzVXKpWGhISge3xGSKEu\nJRMTE6dySG/fu4nLazQaJpO5atUq89zPP//c3t7e/NNPenDgGSG4GwhCi0AF4e0wDHv99depMtR9\nv/j4+KmscN++fQihF154wTyF+uuZn59/x/IEQfj4+MydO5faEJvN/utf/+rt7Z2SkkKSZFdXF0Lo\nV7/6FVV40iC8efOmec1nzpxBCB06dOiO26WC8I6o27nmkhKJZHBw0PzVZDJRl7+jo6P3Wv9xqCC8\no/Ly8tt3+W5BqNfrmUxmZmamZgyVSiUQCJYvXz72+DQ3N1NfJ13k5MmTCKF//OMfYytM/bjTDsKk\npKSpHFLyth934vJ6vZ7NZru5udXV1d2+6akcHAhCcDfw+oQFiY2NHduC393dffHixeYboVZWVl5e\nXq2trXdbvKqqymAwUK8MUjeXFi5cSL0GhxCKj4+vra3dv3//U089dfuyGIalpKT861//Gh4erq6u\n1uv1ixcvrqurO3TokNFopHKaan06KS6X6+npaf46lZcffv7zn1MteigKhSInJyctLW3Dhg3mlh12\ndnY9PT1ZWVlVVVWVlZVFRUVjb7fef/137dr1zjvvTGUH76atrc1oNO7du3fv3r3jZg0ODpo/M5lM\nX1/fKS5CXcrPnTt37KzIyMjp1ZD6D4Gfnx/1deJDeruJy7NYrN27d7/88svBwcFz5syJj49ftmzZ\nsmXLrKysprKnAEwAgtCCJCYm/ulPf5qgQEBAQH5+/s2bN729vcfNkslkc+fODQkJqaysHBgYyM/P\nRwht3759XLEjR458/PHHXC739pWnpKR8/PHHFy5cqKmpsbGxiYiIWLJkyZdffnn16tULFy7w+Xzq\nvbRJcTgcDMOmUvJuBALBpk2bjhw5cvjw4YGBAaoF0O7du3/9619bW1svXbo0Pj7+lVde+f777z/+\n+OMHXv9pMxqNCKHt27fv2LFj3CwqDCg8Hs/8n4NJF2Gz2bdvaNovVlJPTM0xPOkhHWfS8jt37ly9\nevWxY8fy8vIOHDjw0UcfeXh4HD9+PDIycooHB4A7giAEP9iwYUN+fv7rr79ONR8d68iRIyaTibra\nO3jwoMlkSklJeeGFF8aWee211+rr63Nzc6m2M+MkJiZaWVnl5+c3NjbGxcUxmcykpCSE0IULF86f\nP798+fI7/lF+eKi/10NDQ87OzqOjo//n//yfpKSknJwcFotFFThw4MAjVX8/Pz8Mw+Ry+fz5880T\njUZjfn6+QCCY3iJisRghdO3atbEpPr33H0iS/Mc//oEQWrlyJUJoKod0rEnLS6XSjo4OLy+vF198\n8cUXXzSZTNnZ2c8888zrr79+8uTJaRwcAMyg1Sj4wdatW8PCwvbt2/faa68ZDAbz9Orq6ldffdXa\n2vrFF19Et9qLvvbaa2t/jJp7t7ajVlZWS5YsOX369JUrVxYvXowQcnNzmz179ldffXXz5s2J7ysS\nBPEAd5NCXfdQPcyNjo4aDIbAwEDzn+CbN29mZ2eP3fT91P8+UXXgcrkrV67Mzs4e+5rKZ599tmLF\nCuo1idtNusjChQudnJz+8pe/9Pb2UnPlcjnV/uieGAyGX/3qV2VlZampqXPmzEFTO6Rj927S8k1N\nTZGRkeZ3exgMxooVK/h8/r0enIdxLoHHHVwRgh9wudzjx48vX778vffe+/bbb2NjY93c3FpbW3Nz\nczEM++KLLwICAlpaWsrKykJCQqieWcbasmXLb3/72xMnTsjlcnOfbWOlpKTk5OQghKggQQgtWbKE\nuvdFXUbcjnrP7+OPP+7u7h7bkv7+UTVsb29PSkry8fEJCAj497//PTAwEBIS0tHRceTIEV9f3/7+\n/pdeeuntt9+mLh+nUf/7NG73//73v1++fDkuLi49Pd3X17e6uvrgwYM/+clP1q9ff7c1TLyIQCB4\n//33t23bFhkZ+eyzz1pZWWVlZdnZ2fH5/Ikrdv78eepuJEEQ7e3tV65c6ejoCA0N/eyzz6gCUzmk\nY/fu6aefnrh8REREaGjoxx9/LJVKw8LCOjo68vPzlUql+V7opAfn4Z1L4LFHd2sdMBPu2MXa3ajV\n6r/97W+xsbFCodDKyio0NDQtLa2xsZGaS/XF/MEHH9xx2c2bNyOEvvrqqzvOvXnzJkKIx+NRL12Q\nJHnkyBGEUExMzNhiY5tQSiSSuLg4NpsdGRlJ3qmLNapN5sStRm/vYo281dzU1dWV6kyuubl57dq1\n9vb29vb2y5Yty8vLUygUS5YscXFxKSsru6f6jzOVLtbu1mp03O6TJNnd3Z2WliYWi62srIKDg999\n913z2wh3PD6TLkKSpPmFel9f3xdeeEEul9vb20/canQsLpcbEhLy9ttvj1vtpId03N5NWr6zs3P7\n9u1eXl5sNtvZ2XnZsmW5ublT39PbDyYAFIx8fLrFAgAAAB44eEYIAADAokEQAgAAsGgQhAAAACwa\nBCEAAACLBkEIAADAokEQAgAAsGgQhAAAACwaBCEAAACL9iR0sbZ379729na6awEAAE+C+Pj4Ow6m\n9gR7EoJQpVI999xzY8eouydqtdra2vrBVmmKTCYThmHTHvXmPtG44yRJmkwmJpOe00+v12MYZu7c\neYYZjUYGg3GfI0lNG71nO0KIwWDQsnV6z3aj0Ujj+WYymah+VqeirKxMJpM91Co9gp6EIEQIcTic\naY86RhAEXSOW0RuENO44vUFI5ZBlBiG9ZzuiLwjpPdsfoyCk618lveAZIQAAAIsGQQgAAMCiQRAC\nAACwaBCEAAAALBoEIQAAAIsGQQgAAMCiQRACAACwaBCEAAAALBoEIQAAAIsGQQgAAMCiQRACAACw\naBCEAAAALBoEIQAAAIv2hHQ0ThAEQRDTW5YkyWkve59IkkQI0bh1y9xxgiAwDKNr6+YK0LJpi/3R\nLXbHCYK4p32namtpnpAgRPfx+5EkSeNvT++m6T3pYeu0bBd2fOY3jeje8alvHYLwMYbj+LTHObuf\nZe8TveMR0rjj1HiE9B52urZuNBpxHKfripDesx3RNx4hvWc7NQglXVtHCDEYDJIkr127VllZSZJk\nTExMeHj4HcvT9eeIXpa4zwAAYGny8vJOnDhhZWVFkmRWVlZ3dzfdNXqEQBACAMATTiqVlpSULFmy\nZPPmzZmZmVwuVyKRUNfoAD0xt0YBAAAggiCqbxADfbibJx4Shm7dgR8YGEAIhYWFIYQGBwe1Wu2J\nEyeys7M9PDw2bdrE5/PprPMjAIIQAACeCCRp+HYPUVOJEDIhxJg7j7kxjZrj6OiI43hWVpZYLC4p\nKbGzs1u2bJlKpTp9+vTZs2fXrVtHa73pB0EIAABPAqKzg0pBhBCJUPmpk3X9w4Q1LzIyMjo6eu3a\ntYWFheXl5QRBZGZmCoVChNDg4GBrayuttX4kQBACAMATQSE3f8xv7yzt7Zs9N1ZJkseOHXNxcQkP\nDw8PDy8sLGxoaKBSkCRJqVRqa2tLX40fFdBYBgAAngSYpw/1QarTlfb2JXp5PLfzxfT0dA6HY24a\ng+N4b29vVVWVVqu9cOFCU1PT/Pnz6az0owGuCAEA4EmACYWs57Ya9n81qFIjhCJ+9RtkZTV486ZO\npzM3jVmzZk11dXVWVhZCiMlkpqSk+Pv7011x+kEQAgDAEwKfM5cTFuF2s4P19b6j1XVijf7y5csi\nkWjFihVU05iioqIdO3a0t7fr9Xpvb28ej0d3lR8JEIQAAPB402q1ly5damtr43A4YrE4NjZ27bp1\n5qYxaWlpTk5O6FbTGAaDIRaL6a7yowWCEAAAHmNSqXTPnj1qtVosFisUiqKiIh8fH3PTmLq6Omga\nMykIQgAAeFxR/aUZjcadO3c6OjoihFQqlfmGJ9U0prq6OjQ0tLi4uKmpacuWLbTW9xEFQQgAAI8r\nqVTa2dm5bt06KgURQuYUvHTpkre3t7Oz87Fjx06ePAlNYyYAQQgAAI8rmUyGEPLw8Bg3XaVSFRYW\ndnR0PP/889Qr89A0ZgIQhAAA8Lhis9kIocHBQXt7+7HTeTxeVFRUWVkZQRB+fn4cDoemCj4e4IV6\nAAB4XLm6ugoEguLi4tuHknBxcSFJUqPR0FKxxwsEIQAAPK4wDEtISOjq6srNzR07uDxJkrW1tba2\ntgKBgMbqPS7g1igAADzGoqOjOzo6Kioq5HJ5cnKynZ2dVqvNzc1taWnZtGkTdmskJjABCEIAAHiM\nYRi2YcMGBweHgoKCpqYmPp+vUqnYbHZqampQUJDRaKS7go8BCEIAAHi84TielJQUERHR1NQ0Ojoq\nEomCgoKo9+jBVEAQAgDuC0EQSqXSxsaG7opYOpFIFBsbS3ctHkvQWAYAcF8KCws///xziURCd0UA\nmCYIQgDA/ZLL5Xv37oUsBI8pCEIAwH2h+nE2GAzmLIQGGuDxAkEIAJgcKZeZCs6bzueRQwPjZgmF\nQgaDkZ6eTmVhdXX17t27+/r6aKknANMAQQgAmAQ50Kf/4+vGnKPGMyf1f/0j0VQ/dq5QKDSZTHw+\nPyMjQ6/XHzlyhMfjiUQiumoLwL2CVqMAgEmYCi+YP5MIle//ptbLjyTJmJiY8PBwGxsbDMNkMhmG\nYVTnJkqlUqVScblc+qoMwD2AK0IAwCRIlcr8Ob+982R+vpWVFTUSXnd3N4PB4PP5NTU1+/btc3Bw\nyMzMpO6RqtVqGusMwNRBEAIAJoF7z6I+SHW60t6+JcnJmzdvzszM5HK5EonEZDLZ2tqWlpY6ODik\np6fPmjUrIyNj3rx51tbW9FYbgCmCW6MAgEkw4pPIgT7TtfJBlRpzdJ6zfSdCaHBwUKvVnjhxIjs7\nGyHk7Oycnp5O3Q51dXV1dXWludIATBkEIQBgMgwGc2M6c8Mmd4mE+e//ZOeeFovFJSUldnZ2y5Yt\nU6lUp0+fdnFxgYeC4DEFQQgAuCutVnvp0qW2tjYOhyMWi2NjY9euXVtYWFheXk4QRGZmJtWh5eDg\nIDUMOgCPIwhCAMCdSaXSPXv2qNVqsVisUCiKiop8fHzCw8PDw8MLCwsbGhqoFCRJUiqVUq/VA/A4\ngiAEANwB1SjUaDTu3LnT0dERIaRSqXg8HjUXx/He3t6qqqqAgIDi4uKmpqYtW7bQWl8Apg+CEABw\nB1KptLOzc926dVQKIoTMKXjp0iVvb29nZ+esrCyEEJPJTElJ8ff3p62uANwfCEIAwB3IZDKEkIeH\nx7jpKpWqsLCwo6Njx44d7e3ter3e29vbnJEAPI4gCAEAd8BmsxFCg4OD9vb2Y6fzeLyoqKiysjKC\nIMRiMU21A+BBghfqAQB34OrqKhAIiouLTSbTuFkuLi4kSWo0GloqBsADB0EIALgDDMMSEhK6urpy\nc3OpHkQpJEnW1tba2toKBAIaqwfAAwS3RgEAdxYdHd3R0VFRUSGXy5OTk+3s7LRabW5ubktLy6ZN\nmzAMo7uCADwYEIQAgDvDMGzDhg0ODg4FBQVNTU18Pl+lUrHZ7NTU1KCgILprB8ADA0EIALgrHMeT\nkpIiIiKamppGR0dFIlFQUBD1Hj0ATwwIQgDAJEQiUWxsLN21AOBhgcYyAAAALBoEIQAAAIsGQQgA\nAMCiQRACAACwaBCEAAAALBoEIQAAAIsGQQgAAMCizVwQtra2Ll682MbGZtGiRQ0NDbcXKCgoiImJ\nsba2DgwM/Prrr6mJW7Zs4Y7R398/YxUGAABgCWYoCEmSXL16dVxcXEtLy/Lly1NSUgiCGFtALpev\nX78+IyNjYGDgww8//NnPflZUVIQQamho+M9//nPjFgcHh5mpMAAAAAsxQ0FYVlbW09Pz5ptvOjk5\n7dq1S6FQXL58eWyBiooKa2vrX/ziFwKBIDk5OTk5+dy5cyRJtra2JiYmBt3CZEJXOAAAAB6kGQrC\n5ubmkJAQaqhPBoMRHh7e3Nw8tkBMTExxcTHVn71KpaqoqBCLxRKJRCaTvfbaa35+fomJiXl5eeby\nBoNh9BadTjczewEAAODJM0MXWCMjI2NHLxMKhcPDw2MLCAQCqkBlZeW2bdueeuqpZ599tq6uzsfH\nZ/ny5W+99VZeXl5qamp5eXlYWBhC6Pz588nJydSyoaGhK1asUCqV06ubXq8fO9zaTCIIAsMwuoaz\noXHHSZIkSRLHJ/l/GEmSdXV1zc3NTCYzPDzc09PzgRwraqRZBoNx/6uaBov90amnIZP+6A8JvWc7\nQRA0nm8EQRgMhimW1+v1D7U+j6YZCkI7O7uxQSWXy+3t7ceVUalUu3btOnLkyJtvvvn8889jGDZn\nzpz29nZqrlgszsnJOX78OBWEy5cvN5/W//rXv6ysrPh8/vTqplKpeDze9Ja9TyaTCcMwuv400Ljj\nJEmaTKZJb3SfP3++oKDAx8dndHT0+++/Dw8PX716NYvFus+t6/V6DMPufz3TYzQaGQwGXUFI79mO\n6Pv/B71nu9FopPF8M5lMHA5niuWp+3aWZoaCUCwW19fXG41GJpNJEERdXZ2fn9/YAiaTKTU11cnJ\nqa6uznztWFFR0dfXt3r1auqrlZUVXScTmHkqlaqoqGjp0qVxcXEEQRQXF+fn54+MjGRmZsJpAAB4\ngGboWiQ2NtbFxeWf//ynwWD497//zeVyExISEEJZWVn19fUIodOnT9fW1r799tsSiaSjo6Ojo0Mm\nkxEE8fTTTx87dkyhUJw4ceLs2bNr1qyZmQqDmaYf/6BXo9EQBOHp6YkQwnE8Li5u/fr13d3dJ0+e\npOseFwDgiTRDQYhh2PHjx7Ozs52cnL755pucnBzqfuDLL7985swZhFB5efnAwIC/v/+sWz755JN5\n8+Z9+umnr7zyipOT05tvvnno0KHAwMCZqTCYMURbi+Gzj3Sv/1b3u5eJhjrzdJFIxOPxrl69ap4S\nHh7+1FNPVVZW9vb20lFTAMCTCXsC/nP9r3/9KyUlxdvbe3qLwzPCmffDM0KdTvfGb8fOYr/+R4z/\n33vjhYWF586dW7VqVXR0NDWFIIjdu3cHBQUtX7582luHZ4S0bBqeEdKy9Xt9Rnj58mWJRJKamvpQ\na/WogdfyAJ2I3i7zZxKhmqHh5k8+QR6eISEhoaGhixYt6uzszMnJIUkyJiYGIYTjOI/HG9cbAwAA\n3A/oaxTQCbP+4T/p5zo6jzW1EmyWwWDIysr69ttvtVrt008/7e/vn5OTc/jw4fb29kuXLnV3d4eE\nhNBYZwDAEwauCAGdMCcXPCyCqL4h0WhKevpSlixZ8NIvEY7n5+cXFRV9880327Zt27x5c0lJSWlp\naU1NDZfLXbNmjZeXF90VBwA8OSAIAa0wjPXcVtPV0oHLlxlqU/T/fQPhOEIoICCgqKiot7c3Nzc3\nNTV14cKF8+fP12g0XC6XrodMAIAnFdwaBXTDcUbMAtHapzEPr/rmZoQQSZLl5eXe3t6JiYnXr1+X\nyWTo1tNBSEEAwAMHV4SAHlqt9vLly21tbRwORywWx8TEeHp6ZmdnNzc3S6XS7u7utLQ0Ly+v0tLS\n+vr6+fPn011fAMATC64IAQ2kUum///3viooKOzs7giCKioqGhoa2bNkyb968oaEhLpebmZnp5+fH\nZDJZLBZdbxoAACwEXBGCmUaSZFZWltFo3Llzp6OjIxrzjteKFSuoMs3NzQMDA9euXVOpVON64wMA\ngAcLghDMNKlU2tXVlZqaSqUgQsj8pvOlS5cCAwN1Ot3+/fup4SmSk5NhNGYAwEMFQQhmGtX4xd3d\nfdx0lUpVWFjY0dGRmZn50ksvSSQSZ2dnGxsbOuoIALAg8IwQzDRqnJehoaFx03k8XlRUVEdHh8Fg\nsLOz8/f3hxQEAMwACEIw01xdXfl8fklJCdX55FguLi4kSWo0GloqBgCwTBCEYKZhGJaQkNDV1ZWb\nmzu2z3eSJGtra21tbc0DUgIAwAyAZ4SABtHR0W1tbRUVFXK5PDk52c7OTqvV5ubmtrS0bNq0Cd6X\nAADMJAhCQAMMw9avX3/58uWCgoKmpiY+n69SqdhsdmpqalBQEN21AwBYFghCQA8cx5OSkiIiIpqa\nmkZHR0UiUVBQkFAopLteAACLA0EI6CQSiWJjY+muBQDAokFjGQAAABYNghAAAIBFgyAEAABg0SAI\nAQAAWDQIQgAAABYNghAAAIBFgyAEAABg0SAIAQAAWDQIQgAAIghCLpfTXQsA6AFBCABAhYWFn3/+\nuUQiobsiANAAghAAgBBCcrl87969kIXAAkEQAgCQra0tQshgMJiz0Gg00l0pAGYIBCEAAAmFQgaD\nkZ6eTmVhdXX17t27+/r66K4XADMBghAAi0B2dxq++IT5//6v7ncvE5XXxs0VCoUmk4nP52dkZOj1\n+iNHjvB4PJFIREtVAZhhEIQAWACC0O/+K9FUT30z7P+KlAyNnW9jY4NhmEwmM5lMJEkihJRKpUql\noqGqAMw4CEIAnnzk0ID5M0GSJT19H//5/U8++aS4uJiKPQaDwefza2pq9u3b5+DgkJmZSd0jVavV\n9NUagBkCA/MCYAGseeaP5X0D5zq6wqPm6wQCoVCIYRg13dbWtrS01N3dPT09ncvlZmRktLa2Wltb\n01RjAGYOBCEATz5MYMOInm+quIIQKuzqiZ8f+5NfvIxYrJGRkevXr1tbW4vF4uXLl1+4cOGZZ57h\ncrkIIVdXV1dXV7orDsBMgCAEwCIwn96MiQN0zY36IblH+nbEYpWVlZ0+fZogCISQk5PTtm3b0tPT\n6a4mADSAIATAMmAYIzKaDJjtIVNX1dc7uXPKydUAACAASURBVLufPn06Li4uPj6+tbX10KFDFy5c\nWLlyJd21BIAGEIQAPMm0Wu2lS5fa2to4HI5YLA4NDY2JiTl69KhSqXR3d09KSsIwLCgoKDw8vKur\ni+7KAkAPaDUKwBNLKpV+8sknFRUVdnZ2BEEUFRUNDg7OmTMnJCSku7t7bEmlUikQCOiqJwD0gitC\nAJ5MJElmZWUZjcadO3c6OjoihKj3AjEMW7NmjVqtbm9vz83NDQsLq62tbW5uTktLo7vKANADghCA\nJ5NUKu3s7Fy3bh2VggghHo9HZWFJScmyZcsaGhrKy8vLyso4HM6aNWvEYjGt9QWANhCEADyZZDIZ\nQsjDw2PcdJVKVVhY2NHRkZmZmZCQoFAo+Hw+i8Wio44APBLgGSEATyY2m40QGhwcHDedx+NFRUV1\ndHQYDAYmkykSiSAFgYWDIATgyeTq6ioQCIqLi00m07hZLi4uJElqNBpaKgbAowaCEIAnE4ZhCQkJ\nXV1dubm5VIeiFJIka2trbW1toZkoABR4RgjAEys6Orqjo6OiokIulycnJ9vZ2Wm12ry8vJaWlk2b\nNpl7GQXAwkEQAvDEwjBsw4YNDg4OBQUFTU1NfD5/dHSUz+enpqYGBQXRXTsAHhUQhAA8yXAcT0pK\nioiIaGpqGh0d5XK5kZGRQqGQ7noB8AiBIATgyScSiWJjYxFCKpWKx+NNWh4AiwKNZQAAAFi0J+SK\nkCCI29uIz8Cy94m8hZat07jjFLq2ThAEhmE07js18hFdm6bxbEe0/uiWueMmk+me9p3Gk5NGT0gQ\nIoSm3QQOwzC6ms+RJEnj1undcXQfP9kDYZn7brE/Oo07jm79S6dl09gtUy//UOvzaHpCghDHcRyf\n5m1eDMOmvex9ov550LV1enfcZDLRtXUcx2ncd4IgqArQsnV6f3SEkGWe7dSPTsvWcRwnSXLqW7fM\nIIRnhAAAACwaBCEAAACLBkEIfoQgCLlcTnctAABg5kAQgh8pLCz8/PPPJRIJ3RUBAIAZAkEIxpPL\n5Xv37oUsBABYCAhC8CO2trYIIYPBYM5Co9FId6UAAOAhgiC0RGRfD9HajFTK22cJhUIGg5Genk5l\nYXV19e7du/v6+ma+kgAAMDOekPcIwVSRpPH7b03XyhBCTISI9Bfw0PCx84VCoclk4vP5GRkZe/fu\nPXLkiJubm0gkoqm6AADw0EEQWhairppKQYQQidC1v/+pZdlqEqGQkJDQ0FAcx21sbBBC7e3t9vb2\n1BvQSqVSpVJxuVw66w0AAA8N3Bp9MpEkWVpa+tFHH/35z38+fPiwUvnfu6Bkf6+5zLmOzmNNrSa1\nymAwZGVlffvtt2q1GsOw9vb2Dz744PPPP3dwcMjMzKTukarVapp2BQAAHi64InzoCII4fvx4UFDQ\ngxoKlSTJa9euVVZWkiQZExMTHh5+e5n8/Pzi4uKIiAgWi3X16lW1Wp2RkYEQwhydqQISjeZKT1+y\nr8+i7c8jDMvPzy8qKtq3b59IJJJKpTKZTCwWp6enc7ncjIyM1tZWa2vr+6lwTU1NY2Mjk8mMjIz0\n8vKa9qoAAOCBgyB8uAiCyM7Orq6utrGxeVBBmJeXV1JSEhgYqFKpsrKy7OzsPDw8xhaQyWSXL19O\nTk6eP38+QojD4RQWFmq1Wi6Xi4dFMCJjTNfLe5UqhFD0K79FGIYQCggIKCwsPH/+PI7jmzZtys/P\nT0hIoG6Hurq6urq63k+FL1y4UFBQ4OPjo1Aobty4ER4evmrVKrq6XgQAgHEgCB8icwrGxcUlJSU9\nkHVKpdKSkpIlS5YkJCQYDIa//e1vEonE1dWVwWCYywwNDSGEfH19qa92dnboVpfHCMOYm9IZ8Yl2\ndXVEXn4jgx2OEEmSZWVlEonEysqKJMkVK1ZYW1sPDg4+kAqrVKqioqKlS5fGxcURBFFcXJyfny+R\nSNLS0phMOP0AAPSDv0QPC0mS5hR86qmnuru7S0pKFAqFl5dXQkICh8OZfA0D/URzA8bn46ER6FZm\nDAwMIITCwsIQQv3lpeorRdmnjh1xcvFMSNq8fTufz0cIOTs7Yxh25cqV1NRUahE7OzsMw86cOdPa\n2spgMGb7+UaHhrk3NmdnZzc3N0ul0uLiYj6fv2nTphs3bjQ0NHh4eJw7d+6BjB2j0WgIgvD09EQI\n4TgeFxdnY2OTlZWVk5OzYcMGy+zqHgDwSIEgfEC0GqK3B+PxMOf/3kXEMIy6SgsKCqqvrz906JC9\nvb2NjU1xcXFjY+POnTvZbPYE6zNdLTN+/w31GRdfYaU/j7hWCCFHR0ccx7OysvycnYr+8idbFvOp\nAD+VwXj2Qv5ZT89169YhhAQCwaJFi4xGI5VkTU1NHh4en3zyiUqlGhkZGayq7NOr3UKCtlhbXw6d\n2zk0JBAIfv/73ysUioiIiLq6OgzD/P39T5w40dHRMWvWrKkeAZIkKq8RjfWIx2MsTMDs7KnJIpGI\nx+NdvXrV29ubmhIeHi6VSvPz8xcsWODu7n6vRxoAAB4sCMIHgOzt0X/4J4QwhBAeHsnamE5dwK1c\nubK7uzsrK8tkMoWGhq5btw7H8ZaWlv379xcXFycmJk6wTnMKkghVFBXVtN5kBMxeuHBhUFDQ2rVr\nCwsLy86eIRCZFjpbyGETJFnSU9XZ2GhefMmSJdRonEqlcmRkRKVSMRgMd3d3w/DQCkfbKNcAHouF\nEHqqoZLzpw8RhjU3N7u7u585c0alUvn5+dnY2Li7u5eWlk49CE3nThvP5v73c+EF9v/9X8xWhBBi\nMBjz588/d+6cl5dXdHQ0VWDRokVXr16tqamBIAQA0A6C8AEw5uVQKUgiVJF3pra6DvMPotpzPvPM\nM5999pnRaExISKCah4jF4uDg4K6urgnX+EOvZmfbb5b19vvaOanU6oMHD7700kvh4eHh4eGXvvys\nrrFeyGEjhNRGY3F3r/vNm+bxP81NUeRyuVarVavVzs7ON2/eXDzLexGHpGYRJPmfG9Xqt94k2Jz2\n9nYPDw8cx5OTkx0cHBBCMTExx44dGx4epr5OgiTNKUgiVNHXf+31P2g8vPz8/JKTkxctWtTZ2ZmT\nk0M1c6Wqx+PxCIK4xyMNAAAPHrTcewCI+hrqQ35756nWdq5aTZJkVlZWd3e3s7NzcnKyq6sr1Ycn\nQogkSaVSST3MuysmEw8KRgj1K1Vlvf0r/GZlbtu2Y8cODofT19en0WgQQkwf316lqnpoWGs0VfQO\nGPgCxGJlZ2ePSxeBQKDT6aqrqwcHBxMTE8dehhZ29RxtbO0ZGvb29t6+ffuWLVt+9atfzZs3j5ob\nEhJibW196tSp/7aymeQQ/LDR8x1dZ9o6PWxtwsLC6uvrs7KycBx/+umn/f39c3JyDh8+3N7efunS\npe7u7pCQkMnXDAAADxlcET4AeGAw0Vgn1elKe/sSvTwS164h4pPM7TmjoqLCwsIwDGtoaHB1dS0u\nLu7o6Ni6devE62QmrzZqtb1XriCEzU1di4dF9LS363S6Q4cOIYT8/f1TU1OrNjx7vPQKktw08AVu\nc6P9/f2rq6sRQtQ9WGo9AoEgMDDw+vXrwcHBiYmJyGjEg0KIhlqEUHnfwLLkZIOz88aNGxUKxc2b\nNwmC4PF41KNNFou1ePHiU6dO9fT0jHs94w4YDDw0nKipkun0JT29P5nlvSg9Ew8IGvvmxubNm0tK\nSkpLS2tqajgcTmpqKrxQCAB4FEAQPgCMp5YTjXWDKjVCKGzhIkZcYv/goFarPXHiRHZ2toeHx6ZN\nmy5evFhSUoIQwnF8xYoVPj4+E68Tc3Vn7XzJPWEpY/9319285xHE0aNHnZ2d169f39raeubMmatX\nr+58+Zft7al6vb6jo0Mmkz3zzDNFRUUXL15EP87CtLS0uro6o9FoMpkYTCYrcwdRV02MSNQacu7i\nxOrq6qKiosLCQupS0snJadu2bVZWVgihqKgoHMcnT0GEEELMtc8aGczhixcQQuItGXhAEPrxmxsY\nhi1cuHD+/PkajWYqjWYBAGBmMN566y2663C/ysvLAwICzPce75XBYJi4AeekMFsRMy6R4e55VSrv\nd/VUqNU5OTk2NjarVq2aNWvWjRs3FApFSkqKn5+fWCxevny5n58ftSDVqnPsKwRarfbcuXNnz56t\nrq5WqFQBYWFGkgwPD+fxeC4uLgsXLhSJRI6OjmfPni0sLJTJZCwWKywsLDAwMCQkhMlkUvlaVlY2\nMjISFBRErZnNZjs4OFRUVKjVan9/fwzHMScX3HtWXUtrRUWFUChsb2+Pj49PS0tzd3evqKjQ6XT+\n/v4IIRzH3dzcpnoQOBxGWAQ3cWnp4DDh5BIYGIgQqqysVKvVcXFxCCGCIHQ6HYvFYrPZGIaRJEnX\nO/Umk8ncpnfmUc9x6Xpv5P7P9mmj/j9E149O444jhAiCoPF8I0ly6u/sdnV1aTQa6t+v5YArwgfE\nysohdsE6K15hYWF5eTlBEJmZmUKhECE0ODjY2tqKYdikdwKlUumePXvUarVYLFYoFEVFRT4+PsnJ\nydRcqgGnVCr98ssvW1pavLy8CIKgyoxdM/UUcNx1YXR0dHNz88GDB7OysiIjI+fMmTNnzhyVSlVT\nU+Pi4uLh4ZGUlIRhWFBQUHh4+CQNeX5Mq9VeunSpra2Nw+GIxeLY2Nhxb25Q/elQfQsMDw9v376d\nxWLdw4EFAICHDIJw+m7PAKo9Z2FhYUNDA5WCJElKpdKpXK1S7WuMRuPOnTsdHR0RQiqVisfjmQsU\nFRXZ29sXFhZWVlb6+fm9+OKLvr6+48pQzFk4e/bs4OBghJBMJuvv72ez2XK5/OzZs6dOnZo9e7ZQ\nKFy5cuW43rSVSqVAIJjiEbhjco97c2PWrFnmHnYSExMhBQEAjxoIwmm6YwZQV2Y4jvf29lZVVQUE\nBBQXFzc1NW3ZsmUqK+zs7Fy3bh2Vggghc8JdunTJy8urtra2ra0tJyeHz+dv3bqV6kFtbJnAwEAX\nFxfqa2Jioo+PD3WnlIpYk8n0pz/9iclkNjU19fX1ubq6enl52dvbf/fdd+3t7bm5uWFhYbW1tc3N\nzWlpaVM5ApMmt1wuRwjZ2tqaU3DiVycBAIAWEITTMXEGzJ07t7q6OisrCyHEZDJTUlKoR24Tk8lk\nCKHbW6aoVKrCwkJPT8/t27e/9957XC43NjaW6rFsXJmOjo7MzEzzRHN7nHERGxsba16qpKRk2bJl\nDQ0N5eXlZWVlHA5nzZo1YrF4Kgdh4uQODAwUCAQYhn333XcjIyOQggCARxYE4XRMmgE7duxob2/X\n6/Xe3t6337q8I+pJ/uDgoL29/djpPB4vKiqqtLQ0OztbJpMFBwc7ODhIJJKxcUWVKSsrMxgMt997\nnDhiqfhMSEhQKBR8Pn/qty6nslqqezlIQQDAowxeqJ+OiTPgzJkzDAaD6kFmiimIEHJ1dRUIBMXF\nxSaTadwsJyen+vr669evJycnh4eHDw4OdnZ2jivj4uJCkiT1rv045ogdN52Kz46ODoPBwGQyRSLR\nPT3Am8pqV61alZycDCkIAHiUQRBOx1Qy4F7XiWFYQkJCV1dX7qlTxEA/KZdR00mSPHbsmEwmW7Jk\nydKlSxcvXqzX60+dOjW2BxmSJGtra21tbe/YzmWCiJ0gPic1ldXyeLx58+ZVVFRAb2oAgEcWBOF0\nPKRoiY6ODvb0KPvsk30vbO1//bfGQ/u1avXRo0dJkty5c+fSpUsxDIuOjl60aFF7e/vu3btHRkYQ\nQlqt9ujRoy0tLcnJyXd8O+2HiM3NHdtf2sTxOakprlaj0Zw+ffr2vt8AAOARAc8Ip4PKgJycnNzc\n3JSUFHP83H+0rGGStjxuYVdvy9VRXlWd5ux5jo+vv79/b2/v+++/T/VhnZaWVlVVVVRUNDo6yufz\nVSoVm81OTU2l3ti7o+jo6I6OjoqKCrlcnpycbGdnR3V809LSsmnTpmm/3H3H1ebm5o5dLY/Hc3Nz\nu73vNwAAeERAEE7TVDJgOuqqF3t5hDs5No+OlvX2Dw30t2v1lZWV69ev9/X1vXr1qlqtzsjISEtL\nO3DgwIIFC0iSFIlEQUFB1GuLd4Nh2IYNGxwcHAoKCpqamqj4RAhNHJ+T+tFqqyqtVUqVTs/19Bq7\nWolE0tPTY29vD1kIAHg0QRBO0x2jxXxlRhDEtWvXIiIiTp48GRQUdK9hI+JyFDqDVKufOy/01M0u\noVCo0+k2btxo7sM6JCTE3t5+YGAgPT19iqGL43hSUlJERERTU9Po6KhIJPLy8nJ1dZ3W3o9fbbiN\noP7D90e1OhGXEyDpsffxlsvlNjY2CKHKykqxWDyuH1QYmB4A8OiAIJy+26PFfGWm0Whyc3MPHjxo\nbW1tY2Mz9SBkpT9v2PeFeQwHl81bmo8cCQsLq66u1mq15j6s721oiDFEItHY9wjvcY/vyqbqaoyr\ni/nrpU//dYPLy8zMtLe3T0xMJEmSwWCM7ftt7dq1D2rTAABwnyAI79fYaDGzsrLq6+srKChYvHjx\n4sWLJ1+LVmMsvEgO9mMiO9bPfz1aUoJrTIFvvGVlY4NhGNX5J0mSAwMDdnZ21NAQkZGRRqPxnlLw\n4aHGdTIjVQq53rR3714qC83TzVlIkmRqauoMVxIAAO4IgvB+EQQxMjIydhh3kiT37t1bWFhobW2t\nVCo///xzW1tbpVLp5eWVkJBwhxGISNJwYJ95dF9TT5fgJ6vxhqYrV6+mpqYuWrToypUrIpHIysqq\nqakpMDBQq9Wy2ezjx48PDw/HxMRM/eU/kiRramoaGxuZTGZkZOS4N/fvBx4ZTVyvMH+18xOjzl6D\nwWDOQqPRSPV/b87CwMDAsLCwB1UBAACYNgjC6SPlMv2JrOzDh5tHRl/6f+/YLl7y3+kkmZeXZzKZ\nfvnLX1Jh4OfnFx8fX1xc3NjYuHPnznHDwZCD/erqysKunnapjMNkqK5W6ps7wuYvZLPZJEkuWbKk\ntrZ29uzZSqVSIpE0Nze3t7fb2dnV1dXdax/WFy5cKCgo8PHxUSgU169fVyqV27ZtCw0Nvf9DwUxe\nbZTLiNZmhBAjMkYUs5Cxf396evq+ffv27t27bNmy/Pz8TZs2UY8kExMTvb29x/USBwAAdIEgnD79\nscPZWVm1Q8MIod5v9/Ld3NTObgKBIDs728/Pz87Obnh4eGRkxNHRkRqJ6dlnn83KyiouLh7X04p0\nePjzG9Vqg8FPZKvU66uHJTb9A8z29szMTGoMh7q6OrVaXVVVdeXKlTlz5vD5/P7+/nvtt0ylUhUV\nFS1dujQuLs5oNP75z38uLy/HMOy99967/xEhMFsRa8cvSLkM43ARlyscHTWZTHw+PyMjY+/evUeO\nHHFzcxOJRObyPj4+t7+CCQAAtIAgnCZCrzty5MiBukZ3Pr9fpdpdcX3Fn99tmT0nc+tWBoPBZDJ3\n7Njx5ZdfFhcXU3/0T58+3dbWlpKSMm60P5Iks0uuGAnihTmhDtZWCKFgB/vzHp5Go/Hrr7/OzMw8\nc+ZMQ0PD3Llzy8vLm5qa3N3dra2tJ09BgwExmWhM40yNRkMQhKenJ0EQx44dMxgMzz777PDw8MmT\nJ9euXTu9Zpy3D0RF3fi1sbHBMEwmk1ED8CKElEqlSqXicrnT2AoAADxUMEL9dIauJggi++ixrMOH\n20ZlQi5HotUWdvUUt3YMSaXHc05RN/26u7urqqpkMtlzzz3n7e1dW1tbX19vNBrDw8Nnz55NrYca\nrfBsfv7qrVtn2QjIwX58dii2MrWms2vr1q1Hjhy5fv16bW1tb2+vh4eHtbW1SqVycnJauXJlUlLS\n3epGSkeNB78xHtxnyj+NMZn4LD9qOofDuXr1qkajaWhoqK6ujouLW7VqlUKh+O677zo7O1UqlYeH\nx9SHsUYISaXSzz777ObNm+7u7jqdrr6+3s/Pj2o0i+P41atXqVvEDg4O69evr66urq6unjNnjvnq\nE0aop2XrMEI9LWCE+kccXBFOR3FxcUVFBYvDCXIQdSsUXjY2raOyLrlC33EzLjExKysrMTFxcHBQ\nIpG8/fbbTz311EcffcTlcvl8fk9PT2Rk5NhVUf13e4bNYc4OJtw8EJMpdPcwmUw4jtva2t64cUOj\n0QQFBWm12oiICJPJNDw8XFBQ0NbWtmnTJj6ff3vdjMcOE3XVJELX+wer3v8zs6hk0TPPBgUFMRiM\n+fPnnzt3jloqKCiosbGxurqaaox6t+eXdzPpYIS2tralpaXu7u7p6elcLjcjI6O1tdXa2vp+DjsA\nADwMcEU4yX8VSZnMdPyIqbiAqLqGO7uSPL6kp9vXw1MzNNR3taxxeDTK1QkhVN43wGIwtCS5bNky\nhBD12ntNTc369ev37dt34sQJpVI5NDTE4XDc3NyCg4Op/xdrNJr8/Py8vLz+mmpj1gHH/m68pYlR\nerlITwyMjGAYVldXh2EYi8USi8XDw8MikWjLli2zZs26ceOGQqEwX1n+gCCMB75GCJ1tv3nhZrdC\nb8BlIxWXLoWHh1k7OHp6evb29vb29qrV6t7e3qamptmzZ7u7u8+ZMycxMbGsrAzHcfMohhOTSqV5\neXmrVq2ixgdGtzoiRwhdunSJy+X6+vrKZLLNmzdTt0MFAgE1avEPBxauCOkAV4S0gCvCRxxcEU6I\nIAx7/032dFPfNPW1/+joKbl+w0PAx2yEpd19JhPxcnTk7y8UGQhCwGZjOH7mzBkPD4/+/n4+n29r\na/vuu+/OmjWLyWR6eHh4enoGBQU1NzcXFRUlJiZKpdIvvvhCrVaLRKKm8jITA3kLbTxtBBhC7eXl\nrX39KpVKKBRmZmbu27evoKAgKirqZz/7GXXvkWp9c4cK4zhCqF+pKuvtX+LjWdDV40ASysH+jvf/\naHztf53F4qeffpq641pSUmJnZxcUFNTT07N8+XIvL6/g4OBxzy8nMJXBCNPT06d73AEAYOZAEE6E\nHBwwpyBBkv+v6Mp3dY2IRLVDEhaOa4wmhV6Xeuh4gJ3Ilc9TGQwMtbqrq6u2tpbD4URHR1dXV7e1\ntcnlcpIkfXx8li5dWlVVNTIycuTIER8fn3PnzlG3Fjs7O4+/8osAOwcPGwFBksea20blMhHubWdn\nN3v27PXr1y9cuDArK4vNZlMpSD1ZvNsVMDN1Q+9/PkEIm+fm0jIqLe3pG1Cpm0ekA2lb1m/b/vzz\nz2/evNnHx+fw4cMNDQ01NTVr1qzx8vIiSVKpVE7cYelYEw8jfLchggEA4BEEQTgRUiGlPhAk+W1N\nw5eVtUq9HiGMx2ZxGEyJVsNlsnqVyl6l0t7amoFhMp0OkwxrDEaSJC9duhQVFXX9+vXh4WF7e/vY\n2FiRSHTt2jVnZ2cej/fpp5+2t7f/5je/cXR0dHJyap2/oPZKic5kMhKmgs6eJUlLbuJM6nbKl19+\n6eXlNXv27IKCggMHDshkspaWFqVS+etf/xoZjaYrRURXJ+7oxFi0GFlZIYQYixa7GUnsXx9f7Ojq\nliu7FQq9ifAW2jAdHLKyslxdXVNTUxcuXBgbG9vX1+fo6Nje3i6TyYqLizs6OrZu3TrFI2MeiCog\nIGDcPR/zQFQQhACAxwIE4URwD2+EkJEgvqlp+PRalUynI0hEIhI3GuU6vYDNinRxUup0XQrFsEqN\nYRhCmMFoYjEYfD5/cHDw+PHjOp0uLCyMz+fn5+fLZDK1Wt3X1/fpp5/W1dVduXKlqKgoPDycw+E8\n8+bb9u+8VXD5ct2QpFKuGGxsnh0crNfrWSwWi8UqLi62trYeHBy8ePGig4ODyWRydHS04nIN331F\n1FSRCJX29Vd8/oU8MDg4LCw5OdkrYfFCjVaZc1ys020MDmgblZ1oaVMyWSK+YP/+/YsWLbK3t6eG\nrbh48WJJSQlCCMfxFStWTPEBIXpoA1EBAMDMg8Yy45+ik2qV6WK+qegiUVFK9nabRiV/Lij5Z8UN\nhU6nMhhJkkQYMhGEkSRNBDmkVuM4Hu/p3jgyqjUYbTlsIZfDwjCMwzUYDNRN0RdeeKGzs7OysrK/\nv9/R0VGj0TCZzOXLlzc2No6MjJhMpqCgIJzD8Vu2PHLVar8FCxs7bkql0jlz5sTGxmZkZERERHh4\neJSUlEgkko0bN0ZFRfn6+jY1NTnxeB7Xy3AMO9/RdbGzRyuXl3Z0Xq2uLikpEYlEcXFxIYlLgm34\ndpIh/7gETUh40Y1KBwcHkiQHBwf5fP6BAwfc3d2jo6P9/PzEYvHy5cv9/Pzu6bi5ubkNDw9XVlb2\n9fW5u7tbWVlptdqTJ0/W19evXbuWako6AWgsQ8vWobEMLaCxzCMOrgh/TKs1frOHaG1CCBEkeay5\ntWlkNL/9ptZoRAghRGI4ThAEQZIIw3Qmk95kUhuNPXKlzmQiEZJodV5CAQPDbVjMDonE2sra2sam\nsbHR3t7e29ubz+c/88wzjY2NZWVlsbGxYrF4YGCgoqJCq9XKZLL/vpM+f/4vf/nLd999V6VSJSQk\nUH81xGKxi4tLQ0MDj8e7fv26UCgcHh7+y4cffCkZinV3GVJpFnt7nGhuG9Yq1CMjvb291tbWPj4+\nXl5ejPUbi5zcHRwcnrKz23vyVEtLy8qVK3U6HdXVi62tLYZh4xpzmpEkWV1dXVdXR5JkSEhIaGjo\nuD9hEw9EdU9HnSCIkpKSyspKDMPmzJmzYMECGKcJADBjIAh/hGhuoFJQqTf8+tylky3tMq2OQIiF\nYQwcJ0iEEEkihOO4iSCoP9YGE2EgTUwcxzDMSBA9cmWUq3OXXE4YjdYmg6+jw9WrV6meVjIyMhoa\nGmbPnt3W1lZVVZWQkHDo0KELFy709vauWLFCoVAUFRV5e7j3XjwfbsX2lEqExH87IaOGMWIwGHv2\n7Fm6dKlSqVy+fLmDSHT+7+/ntnbYcjhOPKtehZJt5yiysens7Ozs7Pz+++9tbW29vLyam5uHhoZI\nktTr9StWrKA+C4VCpVKpVqsn6Hf7tgMDHgAAIABJREFU7Nmz1CNAo9GYlZVVWVm5YcOGcS8CTjAQ\n1T0pLS3Nz88PCQnRarVCoRBSEAAwkyAIxyAIY+EFhNCIVvuLMxePN7cYCZKaYyBJI0IkQoggEEII\nw3AMY+K4gSBIksQwjMlgWDFwhcGgNZmu9w8ycGyuixOTwbhaU0uyWDY2NosWLYqJiTGZTE1NTUKh\n0MvLKyoq6osvvpDL5QKBYNmyZXZ2dhKJ5MKf3mkuKd4WFDCLxzL89Z3Wzdvc/MTFxcVDnTfXuTjs\nu1TYcxHxQ+ds/dnPhEJhqIf7O7//XePNm0Yra/7sYFdX96ioqA8++KChoUEsFl+5coXFYr3yyiv2\n9vZyuZwaceLChQv+/v5Lly79/vvvv/3225dffvmOL7kPDw8XFxenpKTExMQghPLz84uKir755ptt\n27bd3gTmjgNR3ZOCgoL4+PglS5YghEZGRq5fv25tbS0Wi+m6mwQAsCgQhAghhDRqorvLeOoo2dtD\nIPSbc4WnWtpMt1IQIYRu9ZlJIoRhGEEQCMMMVCgihBAyEYTCZDISBMKQxmhk4TibwRBxuVZstquv\n78aNG62trY8dO5aWllZVVdXc3Nzf319VVaXRaPz8/Fgs1j//+U8+n6+SyfCy0hSx75BaLeSws5ta\nJU1vmoLDrFjMpX0dc12cb9rb6gcH7NqabRg4SZLI3oHw9FZJRvvZ3FjvWba2tnl5eWKxWCqVVlRU\nyOVygiCqqqpeeOEFhNC1a9dKS0v9/f3NXb3U1tZaW1sjnY6UDCMrK0xkZ96d3t5ehFBERAT1NSAg\noKioqLe3Nzc394GPI0g1MXVxcUEIlZWVnT59miAIhJCTk9O2bduowRcBAODhmbkgbG1t3b59+/Xr\n18PCwr744ovbHyPdscCkS90/vKRQd/aU+atUqyvs6tYTBELI3Gc0SZIYdUV465k/Ro0Uz2AQBEEg\nZCQI6nYejuEMDMMx7Hr/oM5EcHg8uVy+Z88eNpstl8uvXLkSFBSUlpamUCi++uqr7u7uXbt2RUZG\nUrcWbbkcX8UQE8f+WXGjaWRUptMpZXKWj1/a6pXuOhlCKNbN5evq+vKSK+djorqMhBRn2Ds4zJo1\nq6enp66uzs7Orqury8fHh7q1qNForKysLl++nJmZWVxcPGfOHDab/cwzz1Bdvbi6utrY2BAtTYbP\nPqJ2nDE/jrn2GaqfbhsbG4RQfX19eHg4SZLl5eXe3t6zZs26dOnS4sWLp3HzEyFENNWbrlxGOI6J\n7BhLkwkMHx4ednBwwDDM3d29qqrKycnp9OnTcXFx8fHxra2t1H3jlStXTveHBQCAKZmhICRJcvXq\n1evWrTt06NCnn36akpLS3Nw8tvHFHQtgGDbxUvdLOqL76zu4gWoIg/Qm05n2TgGLOarRmS8GzVlo\n/krVlppoMJmo9n8MDDORJAPDOFyun5tbT3+fRm/QGo16tVprMGAYZmVlxePx+vr6YmJiMjIymExm\nW1vb3r17nZ2dx95aNNwoM/X1zHV2Ku8bcBPw2kZl2taWrLyzW7RaOyvuAnfXww3NZa39fDbLSJAM\nLtfK03PZsmUFBQUjIyN6vV6lUvH5fIFAwOPxnJ2du7q6hoeHJRJJYWGhp6dnZmbmuANgTkESofKj\nWbU19bi758KFCwMCAjw9PbOzs5ubm6VSaXd3d1pampeXV2lpaX19/fz58+/1SJO93YYvPjF/JdSq\noySrtbX1F7/4hUAgiImJOXr0qFKpdHd3T0pKwjAsKCgoPDx86j3dAADAtM1QU+aysrKenp4333zT\nyclp165dCoXi8uXLkxaYdKn7olLp/vy/6FYKqg3GHafy/+f0uWePnlIZDNT139gIHPNpTC4iRJAk\njmEMHBfxeEw2m0RoXmKihiC1JhPOYHA4HBsbG3d39/j4+KioKKr3teLiYhzH/fz8bGxsSkpKxo7M\nx0jdgCHkzLPWGAw3ZQpnnvWgXFHZ3ft1Tf2IRivXGzqk8nAnh1lCoYjLifLyHB0dLSwspMY/ksvl\ng4ODfX19zs7Oa9eufeONN1gslr29vbOzc1RUVEdHh8Fg+NERUKvMH8+23zzV2s5QyNVq9cGDB6VS\n6ZYtW+bNmzc0NMTlcjMzM/38/JhMJovFml5LFlP1DfNngiSz9u+vqa7W6XQDAwMIoTlz5oSEhHR3\nd49dRKlUwsuIAIAZMJ0rQoPB0NPT4+LiwuFwpvhnsbm5OSQkhHqPh8FghIeHNzc3x8fHT1yAzWbf\nbSmJRHL9+nVq2XF/QKe6F1kH0K2ngMMazbPZOdf6h3Qmkzn8zLuGYT9kH9U0Brt1s5QqRJAkzmC4\nuLl29PTqdLrDhw8TBMHn89lstkajEQgEWq22s7NzaGgIwzBqwN7o6Gg+n5+QkHDq1Kmx76TjvmJG\n8qr6v/1NaTAscHZcsH2HZP9BoVJepTftGxxxQ0TjyOjmkMBGyUirVHa5ucXE5nh7e1dXV3M4HE9P\nz+7u7v7+frFYHBUV9Y9//OP/s3ffAVGc+f/An9lCrwsiSJdFqogB7CJYMSoK1sSKJiYm0Uu+aorJ\nL2o0MfXSPDUaUzU5YwJGQ7CeCoiKWBDpbRUEQeksLOzOzO+PiXscaqjLgPN+/XG3DDM+n4EN752Z\np9TU1Lz00ksURT18thfDv7rJcHOTTnVzHT5/vsZ3yKefflpaWjpw4MCpU6dqfztlZWVXrlxRKpUd\nHW74F+qvj1zcHHLpd++NGT26qLi4oqJCLpdTFDVz5syGhobCwsK4uLjBgwenp6fn5uYuWrSoM20B\nAHREx4Lwhx9+eOeddwoLCxmGyczM3LZtm4eHxxtvvNFmHFZWVrb8dM+NhGtzBwMDg0cdlZOT8/rr\nr3OvTU1NlyxZUl9f36FzkVRWcEXTLDsvJjaltFxN09prvb9yjnDdZAhFWvSXYVkRRRlKJQ1qDWFZ\nMUVJxSJC09WlpRqVSkRR9fX1YrHYzMxMIpFoNJqioiJDQ0Nra+tRo0Y1Njbm5+fn5eUdOXJkxowZ\n3FpISUlJt2/fNjY2vnfvnkgkqqioULl7P/viGvehAfXZGfYlNwNt+59pVl1Nu5HnPogiJK383nI/\n30Y1XW9to29urlKp8vPzhwwZYmNjU1RUVFRUtHHjxjfeeEMkEs2YMWP48OF1dXWXL182NDTk1rvX\n/gSam5tJ+BzJ4V9L6pWEUIPHjG2Ueyhyc+vq6vbv308IcXNzmz17dklJyf79+7nx75MmTTIwMOjo\nj5oQQrm5S04d1abgiJDQESNHVh89mpeX5+Pjw+0TERGRkJBw+fLlxMREfX39SZMm2dradqKtNnGX\n4DwOcKYoiq8hIs3NzS3vc/Qkrg8UXwPqeTxxlmX5HVDPMEzru0GP1tzcrNN6eqcOBOH+/fu5JQU+\n++yzyMhIQsjkyZNXrFhhYmKyZs2avz9WJpO1/ItWW1vbagTbQ3fQ19d/1FEjR45MSUnhXu/YscPQ\n0PChi/P9Dc3QJ+iSYoZld1xOvXrnLpeC2keA5H4HUe1r8r9PCps0NCGEIsRMX1/N0DTDNtO0sZ7U\nTKrXKBLVq9XV1dUhISHXr1/nEt3Jyen999//8ssvS0tLXV1d7927Z2JiQtP0008/HRcX969//aup\nqUkkEt27d4+m6SVLlviOn6gvkTC/XSuorpWKxA1qTXZllUFGpsrIOKtJfdnCynvajLu1tWZmZkZG\nRpWVlRkZGampqR4eHhKJRKFQ2NnZyeXyrKysPXv22NvbFxcXL1iwoNWdRqVSaTw6mPXysbuQJDp6\n/IbP0GEmJsePH3d0dIyMjMzPzz927Ni1a9dCQkLWrl1bUVHRv39/rhNNZ5i4My+8Er93dwZDxs6K\nnLjmZUYklsvlp06dMjY21qbCtGnTpkyZUldXZ2JiorupSpubm7nFrXT07/89jUYjFov5CsJWy0b2\nJH4/f/B44izLajQaHt9vNE3r6+u3c38e59/hUQeCcNu2bcuWLfv222/J/Y91CxcuTE1N3bVrV5tB\nKJfLufXZJRIJwzAZGRmt7rA9dAcDA4O/P6orJMETG5Ivfno4ds+16yqNplUK/j2GZbl7pKZ6UoZl\nm2iGZdkGtUZfIqZEhGJpbprQK1euGBsbjx8/vqCgQKlU7t27Nzc318nJyczMTDshHEVR5eXl/fr1\nUygUGo1mxIgRcrlcrVaXxB0xOnPih7TMew2NA0yNyxoammmabWo2t7BobGxMvVvBlt+jKGrixIm3\nbt0yNDSsrq42NjYuKytTKpVLly41MzOjafrYsWPcVGobNmx4VIdbSmblPHX6SJF0kIeHWCyOiIiw\nsbExNja2sbEpKCgoLCwMCQmRyWQymeyhh7efyNl11FubzW7c+GtpYpp2d3c/cuSIQqFwdXX97+9F\nIrG0tOxiWwAA7deBICwoKHj77bdbbQwMDNy+fXubxw4fPtzW1vaLL75YvXr1119/bWBgEBwcTAiJ\njo728vLy8vJ66A7cw60Hj+oWZWVlb+feOp9b0Eyz3JUfpR0a0eIuaMsvuaTkPslTFGUgkeiLRE00\noy8SqRmGEFZmaFCraq5Vqy0tLRsbGysqKiQSSXZ2tqWlpZWVFcMwenp6XK5rO15WV1efP3++uLjY\n1dV15MiRwcHBPj4+9bnZkq//9daFFH2xZJKrU01T0yh7OzFFWbl7GMvdL1y4UFtba0RYTVVlzh+/\nX6uuVavVTU1NpqamgwcPPnv27GuvvSaVSs+dO5eWllZUVFRRUfHjjz96e3trL5pVKtXZs2ezsrJM\nTU3lcvnw4cPDwsK4b2kziWVZtVrd0evsB3FtFRQU/DWH3PDh2p8n143o4sWLLYMQAKCHdeB+vYeH\nR2JiYquNN27ccHd3b/NYiqIOHz4cExNjY2Ozb9++2NhY7ppyzZo1x44de9QOjzqqi27dujVmzBgH\nB4c9e/akl9+taGxkuXGBXB0Udf//qZbXiNxrsVgskUoNDA0NDQ2NDA3qaFqpVlMiIhWLxCJRVaNK\nqVazIlFoaOjMmTPNzMzq6upqa2vv3bt38uTJH374IT8/38bGZvHixdofGresUv/+/deuXbtq1So7\nO7urV6/eunq5qLYupbS8pK5upMOAZpr5j6IoXdloN/QJb29vIyOj+uJbedevW6mbK3JzmnOy+hsZ\nicXiuXPncsPzd+7cef369bS0NJlMFhkZuWjRooyMjN9//51rsbq6eufOnSkpKZaWlgzDJCYmcl03\nOYmJiVlZWWq1+vTp0wqFIiAgoCs/am1bMpnswbYIIUFBQdnZ2a0eGAMA9KQOXBG++OKLzz77LDcp\nCcuyZWVl586d++STT7Zt29aew+VyeUJCQquNLTt8PnSHh27siqSkpJkzZ1ZWVv7PNR8hzP1LPcKy\nXAy2vBzkApiiKC8vr5qamnv37hkZGjarVAzDEopqZilDPalao6FZQkmlDnZ2KpVq2LBhKSkpVVVV\ndnZ2EyZMuH37trOVLMLbw9Pd3ex+9xBCiFQqValUIpHIwcFBO68Ke7ecupElMzTIrar5JTO7oKrG\nfGiAa0NDTm7uwkWLbuZkZxTfypeIr5bdVdO0iqYlNZdpQ6OEhARDQ8Pg4ODk5OSCgoK6ujoPD4/5\n8+dz4/+uXLmycOFClmWjo6O5BYGNjIyMjY1bPjtpaGhIT08vLS0l9xdmGjhwYKd/1C3b4hajePA5\njY+Pz4kTJ/7888/FixdjilEA4EUHgnDFihU1NTVbtmz56KOPCCEhISFGRkbr1q1bvXq1zsrrZhqN\nZunSpdXV1dzVnp6eHkVREomkqamJpmluvRL2/qwxRDtY4n4oikWi0aNHX7x4sbS0tLqpiSKEZllj\nqcTGyKhOKmVUTY2NjWJC+vXrJxaLr1275uTkVFVV1dzcbGxs7KCvZ5+TPsRIRBS5jcdjjw4Z7hUQ\n6O7uPmDAgH79+uXk5GRkZJw+ffqveVVyc3e9suZWTWGDRv1HbqF84MC3X3+9oqLil19+aWhoGCGX\np2jU60cEXbxdcrao+G5NI9vU7NjflqZpmUwWFBQ0c+bMU6dOSaXS5557zsXFhWGY5uZmrqdMdXX1\nrVu3IiIi+vXrp1QqCSHaZDp79qyHh8czzzyjUChqa2udnZ27+FywZVvclpZtyeVyBwcHqVQ6bty4\nP//88/bt2w4ODl1pDgCgczoQhBRFrV27duXKlRkZGUVFRQMGDPD29u70KoC8SEpKKisrE4lE3Jpw\nBgYGhoaG9+7dY+7PGsplXst7oX9tJ0QqEtEME3PggLe/v6+39/Xr1ymRSF8sZli2tF7ZTNNGhoZc\nP2kDAwOuC2h5eTlN06GhoSEhIZ+vfaVRRBiWJYT8nluQUVJhbjfA3d2doqj58+dv2rRp8+bN8+fP\n5+ZVsR0woHaA451r100MjRiWKA2NuCGVTk5Of/zxR0F2VmF1jcxQr6yhUUyJHExNI8eM6j9hskwm\n69+/f0JCwsKFC83MzE6ePHnjxg2RSPTTTz+VlJRwHZpqamoIIQ9GjlKpTEhIUCgU3Nj5bvlp/31b\nBQUFy5cvJ4QEBARwF8Td0igAQEd17JHbnTt3Dh48OGjQoDlz5jg7O+/evbtvTYKVnp6u0Wi0/ULN\nzc1nzZqlr68vlUq1V4TaZ4FE+6SQEGOpdKr7QCOJpLmxwdTUpPTOHZFI5G5pbqYn5f4xiqKampoM\nJJIBAwZcu3bt/PnzN2/erKqqYhhm2LBhKpXKtLmJYZlDOfmHcvLT794b7eocGhrKlfH0008PHz48\nLy/vt99+q6ioYFn2wIEDuXl5roM8bAe6mclk5ubmqamp+/fv5575WdnaMUbG8bdul9TXN2o0Q52d\nXvr085EjRxYXF3P9LVmWfeKJJ6RS6SeffLJkyZITJ04sWbJk2LBh5H7f6PLy8lY/GWNj44fPPtMF\nf9/WzZs3ubbEYnFgYGB3NQoA0FEduCK8ePHi1KlTVSrVlClTLC0tm5ubP/744/fee+/EiRPcYj29\nH3f9yo1o5q4IDQwMjI2NzczMFAoFN5K01SAKihCRiDLT11M1N5sb6ElFYnMjo2a1up+Z6V1lIyHE\nSCppoml9sdTCUL9c2WggEUulUu5upEwms7S0jI6ODgoKWv3U/Kq063uuXWcJFeXnPSE8vLi4+Ny5\nc/X19c7Ozv/617+ioqKuXbs2Y8aMfv36XblyRU9Pb+TIkU1NTXZ2dvn5+dHR0YGBgdzTPqlU6jX0\niRxC+ptYOpmaOHl4XryeFhIScuTIkevXr8tkMm7FhvXr14eEhNTU1AwdOpRb24EQYmdnZ2pqyi00\n2OqH8/DZZ7qgZVutRo91e1sAAJ3WgSvC9evXDx48+M6dO/b29oQQV1fXW7dujRo16tVXX9VZed1s\n7ty5hoaG3JWfSCTihi6o1erg4GAXFxdyfw4I0iIIDUQiiUhU2agqVTbQLHEyM2XEEkNDw2dXvWBs\namJgoO9mYeFgamKqJw11dtQwTPndey4uLnK5fN68eRcuXNixY4eHh8fTTz8d+MKa0EkTfa2t7yob\n9P38c4zMvvnmm/LycqlUmpSUtG/fvv379y9cuLCkpOTChQsqlcrGxubevXvLli1buXKlv7//zZs3\nY2NjP//887179xYUFOTk5NzIyjLo1++JkaNCQ0P19PRMTU3t7e3PnDnDjRdkGObw4cNpaWkTJ07U\npiAhhKKo4ODgoqKiuLi4/5lJlWXT09MtLCy6cXrPnmwLAKDTxJs2bWrnrqtXr966dWvL/vQSiURP\nT2/Xrl1vvPGGTqprn0uXLg0aNKg9TytFIlFqauqNGze4IGQYRqVSKZXK+vr6e/fuNTY2EkK4TjQm\nJiZcJ5nnhvo2qjUsReqbmu81Npo7OKrU6oqKCueBA01lssq6Ol8TwzA317S7FSX19QzLqkViS0tL\nmUzWr1+/sWPHOjk5yWQyb29vysBQPDRo5KyIkqbmOgur3Lw8d3f3RYsW+fv7Ozo6Jicnc7ONDBw4\ncNasWWKxeNmyZf/4xz8CAgJcXV1Hjx6dnZ2dl5dXW1tbVVXFzdymVqvVavWiRYtCQ0O56Tqbmpqi\no6NnzJhhZ2cXExOTlpYWGBj44AO/AQMG3Lt3LzU19fbt287OzoaGhiqV6o8//sjMzJw1a5a2Y0u3\n0LZVWlpqb2/fsq3w8HAbG5tubKv9aJqmKIrHKa+4oUG8tK5Wq/maOoT7MMTXFGs8njghhN8p1liW\nlUjae/OvqKiosbHRw8NDp1X1Nh24NWpnZ/fgE8HCwkK+/px1zvfff3/8+PGKiorm5maappuamggh\nGRkZ3J8nQohIJJLL5VFRUd9++21pcXFRbb1IRFnqG5Q3NBgZG4sMDJYsWbJt27bo6GhCiEwmk3h6\nXrt9q665aYC19fQJE/5z8ZKTk9OdO3eampq4MekVFRU7d+7UDid/7oUX9+zZo9FogoODuRblcrm3\nt3d2dvadO3ciIiL8/Pxqa2vr6+u1P9i0tLT169cXFBSkpKQwDOPk5DRs2DB7e/uJEyeOHj1ae2pc\nJ8xLly7dunXrxo0bISEhISEhD/4EKIqaPXu2tbX1yZMn/1oQWKnU09MLDw/v9uUetW3Fx8fn5ORo\n25oxY4bQ/ksDgF6rA0G4YMGCzZs329vbz549m5sh5dChQ1u3bn3++ed1V1+3k0qlv/3227Rp0xoa\nGjQazV9zytx/LiiRSExNTf39/WNiYqqrq00tLO4YGMktZSKRqEDZIDY1S09P/+CDD6ytrW/cuNHU\n1OTu7u43cuTNmwMGGRjb2NgkXLlmbm4eFRW1cePGrKys6urqf//730ql0sXFRaVSJSYmuri4ODk5\nhYWFXb58WXsJy7JsfX09V4ODgwN3RzE2NpZblaKhoSEhIUEmk/n7+1dXV9va2kZFRQ0YMMDT05Nb\nIPfs2bOVlZUjR460tLR0cXH5448/vLy8wsPDH5qCHJFIFBoaOmjQoOLi4qqqKktLS+2/1u24tvz9\n/bnFh7m2uBngdNEcAEBHdSAIN27cWFpaunDhwoULF1pZWVVVVdE0PW/evC1btuiuPl0IDg7esGHD\nl19+WVNTw96fNZS7byOVSi0sLG7evKlSqczNzS0sLFasWMEwTGlpKcnOzsnJMTIyqq+vNzIycnZ2\nLi8v524wDhw4cMSIEUeOHFEqldu2baupqbG1tR03blxcXJxarba3t1cqlcuXL+cGFBJCAgICBg8e\nTFFUVlbWgAEDLly4oFAopk6dWlpaWl5ebmVlFRgYqFAoUlJSamtrw8LCAgIC/vzzz8LCQqlUumHD\nhpbr4iqVylOnTqWnp9fW1i5dujQ0NPT06dN/n4JaFhYW3OPeHtBy8WHSvgldAQB6RgeCUCKRfP31\n16+99lpycnJRUZGtrW1gYKCvr6/uitMRiqJee+017olafX29iYlJU1NTc3OzkZGRTCYLDAz09PQ8\nd+5cRkbGqFGjCCGDBg2KjIy8cOHCe++9t3z58tzc3MzMTJFIZGZmVlxcrFAo7t69e/v2bZFItHHj\nRoqiMjIyBg4cyHXjNDMzu3nzZkhIiFQq1faQjI+P9/DwSE1NTUpKIoSIxeKpU6cOGzYsMTFR28ey\n5R3Fmpqay5cvE0IcHR1bXbcZGxuPHDmyoKAgPz9frVaHh4e7urp2YgV5AADB6kBnGY6VlZWfn9+Y\nMWP8/f17ydPB9neW0RKJROPGjXN0dGxoaGBZdsCAARYWFgYGBl5eXq6uroWFhadPn/bx8ZkwYUJT\nU1NmZqanp6dMJktKSqqtrV28eDHLspWVle7u7sbGxi4uLtzk1DNmzGhoaKiqqho5cmRlZaWnp+fv\nv/8ulUonTZrU8vpMqVQeOHDg7t27kZGRXP/SsLAwNzc3bmGglJSUhoYGd3d3kUjk6uo6ZMgQS0vL\nnJwcOzs7Pz+/8vJyiUTi7+/fssdBfX19Wlqanp7eiBEjTE1N2z8ynd/uA9qJ63oeOsvw0jQ6y/DV\nNDrLtKkDV4SVlZX/7//9v5SUlAef7miXBuxDRCJReHj42LFjU1NTVSqVpaWlra3tnTt3Kisrb968\n6eDgsGXLFm6wHTdDZklJiVwuT01N3bFjh5GREbfCQ1JS0ooVKy5evHj8+PERI0YUFxePHz/ezs7u\nwoULsbGxZWVlTz75ZKu7lNxw8uTkZI1G4+TkxI1o5L7V6o4oNyKwtLTUyMhowYIFdXV1RUVFp0+f\ntrW1nT59uva55o0bN+rr6z08PDAaAQCgEzoQhCtXrvzjjz+eeuqp/v37666gHmZpaRkUFKSdA9PZ\n2bmqqurkyZMeHh7avOe+a2dnV1dX5+joyI1hSEhIsLS0nDVrlpOTU0VFRXp6emBgIDenzNKlSz09\nPTMzM729vblVMR81nLzVDNSP6mPJ9edkWbawsPDXX3/97rvvSktL58yZY2RkdPjw4djYWGtr66lT\np2LSagCATuhAEB4/fvyTTz558cUXdVdNb1BTU2NqamphYdFqShTuJmpeXp6FhcUTTzwxb968IUOG\niMVi7fDw/v37c5d6arV6+vTpzs7OYrFY2/lTm1Ith5Nr5zjVemgfS+65IEVRc+bMsba2/uGHH37+\n+edff/2Voiiapr28vKKiorp95AMAgEB0IAjt7e27sihPX8E9SPD09ExPT2+ZYcbGxtOnT7969apU\nKi0qKkpMTBwyZIhKpYqLi8vLy1uwYAG3jDB3qWdmZjZixAiWZVvd6my1/6NqaNXHUkskEo0fP37o\n0KGXL19OT08nhPj6+gYGBupo5AMAgBB0IAiXL1++e/fuyZMn8/XUt2dwM2TW1NR4eXm1yrDs7GwD\nA4PIyMgrV65ER0enpqY6Ojrq6+trb122mjns7291drpCS0vLiRMnTpw4sZvOGABA0Do2s0x5efng\nwYOnT5/er1+/lhc069at00Ft/NCOZw8ICAgODk5ISOAyrL6+PjMzMzAw8Lnnnquqqtq1a1dGRoZI\nJHrqqae48fIPvdTT3urMysqTGU2LAAAgAElEQVQ6f/68Uqk0NTVVqVTaUfwAAMCvDgThW2+9xb34\n5ZdfWn3rcQpCcr/35uXLlwcNGrR48eK7d++WlZVlZmYOHDhw6dKlFEXJZLLXX3/97Nmz8fHx3333\nXZuXetzqSHV1ddySTObm5khBAIBeogNBqFAodFZG7/KoW5pz587V5tzf9Gp5qPj4+LFjx44fP54Q\nUllZefXqVSMjI1dX1/aP7wEAAF3o6l/h1NTUEydOPGZXhKTdOfeoXi2tcD1ouOWQkpOTjx49yvUX\ntba2joqKajWIAgAAelIHgpBhmJ9//jk1NbVlp/8zZ86UlpY+fkHIaWfO/RfLMqlXGEUB0dcXjxhD\nWcq4zRRF2dvbX79+3cbG5ujRo2PGjBk7dmx+fv6BAwfOnDkzbdo0nVQPAADt0IEg3Lx58zvvvDN4\n8OCcnBwLCwsnJ6eMjAw9Pb1///vfuquvb9GcPEqfjONe02dO6m14hzL/a+K3oKCgQ4cO1dfX29vb\nh4aGUhTl6enp5+f34MpWAADQkzow79/333+/atWq69evf/PNN0FBQcnJyYWFhdbW1g+OChcoltWm\nIEvIlTvlX29955tvvsnKyiKEDBkyxMfHp7i4uOUR9fX1mBcNAIBfHQjCkpKSsWPHEkJCQ0OTk5NZ\nlu3Xr9+LL764detWnZXXp7T4QHCi8Oaf+YUShm5oaDhw4EBlZSVFUTNnznR1dS0qKoqLiysqKjp6\n9GheXl5QUJCOymFZ9uLFi9u3b//ggw9+/fXX+vp6HTUEANCndSAIrayssrOzCSF2dnYURaWlpRFC\nLC0tr127pqvq+haxWOTjRwi5U69MLrkz1c116fOrnn32WX19/dLS0sbGRj09vcWLFwcHB6elpe3d\nu/fq1avh4eFyuVxH5Zw8efLo0aOOjo6DBw/OzMyMjo7WUUMAAH1aB54RhoeHf/HFFw4ODs8880xA\nQMDnn3++adOmX3/9dcCAAbqrr2+RRMzTSCQlx44SQgU8/6LI3fN2QUFTU9PBgwcJIe7u7vPnzx8/\nfnxwcHBdXZ2JiYnulqSpqak5d+5cWFgYtzahvr5+QkKCSqUyMDDQUYsAAH1UB4Lwo48+omk6Njb2\nmWeeee+990aNGvXNN9+IxeIff/xRd/X1LZSpmfTpZQ5jJ4j37r1CJMNo+tChQ/3794+MjMzPzz92\n7FhiYmJISIhEIuGG2D+4oFUnsGV3mNwsysRE5OtP7o9KvHv3LiFEOzesTCYjWBceAOBhOhCEZmZm\nX3/9Nfd6yJAhJSUlycnJcrnc1dVVN7X1GSqV6uzZswUFBfr6+nK5fNiwYSNGjOBWroiIiLCxsTE2\nNraxsSkoKCgsLGy1PGEX0ZeTNb/s416L5Beki1cQA0NCSP/+/SmKunDhQnh4OCGkrKyMW92QEMIw\nTHNzMy4NAQA4nR9Qb25uPmnSpG4spY+qrq7+9ttvGxoa5HJ5XV1dYmKii4tLWFgY913tpwSWZbmF\n7Lu3dW0KsoSkJCbeyL8pHuQ1atQoT0/P0aNHazQablLTnJwcbk4chmFiYmLu3bu3fPny7q0EAKCP\najsIly1b5u3t/eqrr3788ceP2udxHVDfJpZlo6OjNRrNypUr+/XrR+4vZ6/dITEx0dra2s3NLSEh\nQaFQLFmypDub12i0L08U3kwuuTNQZqNsaDhw4MDq1avHjx9PURRFUfX19ZWVla6urlwKpqWlhYSE\nSKXS5ubm7iwGAKBvajsIv//++wkTJrz66qvbt29/1D6CDcLq6upbt25FRERwKUjuL2dPCDl79qyT\nk1N6enppaSkhRCQSTZ06tZsXdJRIRJ7eTFaGtp/qiKgotYf3p59+WlpaamhoyN0Lra2tJYRYWFho\nU7B7b88CAPRpbQdhXV0dtwDhyZMnZTIZ1+0CODU1NYQQBweHVtuVSmVCQoKjo+MzzzyjUChqa2ud\nnZ118aOThM3QqDUl5xIJoZ4InyUa7H+7sLBVP1VTU1OKon7++efKykqkIABAK2133zcxMeEuLIKC\ngjZu3Kj7kvoSbjn78vLyVtuNjY0DAgIUCgXDMG5ubkOHDtXRBwjKzl66YpX9q2+Jx024OsCZZhiu\nn+oLL7wwZcqU3NzcxMREU1NTT09PpCAAwEN1oLPMmjVrfvzxx/fffx+rJWhxy9knJSVxfURbfsvW\n1pZbdEIqlXZ7uw/ppzp27N/0U50+fbqzszM3phAAAFrqQBB6enoOHDjQx8dn9uzZNjY2Lf/uC/YZ\noXY5+7i4uGnTpmmX22VZNj093cLCQhdTiXain6qxsTFSEADgoToQhG+88Qb34rfffmv1LcEGIbm/\nnH1KSkptbW1YWJhMJlOpVHFxcXl5eQsWLOj2leh57qcKAPDYwQr1XfWo5ezDw8O1y9l3I577qQIA\nPHawQn03aOdy9t2C936qAACPGaxQ3206vJx9p2j7qVpZWbXczvVTTU5O5vqp6roMAIDHBlao72P4\n6qcKAPC4wgr1fQzXT5Vb3bflahI67acKAPAYwwr1fU9gYKCPj09KSgo3WQwhRKVSHTp0KC8vLyws\nrNv7qQIAPN46cGv0wRXq/fz8sEJ9z+vhfqoAAI83rFDfJ/VkP1UAgMcbVqjvw3qmnyoAwOMNK9QD\nAICgdaCzzMqVK+Pj47V9RLkV6pGCAADQp3UgCGNjY8eNG+fi4vLGG2/cuHFDdzUBAAD0mA4EYVFR\nUVJS0rx58w4cODB48GB/f/+PP/64uLhYd8UBAADoWgeCUCQSjRw58uOPP87Pz79y5cr06dP37t3r\n5OQ0fvx43dUHAACgU52ZdJuiKBcXF29v79zc3IKCgvj4+G4vq6NomtZoNJ07lmGYTh/bRSzLUhTF\n19Q8PJ44IYRlWb5ap2maoii+Zh5gWZamaV6aJny/27X/2/P4fbcTQnh8t3foz6MwZwrrWBAqFIoj\nR478/vvvZ8+epShq4sSJO3funDlzpo6Kaz+RSNRq4s32oyiq08d2EfeeE4k6cF3ejXg8cZZlGYbh\nq3UuCHlsXSQS8RXDeLf3PO6jD4+tsyzb/taFOTVVB4LQ398/NTVVT09vypQpe/funTFjhqWlpe4q\n65CufMDn8eKA39b5PXHC339y1H28tE7wS+epXSGfePtbF2YQduDT2cCBA/fv33/37t3Dhw8vWbKk\n96QgAADwgmXZgwcPhoSEWFhYWFhYDB8+fN++fe28v2pra7t9+/YHX/e8DlwRRkdH664OAADoc9as\nWbN9+/bJkye/+eabenp6J06cWLx4cVZWVt9ajKGrK9QDAIAwnTlzZvv27du2bXvttde4e6r/+Mc/\n1q1b9+677y5ZsmTQoEF8F9he/Dy4BgCAvm7jxo2+vr6vvvpqyyeLr7/+uqur65kzZ7gvWZbdu3dv\nQECAsbGxn5/fTz/9xE+tfwtBCAAAHdbc3JyYmLho0aJWPYGtra0LCgpWrlzJffnpp58+88wz3CIN\nQ4cOXbhwYS9cp6Ezt0ZZlo2Jifn11181Gs28efNmz54tzI5GAACCpVAoGIYZOHDg3+xTX1//zjvv\nrF+//sMPPySEzJ8/n2GYzZs3L168uKfKbJfOXBF++umnzz333ODBg5944olVq1b985//7PayAACg\nN1OpVIQQfX39v9knIyOjpqZmyZIl2i3z588vLCxUq9U6r68jOnNF+Nlnn+3Zs2fWrFmEkEGDBq1e\nvXrt2rXdXRgAAPRebm5uhJD8/PwHv/Xdd99RFLV06dKioiJCSEBAgPauIcMwDMPcvn3bxcWlB4tt\nQ9tXhK+99lplZeX/HCMSaSfs0Wg0fM2YAAAAfDE2Nvby8vrll19aTZvX1NT00ksvXbp0iRBia2tL\nCDl+/Pi1+65fv56Zmclt7z3aDkKpVDp48OB33323vr6e2/Laa6+tXLly8+bNmzZteu6551599VUd\nFwkAAL3Opk2bLly48Omnn7bMwg8//FCpVHJTb3p7exsaGubl5Xned/ny5W3btv39DdWe1/at0a1b\nt65evXrr1q1eXl7r169/7rnnVq1a5ezsfODAAZZlf/zxx+nTp/dAoQAA0KvMnTv3l19+Wbt27dGj\nRydMmGBkZHTmzJno6OilS5dOmjSJEGJpablu3bo1a9bcunXL19f38uXLn3/++VtvvdXb+le26xlh\n//79v/zyy7Vr127cuNHLy+vtt99etGjRk08+qeviAACg16Io6sCBA3v37t27d++WLVskEsmgQYN2\n7969fPly7T6bN2+2srL6+uuvP/roI2dn5/fff3/NmjU81vxQVEdXRUlPT3/rrbeys7O3bNkSGRnZ\nG4J9x44d06ZNc3Z27tzhSqXS2Ni4e0tqJ24ZBL7m4+fxxLn5+CUSfiY2am5upihKKpXy0jr3WJ2v\n/3D4fbcTQvjqUsDvu12j0fD4fqNpuv23Is+dO1dRUREeHq7Tqnqbtv8ENzc3b9myxc/Pz9vb++23\n33Zzc4uJifnuu+927NgRFBR0/PhxvhYYAwAA6Lp29Rrdv3//5s2bv/jii4SEBG6kxLBhw06dOvX+\n+++/9dZboaGhuq8TAABAJ9q+N7V///6ffvpp4sSJhBB/f3+5XL59+3buxs7EiRMnTJhw6NAhnZcJ\nAAC9BsMwnRgUL5VK+XoS9PfaDsJ+/fplZWVxQZient6vX7+WjzcoioqIiNBhgQAA0MsoFIqampoO\nPepmWVYmk3W6M4dOtR2En3zyyezZs0+ePGliYhIbG/vDDz/0QFkAANBrMQzj5ORkZWXV/kPKy8uV\nSqXuSuqKtoMwLCzs+vXrR44coWn67bff7kNLTAEAALSpXf3X3dzcXn75ZV2XAgAA0PN643NLAACA\nHoMgfGwxDHPu3LkdO3bs3LkzKSkJwz0BoPf4z3/+M3nyZGtraxsbm7lz5xYWFj5qT4VCYWFh0fJF\nt0MQPrYuXrx48uRJGxsbU1NTc3Pz3jAHEAA8xti6Wk3ML+q9OzXH/iCNjX+z508//RQREREZGXnh\nwoXjx4+LxeIpU6Y0/u0hOoUgfGzFx8ePHTt2zpw5ixYtsrOzu3r1anZ2NjfNFQBAN1M3aw78SF9I\nZHIy6f8cV//8HXnEX5uGhoZ169Z99dVXzz//vFwu9/f337dvn1gsPn36NCHk2rVrY8eONTEx8fX1\njY2Nfei/wLLshg0b+vfvb2JiEh4eXl1d3cXa+ZnsEboRW1tD6usoSytiaPjfjSzb2NjILfqVnJx8\n9OhRhmEIITY2NlFRUYYt9gQA6DrmloLJzf7vl9mZ7J0Syt7xwT2vXbtWV1c3b9487RaJRJKZmUkI\nqaurCwsLe+utt/7444/jx4/PnTv3+vXrD05KHBcXt3v37tOnT1taWs6aNWv79u1vvfVWV4pHEPZt\nmuOxmlPHuJue0oVRIr+h3HaKouzt7a9fv25jY3P06NExY8aMHTs2Pz//4MGDp0+fxsohANDN1JpW\nG1i1+qHPYwoLCx0dHR86xczRo0ednJxeeuklQsjcuXN//vnn+Pj48ePHt25KrWZZtrKy0sfH5/Tp\n09yn/K5oOwjXrVvX5j4ff/xxF+uATmCLbtL3U5AlJOnjbdcCRitVKjc3t7CwsKCgoEOHDtXX19vb\n24eGhlIU5enp6efnV1RUxHPdAPDYETm7/s+X8kGih10OEkLs7OxKS0tZlm3ZcSExMZGiqFu3bqWn\np7u4uHAbGxoagoKCHvwXZsyY8f7777/88su3bt2aP3/++vXrzc3Nu1J820FIUdTevXurqqpkMpml\npeVD90EQ8oIpKda+/o+i6Pzt0oDxU/Xc3S9fvtzQ0LB48eLc3Nz09HRHx/++Hevr601NTfkoFgAe\na4aGemvWa47/yWSli4YGSsZPIY9YeWro0KE0TR85ckS72BNN088+++yLL75oZ2c3atSoEydOcNvL\nysqMjIwqKipa/QtFRUUTJkx49tlni4uLN27cuGjRonPnznWl9raD8KOPPlq5cqWXl9cbb7zRnqtD\n6DGU7K/5jWqams/fLpns6jx2wVPEwEBfXz8hIaGpqWnmzJkNDQ2FhYVxcXGDBw9OT0/Pzc1dtGgR\nv2UDwGOJsneURj3X5m6WlpZbt25dunTphx9+OHHixKampo8//rimpmbBggUURa1Zs+bnn3+eMWNG\nUlLS7Nmzz58/b2Ji0upf+OOPP7Zv3/77779bWVkZGxu3f7XFR2nXM0J3d/fhw4d3sSXodiK5h/iJ\nIPrKpXsNjYQQ+dyniIEBIUQmkxFCWJbV09NbvHjxmTNnLl26lJycrK+vP3PmTLlcznPdACBsa9as\ncXR0/Oc//7l+/Xo9Pb1Ro0adOnXK2tqaEHL48OGXX355xYoVTk5Oe/fu9fX1VSgUrQ5fvnx5UlLS\nsGHDNBrNsGHDduzY0cV62ttZZtu2bdyfV+hFKEoyf7Fo2KgBRTfFf8RdImLuRkNZWZlMJtN2DR01\nalRwcHBdXZ2JiQlfy2QDtMKy7JUrV65du0ZR1KhRozw9PfmuCHpURETEQ1cuGjVqVHJycsstLi4u\n3AAJ7QtDQ8P9+/d3YzHtHUcYHBzs6+vbjQ1DdxG5ulkEjx8zeYqenh43fUxOTg73Z4VhmJiYmO+/\n/55lWUtLS6Qg9B7Hjh07cuSIVCptaGg4cOBAZWUl3xWBcGH4RJ+kUqnOnj1bUFCgr68vl8uHDx8+\nfvx4iqIoiqqvr6+srHR1deVSMC0tLSQkBBEIvUppaemFCxemT58eGBjY1NT06aeflpaWGhoaYoQr\n8AJB2PdUV1d/++23DQ0Ncrm8rq4uMTHRxcXFycmJ+25tbS0hxMLCQpuCISEhfJYLAqdqpK9dJgzD\nug2irPtx20pKSiiK8vf3J4Tcvn27qanp4MGDhBB3d/f58+c/OIAaehuKooqLi0tLS9t/CE3TXRzk\noDt4w/UxLMtGR0drNJqVK1f269ePEKJUKo2NjbU7mJqaUhT1888/V1ZWIgWBX6yyvvmdDdovxYuW\nk8H+hBAbGxuWZS9dujRs2LBDhw71798/MjIyPz//2LFjiYmJeNP2fk5OTmq1uqNH9dpbUxhQ38dU\nV1ffunUrIiKCS0FCiDYFz5496+HhYWtr6+npmZmZiRQE3tFJ8drXLCGpvx3MTc9iWdbb23vYsGGD\nBg0Si8URERE2NjbGxsY2NjYFBQWFhYV43/Z+Uqm016ZaJ2BAfR9TU1NDCHFwcGi1XalUJiQkKBSK\npUuXTp8+3dnZecSIEXwUCNBCzX9nQz6luHXhdqnXIB+NRhMTE+Pm5sY9EXR1/WtGEpZl1Wr1g4PG\nAHSt7V6jH3300cWLF8Vi8RtvvJH3CD1QKHD09PQIIeXl5a22GxsbBwQEKBQKtVptbGyMFITegHL6\nK+QqGhsv3C6dOmnS008/vWTJkjFjxuTn5+/bt0+tVicmJmZlZanV6tOnTysUioCAAH5rBgHCgPo+\nxs7OztTUNCkpibut1PJbtra23KITj9Mti96JYZjDhw97enpi9NvfEweNYEuK6fMJJfVKYtVv6MoX\nuO2DBg1KTEwsKSmJjo6urq7m+lyIRKKpU6cOHDiQ15JBiDCgvo+hKCo4ODg2NjYuLm7atGnaWWtZ\nlk1PT7ewsMBUorqmHZdiZmaGIGwDRUlmzRVPmmp186bowC/ZZeX+dgO4bjLOzs6urq5nz55dvXp1\nVVVVbW2ts7Mz/sgAL9obhMHBwTqtA9ovMDBQoVCkpKTU1taGhYXJZDKVShUXF5eXl8dN1sd3gY8z\nbQqOGTMmNDSU73J6r1ZDXYOCghwcHGJiYvLz86urq4uLixctWuTk5HTx4sWcnBzcyQd+YfhE30NR\n1OzZs62trePj43NyckxMTJRKpZ6eXnh4OC5QdIplWW0KTpgwobi4+Pz583V1dU5OTsHBwV2f+fex\n8dChrk8//fSZM2eKiopMTU2XLl3q4uLCsqxUKsVHN+AdgrBPEolEoaGh/v7+OTk5VVVVlpaWnp6e\nvXawap/EMEz6dba6SuQ2iBpgz22jKIp7LssNUDl48KCVlZWZmVlSUlJ2dvbKlSu5rkwC96ihrjRN\nh4WFcT/A3NzcsrKyK1euKJVKNzc3vksGoUMQ9mGWlpboxKQTGrX6h71Mdgb3lWRGpHhMCPf6ySef\nLC4ujo6Opmna19c3IiJCJBLl5eX99NNPSUlJGABH/naoa3x8vJeXV1NT008//cSyrEgkCgsL49Yc\nAOARghCgNTr1qjYFGZa9uGfXiOGjRVIpIURPT2/u3Ll79uzRaDTBwcEikYgQIpfLvb29i4qK+Cy6\n1/j7oa63bt1aunTp6tWrKyoq+vfvb2ZmxkeNAP+jvatPAAhIi2HgjRrN8cKb0Qf+zTAMt6V///5h\nYWF2dnYWFhbcFpZl6+vrMRKc056hrjKZzN3dHSkIvQSCEKA1ytlF+9pYKh1gYnwjLz8mJkabhQEB\nAcuWLaMoKisrq6am5ujRowqFYujQofyU28toh7rSNN3qW9qhrrwUBvAouDUK0JrIbZB4/GT6P8cJ\nIZWNqjteflZWVmlpaYQQ7qEgRVH6+vrHjh07f/48uT8S3MXFhd+yewkMdYU+B0EI8BCSKdMlwRNY\nVWPm5Svud+/OnTs3MTHxzJkz5H4WEkImT57s5eVVV1fn6OiIu3wtPXSoa2xsbH5+/lNPPYXxEtDb\ndCAIq6urJRJJqwch9fX1Go1G+7AEoE97cMXjkPHjWZYVi8Vcj9CWWUhRlHYZSGjpoUNdJRLJjBkz\nMNQVeqEOPCO0tLScNWtWq42zZs161JIUreTn548bN87MzGz06NFZWVkP7hAfHx8UFGRkZOTh4fHD\nDz9wGxcuXGjQwp07d9pfMECHVFdX79y5MyUlRSaTMQyTmJhYVlYmEom0c7pyK1ulpaW1fF4ID8UN\ndV2zZs3UqVN9fX3DwsJWrVqFx6jQO3XginDFihVeXl6tNrbz0QjLsjNmzIiIiDh48OCuXbumTZuW\nm5vL3V/i1NbWRkZGbty4cdmyZefOnZs9e/bAgQPHjBmTlZW1e/fuYcOGcbthyBHoSJsrHnO014Ve\nXl7e3t49X2ff0nKo64N9ZwB6iQ4E4ddff/3gxrVr17bn2OTk5Nu3b2/cuFFPT+/NN9/cvn37uXPn\nxo4dq90hJSXFyMjopZdeoigqLCwsLCzs1KlTo0ePzs/PDwkJwQ0o0LX2rHjMfRkSEuLi4oKuMQCP\njR7qLJObm+vj48MNMBKLxX5+frm5uS2DMCgoKCkpiXuKrlQqU1JSIiMjKyoqampqNmzYcP78eUdH\nxw0bNkyePJnbv7Cw8JdffuFe5+XlTZ06tdOfNxmG4euzKk3TFEWxLMtL6zyeOMuyPLbOMAxFUa1a\nr6ysZFnWzs6u1XalUsk9NVyyZIl2o6OjY1feb4QQvjqM8Ptu56VdjmDf7TRNd6h1Yd7z76EgrKys\nbNln2tzc/N69ey13MDU15XZITU2NioqaMGHCvHnzMjIyXFxcpkyZsmnTpuPHj4eHh1+6dGnw4MGE\nkMbGxoKCAu7YmpoatVrd3Nzcudpomu70sV3E/UXm628ijyfO/WlotZ5ij9FoNBRFtfoPnovG27dv\nt+rcL5VKhwwZcunSJaVS2S0LPdI0zXW06fo/1bnWeXy3E/7ikPd3O18BQ9N0hz5qC/MOtg6D8Jtv\nvvm///s/Qsi8efOCg4Pr6+u136qtrbWysmq1v1KpfPPNN3/77beNGzeuWLGCoqghQ4YUFhZy35XL\n5bGxsYcPH+aC0Nvb+6uvvuK+tWPHDn19fUNDw87VyTBMp4/tIu6KsOWz0p7E44mzLEvTtETCz+id\n5uZmiqJapZqrq6ulpeWVK1f8/PxaJbSTk9PVq1cJId3y49JoNGKxmMcrQh7f7YQQvj798Ptu12g0\nfC2XrdFoaJpu/9IowlzWW4d/gpcvX15dXV1dXb179265XJ6ZmanRaAghDMNkZGS0mnKepunw8PCy\nsrKMjIxnnnmG+zORkpJy5MgR7T6GhobC/CVBD+CGgRcVFcXFxbX8BI1h4ACPvfYGoUajKSwsbGho\naLVdqVS2Z0jD8OHDbW1tv/jiC7Va/dVXXxkYGHAr/UZHR2dmZhJCjh49mp6evmXLloqKCoVCoVAo\nampqGIaZM2fO77//XldXd+TIkRMnTsycObODJwjQXoGBgT4+PikpKT///HNlZSUhRKVSHTp0KC8v\nLywsDMPAAR5Xbd+bomn63Xffff/99xsbG/X09F544YUPP/xQe2W2c+fO9evXt3kPmqKow4cPR0VF\nbdmyxdvbOzY2lrsfuGbNmnXr1nl5eV26dKmsrMzd3V17yLZt215//fVdu3a98sorpaWlXl5eBw8e\n9PDw6MLJAvwdrHgMIExtB+G//vWvTZs2rV69euzYsefOnfvyyy/v3r27b9++jrYkl8sTEhJabSwu\nLuZebNq0adOmTQ8eFRUVFRUV1dG2ADoHKx4DCFDbQfjVV1+98MILn3/+OSFkzpw5Q4YMiYqKmj17\ndkREhO7LA+ABVjwGEJS2nxEWFhaOGTNG++WyZctmz569bt06lUqly8IAAAB6QttB6OTkdPHixZZb\nPvvss8rKyhdffJGvkeAAAADdpe1bo88+++z69esbGhqefPLJyZMnGxoaOjg4fPXVV/Pnz6coipss\nBgAAoI9qOwj/8Y9/qNXqDz74YPfu3ZmZmVz3uXnz5lEU9dJLL5WXl+u+SAAAAF1p+9aoRCJ5/fXX\nKysri4uLW040PHfu3KKiopMnT+7evVuHBQIAAOhSe+e4oijK3t6+1UY9Pb0JEyZMmDChu6sCAADo\nIfzMcgkAANBLIAgBAEDQEIQAACBoCEIAABC09gZhTk5Oy96h33///UsvvXTy5EndVAUAANBD2g5C\nhmHefPNNDw+P999/X7uRZdnvvvtu0qRJUVFRfK28DAAA0HVtB+G+ffvee++9devWJScnazcuW7as\noqJi8+bN33333bfffmHHrlQAAB2dSURBVKvLCgEAAHSo7SD88ssvp0+f/tFHH1lbW7fcrq+v//bb\nb8+aNWvPnj06Kw8AAEC32g7CnJycSZMmPeq7oaGh2dnZ3VoSAABAz2lXZ5nm5uZHfUupVNI03X31\nAAAA9Ki2g3DIkCGnTp166IpLLMueOnXKx8dHB4UBAAD0hLaD8IUXXjh69OiHH37Yqncoy7L//Oc/\nT506tWLFCp2VBwAAoFttT7o9f/78xMTE119/ff/+/QsXLhw4cKC+vn5+fv6BAwcuXrw4Z86cqKio\nHigUAABAF9oOQoqitm/fPmfOnLfffnvDhg3a60Jvb+9ffvllzpw5FEXpuEgAAABdae8yTCEhIfHx\n8c3NzQqFQqVSubm5GRsb67QyAACAHtDeIOTo6ekNGjRIR6UAAAD0vHYNnzh16tSsWbN8fX3Hjx//\n5ZdfajQaXZcFAADQM9oOwpiYmIkTJyYkJNja2t68eXPNmjXPP/98D1QGAADQA9oOwq1bt/r6+ubn\n5588eTInJ+fZZ5/du3dvSUlJDxQHAACga20HYWZm5rJlyywsLAghYrF4zZo1hJCMjAydlwYAAKB7\nbQdhY2OjlZWV9ktu6m08JgQAgMcDVqgHAABBQxACAICgtWsc4QsvvPDyyy9zr7mZZWbPni2VSrU7\nVFdX66I4AAAAXWs7CNeuXdsDdQAAAPCi7SD8+OOPe6AOAAAAXuAZIQAACBqCEAAABA1BCAAAgoYg\nBAAAQUMQAgCAoCEIAQBA0BCEAAAgaAhCAAAQNAQhAAAIGoIQAAAEDUEIAACChiAEAABBQxACAICg\nIQgBAEDQEIQAACBoCEIAABA0BCEAAAha2yvU9wkajUatVnfuWJqmO31st6Bpmq92+T1xvlrXaDQU\nRfHStLYAvprm/ZfOMAwv7fJ+4ny1TtM0TdMiUXuvefj6c8SvxyQIJRKJVCrt3LHNzc2dPraLaJqm\nKKr979HuxeOJsyxL07REws/bj2VZiqL4OneNRiMWi/lKYn7f7YQQsVjMS+v8vts1Gg1frVMU1aF3\nO1+/IH7h1igAAAgaghAAAAQNQQgAAIKGIAQAAEFDEAIAgKAhCAEAQNAQhAAAIGgIQgAAEDQEIQAA\nCBqCEAAABA1BCAAAgoYgBAAAQUMQAgCAoCEIAQBA0BCEAAAgaAhCAAAQNAQhAAAIGoIQAAAEDUEI\nAACChiAEAABBQxACAICgIQgBAEDQEIQAACBoCEIAABA0BCEAAAgaghAAAAQNQQgAAIKGIAQAAEFD\nEAIAgKAhCAEAQNAQhAAAIGgIQgAAEDQEIQAACBqCEAAABA1BCAAAgoYgBAAAQUMQAgCAoCEIAQBA\n0BCEAAAgaAhCAAAQNAQhAAAIGoIQAAAEDUEIAACChiAEAABBQxACAICgIQgBAEDQEIQAACBoCEIA\nABA0BCEAAAgaghAAAASt54IwPz9/3LhxZmZmo0ePzsrKenCHhQsXGrRw586d9hwFAADQFT0UhCzL\nzpgxY8yYMXl5eVOmTJk2bRrDMK32ycrK2r1797X7rK2t23MUAABAV/RQECYnJ9++fXvjxo02NjZv\nvvlmXV3duXPnWu7Asmx+fn5ISIjnfRKJpM2jAAAAuqiHgjA3N9fHx0dPT48QIhaL/fz8cnNzW+5Q\nUVFRU1OzYcMGNze3kJCQ48ePt+coAACALpL0TDOVlZWmpqbaL83Nze/du9dyh9u3b7u4uEyZMmXT\npk3Hjx8PDw+/dOnS3xyVmpr67rvvcq+bmpomTpzY2NjYudrUanWnj+0ihmEoiqIoipfWeTxxlmUZ\nhhGLxby0rtFoKIrSaDS8tE7TtEgkEuAvnXuuIRLx00GP93c7j+83roB27q9Wq3VaT++kwyD85ptv\n/u///o8QMm/evODg4Pr6eu23amtrraysWu48ZMiQwsJC7rVcLo+NjT18+LCzs/OjjrKyspo4cSL3\n+vr161KplLtw7ASNRtPpY7uIpmmKovj608DjifMbhNyHD4mkhz4FtsJvEPL7bieE8Pjph8d3O03T\nfL3fNBoNwzDtP3e+fkH80uHvZvny5cuXL+deX7hwITMzU6PRSCQShmEyMjLc3Nxa7pySklJaWjpj\nxgzuS0NDQ6lUKpfLH3WUg4PDypUrudc7duwQiUSd/v115diu4zEIeTxxlmUJf//JcZ8/eDx3sVjM\nVxDy+24n/P3S+X23c790vlonHfmx8/XniF89dM7Dhw+3tbX94osv1Gr1V199ZWBgEBwcTAiJjo7O\nzMwkhDAMM2fOnN9//72uru7IkSMnTpyYOXPmo44CAADoLj0UhBRFHT58OCYmxsbGZt++fbGxsdzn\njjVr1hw7dowQMmzYsF27dr3yyis2NjYbN248ePCgh4fHo44CAADoLj1321oulyckJLTaWFxcrH0d\nFRUVFRXVnqMAAAC6Cy6wAABA0BCEAAAgaAhCAAAQNAQhAAAIGoIQAAAEDUEIAACChiAEAABBQxAC\nAICgIQgBAEDQEIQAACBoCEIAABA0BCEAAAgaghAAAAQNQQgAAIKGIAQAAEFDEAIAgKAhCAEAQNAQ\nhAAAIGgIQgAAEDQEIQAACBqCEAAABA1BCAAAgoYgBAAAQUMQAgCAoCEIAQBA0BCEAAAgaAhCAAAQ\nNAQhAAAIGoIQAAAEDUEIAACChiAEAABBQxACAICgIQgBAEDQEIQAACBoCEIAABA0BCEAAAgaghAA\nAAQNQQgAAIKGIAQAAEFDEAIAgKAhCAEAQNAQhAAAIGgIQgAAEDQEIQAACBqCEAAABA1BCAAAgoYg\nBAAAQUMQAgCAoCEIAQBA0BCEAAAgaBK+C+geLMuyLNvzx3Ydj63z27T2f4XWOu9N89g64fWXLsz3\nG3tf+/fXaT2902MShAzD0DTduWNZlu30sV3EsixFUTy2zlfT/LbOMAy/P3aGYXhpmvD9bie85gGP\n73ZCCI/v9g79eUQQ9mFisVgi6eS5iESiTh/bRTRNUxQlEvFzg5rHE+f+KvHVOheEfLWu0WjEYjFF\nUby0zu+7nRAiFot5aZ3fd7tGo+GrdU77W+frzxG/hHjOAAAAWghCAAAQNAQhAAAIGoIQAAAEDUEI\nAACChiAEAABBQxACAICgIQgBAEDQEIQAACBoCEIAABA0BCEAAAgaghAAAAQNQQgAAIKGIAQAAEFD\nEAIAgKAhCAEAQNAQhAAAIGgIQgAAEDQEIQAACBqCEAAABA1BCAAAgoYgBAAAQUMQAgCAoEn4LgAA\noJdiWTYtLS0jI4NlWR8fH19fX5EIFw+PIfxSAQAe7sSJE9HR0QzDqNXq6Ojo/fv3NzQ08F0UdD9c\nEQIAPMS9e/eSkpKmTZsWFBRECDl58mRiYuK+ffuioqKkUinf1UF3whUhAAgde6eEyc9l62pbbiwp\nKSGE+Pv7c18OGjSI2xgXF9fzFYJO4YoQAASMZTW//Zu+dJ77Svr0MtGQJ7jXZmZmhJDMzEw/Pz+W\nZS9duuTs7Ozq6nr27Nlx48aZm5vzVjN0NwQhAAgXk5OlTUGWkIv//DAzeBJLSFBQkK+vr6OjY0xM\nTG5ubnV1dXFx8aJFi5ycnC5evJiZmTlixAh+K4duhFujACBc7J0S7euThbf+zC80YBiWZaOjo0tK\nShYuXDhs2LC7d+8aGBgsXbrUzc1NIpFIpVKKonisGbodrggBQLgom/7ci+qmposlpSFODhOWL1cz\nzCeffFJRUWFnZzd16lRuh9zc3LKysitXriiVSjc3N/5Khu6HIAQA4RJ5+ogDR9ApF8qVDYQQ/+df\nJGJx+Z07KpXqyJEjMTExDg4OCxYsqKio+Omnn1iWFYlEYWFh1tbWfBcO3QlBCAACRlGSuU+LR4+z\nUxRIYo8eyi2Us2fPnz8vk8kmTZqkVCqPHj164sSJiIiI1atXV1RU9O/fn+tEA48TBCEACB01wN56\ngH2EiXlCQsKlS5cYhlm6dCnXL7S8vDw/P58QIpPJZDIZ35WCTiAIAUCIVCrV2bNnCwoK9PX15XL5\n8OHD/fz8/Pz8EhISsrKyuBRkWba6utrCwoLvYkG30GsUAASnurp6586dKSkpMpmMYZjExMSysjLu\nWyKRqKSk5Pr16yqV6vTp0zk5ORgp8djDFSEACAs3OkKj0axcubJfv36EEKVSaWxszH33iSeeSEtL\ni46OJoRIJJJp06a5u7vzWS7oHoIQAISlurr61q1bERERXAoSQrQpePbsWQ8Pj2effbawsLC5udnZ\n2Vn7LXiM4dYoAAhLTU0NIcTBwaHVdqVSmZCQcOzYMbFYLJfLvb29kYICgSAEAGHR09MjhJSXl7fa\nbmxsHBAQoFAo1Go1H3UBbxCEACAsdnZ2pqamSUlJNE23+patrS3Lso2NjbwUBnxBEAKAsFAUFRwc\nXFRUFBcXx7KsdjvLsunp6RYWFqampjyWBz0PnWUAQHACAwMVCkVKSkptbW1YWJhMJlOpVHFxcXl5\neQsWLMCc2kKDIAQAwaEoavbs2dbW1vHx8Tk5OSYmJkqlUk9PLzw83NPTk+/qoKchCAFAiEQiUWho\nqL+/f05OTlVVlaWlpaenJ5bbFSYEIQAIl6Wl5fDhw/muAniGzjIAACBoCEIAABA0BCEAAAgaghAA\nAAQNQQgAAIKGIAQAAEF7TIZPlJSUPDhtYDupVCoDA4PuraedGIahKIqvaSx4PHGWZRmGEYvFvLSu\n0WgIIRIJP29+mqZFIpEAf+kMwxBCRCJ+Pnzz+26naZrH9xvDMFKptJ37V1ZW6rSe3ulxCEIHB4e0\ntLS0tLROHFtcXHzmzJlFixZ1e1W93549eyIjI62srPgupKclJCSIRKLRo0fzXUhPu3v37uHDh1es\nWMF3ITz44YcfJkyYYG9vz3chPe3SpUu1tbUTJkxo/yF+fn66q6d3olrOOStAp0+ffuWVV65du8Z3\nITywt7c/duyYr68v34X0tFdffVUikbz33nt8F9LTrl27NnPmzJs3b/JdCA98fX137tw5duxYvgvp\nadu2bbt58+auXbv4LqRXwzNCAAAQNAQhAAAImtBvjVZVVWVnZ48YMYLvQniQkJAwdOhQExMTvgvp\nabm5uRRFyeVyvgvpaXV1dampqWPGjOG7EB6cP3/ey8vLwsKC70J6mkKhaGxs9PLy4ruQXk3oQQgA\nAAKHW6MAACBoAgrC/Pz8cePGmZmZjR49OisrqxM79F3tPLWtW7d+9tlnPVmYrrV54vHx8UFBQUZG\nRh4eHj/88EPPV6gjbZ7477//7u3tbWRkFBAQcPbs2Z6vUHfa+W5vbGz09vY+dOhQT9amU22e+MKF\nCw1auHPnTs8X2TsJJQhZlp0xY8aYMWPy8vKmTJkybdo0bnhv+3fou9pzaikpKRs2bNi2bRsvFepI\nmydeW1sbGRm5ZMmSsrKyzz//fNWqVYmJiXxV243aPPGSkpKFCxd+8MEHd+/ejYqKioyMVKlUfFXb\nvdr/H/Jrr72WmZnZw+XpTntOPCsra/fu3dfus7a25qXU3ogVhgsXLpiZmTU1NbEsq9Fo+vXrFx8f\n36Ed+q72nNqOHTuee+45Z2fnTz/9lI8adaLNEz916pSjoyPDMNyXkZGRmzZt4qHQ7tbmiR88ePCJ\nJ57gXtM0LZVKs7OzeShUB9r5H/KxY8eGDh3q5eUVExPT4zXqRJsnzjCMubn5zZs3eSqwVxPKFWFu\nbq6Pj4+enh4hRCwW+/n55ebmdmiHvqs9p7Zq1apdu3b5+/vzUaCutHniQUFBSUlJ3GxnSqUyJSXl\n8ehK2uaJz5o16/z584QQtfr/t3e3MU1dbwDATym0lhZa6lAIDAptMkGQMBwK0wEW577wMsQKY65u\nbLg6XkyKYwPGSzTDKUR0ccQ5lCgFxFXBibwsG7DFZRBcxqYVhA0yY5wObHkd0NLz/3Bm0z9gyxjY\nSp/fp3vOPefc5yDx4dze3qOWy+UrV6709PQ0TayLbT6/7YODgxKJ5Ny5c6TZ8mB04oODg0NDQ5mZ\nmXw+PzQ0tKmpyUSRmiNLSYQPHz60s7PTFdls9sDAwL9q8PRaxlMzzOjE7ezsXF1dEUKdnZ2bN28W\nCoUikehJR7kEjE7c2tqaRqMpFAoGgyESiU6cOGGqN2EuOqNzxxjv2bMnJSVl7dq1Tzy6JWR04nfv\n3uXxeNu2bWtsbBSJRJGRkQt7LeWytEx++43icrmjo6O64vDw8Ix3bBpt8PRaxlMzbD4THxsby8rK\nksvlubm5iYmJpnoX9uKa57+4t7f32NjY999/Hx8fz+Pxnn/++ScY41IxOvdz584plcrU1NQnHtrS\nMjpxPz+/vr4+ciwQCOrq6i5fvuzr6/tEozRXlrIiFAgEt27dItsOaLVahULB5/P/VYOn1zKemmFG\nJz49PR0ZGXn//n2FQvH2228vjyyI5jHxsrIy8ngwnU4PDw/39/dva2szTayLzejcm5qa2traVq1a\n9cwzz/z666+7du165513TBTsYjI68Y6Ojq+++kpXZDAY89+SYvkz9YeUT4hWq/Xy8ioqKpqamvrs\ns888PT2np6cxxnK5XKFQGGiwDBidu05UVNRyeljG6MSvXLmyevXqnp6evkdUKpWpo14ERife2Njo\n6OjY3t4+Pj7e0NDAYrF+/vlnU0e9OIzOfXBw8M4jXl5ep06dGhgYMHXUi8DoxNva2mg0Wk1NzfDw\n8OXLl+3t7bu6ukwdtbmwlESIMe7p6dm0aROHwwkODr516xapdHFx0f3XP2eD5cHo3IlllgixsYnn\n5ubO+LuwoKDApPEuGqP/4oWFhW5ubnQ63cfHp7q62nSRLr55/rZjjP38/JbNU6N4HhM/ffq0h4fH\nihUr/P39GxsbTRep2YFXrAEAALBolvIZIQAAADAnSIQAAAAsGiRCAAAAFg0SIQAAAIsGiRAAAIBF\ng0QIAADAokEiBOYiPT2doofJZAYGBtbW1uq3wRhfuHAhNDSUw+FwOJwNGzaUl5cb3TBr//796enp\n/yW2kZGRffv2eXh4MJlMf3//qqoq/bNffvnlxo0bORyOUCjs6OiYc4TGxkZfX18nJyepVDo1NaWr\nz87OzsjI+C+xGTA6OkqhUHbv3r1E4wOwPEAiBGaEwWCceSQ/P398fDw6Ovrbb7/VNUhNTRWJRHQ6\nPSsrKz8/39HRcdeuXTk5OQbG7O3tPX369H8MTCKRlJWVSSSS0tJSPp8fHx9fX19PTtXV1e3cuZPP\n5xcWFqrV6vDw8N7e3hndf//995iYmKSkpPLy8itXrnzyySek/v79+2VlZUuXCKlUqlAo9Pb2XqLx\nAVgmTP2NfgD+IZVK2Wy2fo1SqeRyudHR0aTY3NyMECooKNDtIEh6IYTm3E6vtbV106ZNZF8FqVS6\n4MCUSiVCqKSkhBQ1Go1AIEhISCDFsLCwLVu2kJCGh4ednZ0zMjJmjHD06NGoqChyLJfLfX19yXFK\nSsrBgwcXHBgAYFHAihCYLw6H4+/vr9tWLTc318fH5/3339d/O/YHH3zg4eHR0tIyuzuXy42Kiioo\nKOByuQauQl63rVKpHtfgwYMHISEhoaGhpEilUt3c3MiW7oODg83NzXFxcSQkOzu7iIgIuVw+Y4Tp\n6WkWi0WO7e3t1Wo1Qqi/v//SpUtpaWmPu25MTIy/v7+uqNVqXVxckpKSEELDw8MSicTFxYVOp3t6\neubn5+vuD/N4vMrKyry8PC6X29vby+PxCgsLySnDvcrLywsLC729vdlstkgkevjwITml0WhycnKe\ne+45Ozu7kJCQ9vZ2Uo8xLi0tDQgIYDKZ69atq6ioMPBDBsCsmToTA/CP2StCjUbj6ekZFhaGMZ6c\nnLSysjp06NACRubz+QZWhIWFhUKhcGRkxOg4arX6wYMHlZWVK1asqKysxBh3dnYihK5du6ZrU1xc\nTKfT9desGOPu7m4Oh9Pa2vrXX3+FhoZmZ2djjMVi8bFjxwxcTiaTIYT6+/tJ8YcffkAItba2YowT\nExMdHBxycnJkMplEIkEIyWQy0szd3T08PJzH4+3du3dwcNDd3f3IkSPklOFeL774YmRkZF1d3aef\nfkqj0fbu3avrxWKxDh8+XFVVtXnzZgcHh3v37mGMi4qKEELJyclVVVVvvPEGQujs2bNGf4YAmCFI\nhMBcSKVSOzu7e4/cuHGDrH6OHz+OMe7u7kYILez10IYT4fwVFBSQPx/fe+89kurINt/6m3icPXsW\nITQ0NDSjb21tLZ/Pt7e3f/fdd//++++bN296eHhMTEwYuNzQ0BCNRiPTxxhLpdJnn32WbCkQHR19\n6tQpXUtfX9/9+/eTY3d3d3d399HRUV1RlwgN91qzZo1GoyFFsVgcGBiIMe7p6bGysqqoqCD1KpWK\nwWCUlpaOjIyw2Wxdd4zx66+/zufzjf4MATBDlrIxL3gqjIyMODs769e8+eabZO1CbkXS6XTTRIYQ\nQkgsFgcHB1+7di0/P5/FYh06dIjU69+qxRgjhMjNT32RkZGRkZEYY9I4Ozv7o48+otPpJSUlR44c\n0Wq1aWlp+/bt0x/K3t5+27ZttbW1KSkpGGO5XJ6QkGBlZYUQunTpEmlz79695ubm27dvv/zyy7qO\n27dvZzKZs+M33CsiIoJKpZJjJyenGzduIITa2towxtu3byf1bDa7p6fH1tZWoVAMDQ2RhSCxc+fO\niooKtVoNu9yBpw4kQmBGbG1tdR+wUalULy8vV1dXUiS7jP7222+ze5WVlVEoFLFYvNThOTs7Ozs7\nv/TSS0NDQ0ePHj1w4MCqVasQQuRpGkKlUtFotMd9KknyXHt7e1dXV3V1dX19fVZW1smTJ62trZOS\nkng83quvvqrfPjY2NjExUaVS9fX19ff3JyQkkPpffvnlww8//Omnn6anpwMDAzkcjn4vEtVshns5\nOjrO7nLnzh0ul0uj0XQ1Li4upB4hFBAQoMvcWq1Wq9XevXuXx+PNeXUAzBYkQmBGbGxsXnnllTlP\nMZlMLy+v6urqGcumycnJ5OTk3bt3L10irKqqOnjwYGdnp27BJBAIpqamJicnSZ7u7u4OCgoip27f\nvu3q6mp4s/vMzMwDBw5YW1tXVVWlpaXt2LEDIdTV1XXhwoUZiTAiIoJCoVy9elWhUPj5+fn4+CCE\nhoeHg4KC4uLiGhoa1q1bR6FQAgMD9Xvp4tRntNecMTs5OalUKo1GQx6+RQj19fXRaDQnJyeEUFNT\n0+rVq2e0NzBxAMwTPDUKnhp5eXk//vgj2WVUV3n48OGxsbGoqKilu66bm9vNmze/++47XU1LSwuP\nx2OxWCtXrgwNDdWtYicnJ+vq6nQ3Euf0zTffqFSqmJgY9OhRNVJPoVA0Gs2Mxg4ODuHh4bW1teS+\nKKlsa2sbHx/Py8vz8/OjUCgDAwMKhcLoLBbWKyAgYHp6uqamhhQnJiY2bNggk8m8vb0ZDEZvb++a\nR65fv15QUGDae9cALAysCMFTY8eOHdXV1VKptKGhQSgU2tratrS0XLx4USwWb926dcHDFhUV1dfX\n19TU6L7hMMPGjRuDgoJee+21jIwMJyenr7/+WiaTnTlzhpxNT0+PiIiQSqVbtmz54osvlErlnj17\nHnctjHFmZubHH39Mll+xsbFisXjt2rU2NjZFRUUlJSWzu8TGxkokErVaHR8fT2oEAgGVSk1OTo6N\njVUqlSdOnLCxsWlpaens7PTz83vcpRfWy9fXVyQSvfXWW/39/Twe78yZM5OTk3FxcQ4ODunp6amp\nqX/88YePj8/169ePHTuWnZ1teCkMgJky1VM6AMww++sTs2k0mpMnTwYGBjKZTDab/cILL3z++ee6\nZx0fx/BTo4mJiQghpVJpYIQ///xTLBa7uLgwmcz169efP39e/wsS58+fX79+vb29fVhYWEdHh4Fx\nLl68GBISot/3+PHjrq6uzs7O5JGZ2V0GBgaoVCr5DomOTCYTCARMJjM4OPjq1auNjY1kBPz/j4nO\nKM6/V0ZGRkBAADmemJiQSqWenp4sFiskJIQ8PoMx1mq1xcXFPj4+DAZjzZo1xcXFc8YPgPmjYL27\nTAAAAIClgc8IAQAAWDRIhAAAACwaJEIAAAAWDRIhAAAAiwaJEAAAgEWDRAgAAMCiQSIEAABg0SAR\nAgAAsGiQCAEAAFg0SIQAAAAs2v8ATn0nNz2HaSAAAAAASUVORK5CYII=\n"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -w 600 -h 600\n",
    "FinalPCAData <- cdBactAnnot\n",
    "# # Used this function to save the file in pdf format\n",
    "#FinalPCAData <- na.omit(FinalPCAData)\n",
    "#print(dim(FinalPCAData))\n",
    "strain <- factor(c(rep(\"Cells\", 96)), levels=c(\"Cells\"), labels=c(\"Cells\"))\n",
    "PCx=1\n",
    "PCy=2\n",
    "pcaPRComp <- prcomp((FinalPCAData))\n",
    "percentVar <- pcaPRComp$sdev^2/sum(pcaPRComp$sdev^2)\n",
    "df<-data.frame(PCX=pcaPRComp$rotation[,PCx],PCY=pcaPRComp$rotation[,PCy],Cell=strain)\n",
    "p<-ggplot(df, aes_string(x=\"PCX\",y =\"PCY\",color=\"Cell\"))\n",
    "p<-p+ggtitle(\"PCA with Raw Filtered Dataset\")\n",
    "p<-p+geom_point(size = 2)\n",
    "p<-p+xlab(paste(\"PC\",PCx,\": \", round(percentVar[PCx] * 100), \"% variance\"))\n",
    "p<-p    + ylab(paste(\"PC\",PCy,\": \", round(percentVar[PCy] * 100), \"% variance\"))\n",
    "p<-p + geom_text(data=df,aes(label=colnames(cdBactAnnot)),alpha=0.5,size=5, vjust=1,hjust=0.5,angle=45, color=\"black\")\n",
    "print(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "              C01 C02 C03 C04 C05 C06 C07 C08 C09 C10 C11 C12 C13 C14 C15 C16\n",
       "Mx1             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Pdgfb           0   0   0   4   0   0   0   0   0   0   0   0   0   0  24   0\n",
       "Mmp14           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ccl3            0   0   0   1   5  36   0  42  22   7  60  64   1   0  13   0\n",
       "Ckb             0   0   0   0   0   1   0   1   0   1   0   0   0   0  38   0\n",
       "Relb            0   0   1   0   0   0   0   0  11  91   0   0   7   0   0   0\n",
       "Ier3            0   0   0   0   0   0   0   0   0   0   0   0   3   0   0   0\n",
       "Il12b           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Adamts4         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Sod2            2   0   0   0   0   1   0   0   0 204   0   0   0   0   0   0\n",
       "Hap1            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Cav1            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Pou2f2          0   1   0   0   0   0   0   0   0   0   0   0   3   0   1   0\n",
       "St3gal1         1   0   0  55   0   0   2  38   0  84   0   1   1  12   5   0\n",
       "Mertk           0   0   0   0   0   0   0   0   0   0   0   0   0   2   0   0\n",
       "Csf1            1   0   1   0   0   0   0   5   0   0   0   0   1   0   5   0\n",
       "Gadd45b         0   0   0   0   1   0   0   0   1  18   0   0   0   0   0   0\n",
       "Cd83            1   1   0   0   0  64  13 130   0   7   0  43   0   0 123   0\n",
       "Sqstm1          0   4  22   0   0   0   3   8   0  28  32   8   0   0   1   0\n",
       "Cd274           0  32   0  13   0   0   0   0   0   0   1   1   0   0   0   2\n",
       "Il10            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Slpi            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Mafk            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Slc13a3         0   0   0   0   0   0   0   0   0   0   3   0   0   0   0   0\n",
       "Adora2b         0   0   0   0   0   0   0   0   0   0   0   0   0   0  12   0\n",
       "Irf1            0   0   0   0   0  96   0   0   0   0   0  22   0   0   0   2\n",
       "Csf2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Cxcl16          0   0  75   2  11   3   0   8   0   0   0   0   0   8 550   0\n",
       "Ccl4            0  32   0   0   0  14   0  46   0 357   0  16   0   5 514 125\n",
       "Scn1b           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Tnfaip3         1   0   0  59   0   0  38   0   0   8   0   3   0   0   0   8\n",
       "Dusp6           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Peli1           6   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Rel             0  11  23   0   0  86   5   6 199 104  17  26   0 187  12   0\n",
       "Tnip1           0   0   0   0   0   0   0 103   2   0   0   1  23   2  41   5\n",
       "Rsad2           0   0   0   0   0   0   1  60   0   0   0   0   0   0   0   0\n",
       "Nfkbia          0  16   4   2   0   0  33   7  13   0   5   7  15   0 164   0\n",
       "Arg2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Tnfaip2        14   2   0  82   1   3   0  45   0   0   0   3   0 484  11   0\n",
       "Nedd9           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Edn1            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Arhgef3         0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0\n",
       "Wnt5a           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ednrb           0   0   0   0   0   0   0   0  16   0   0  53   0   0   0   0\n",
       "Irg1            7  29   0   2   0  90   9  55 100   7   0  18   0   0 842   0\n",
       "Zfp263          0   0   0   0   0   0   8 144   0   0   4   0   1   0 107   0\n",
       "Ets2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   7   0\n",
       "Denr            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Il15ra          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Mx2             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Nfkbie          0   9   0   0   0   0   0   0   0   2   0  12   0   0   1   0\n",
       "Enpp4           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Pim1            9   0   0  23   0   0   0   1   0   0   0   1   0   3  16   0\n",
       "C3              0  37   0   0   1   0   0   0   0   0   0   0   3   0   0   0\n",
       "Map3k8          0   0   0   0   0   0   5   0   0   0   0   0   0   0   0   0\n",
       "Tnf            30   5   8   2  58   7  46  12  13 101 186 308   4 126 235   0\n",
       "Lox             0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ehd1            1   2  18   0   0  78   0   0  20   2   0   0   0   0   0   0\n",
       "Fas             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Jak2            0   0   0   0   0   0   0   0   0   0   0   0   0  26   0   0\n",
       "Nfkb2           0   0   0   0   0   0   0   0   0   0   1   1   0   0   1   0\n",
       "Il6             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Slco3a1         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Cflar           0   2   0   1  66   1   0  64  57  38  58  13 117   0  12   0\n",
       "Rgs2            0   0   0   0   0   1   0   0  27   0  37   0   0   0   0   0\n",
       "Marco           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Fcgr2b          0   0  43  28   0   0   0   0   0   1   0   0   0   0   0   0\n",
       "Lcn2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Traf1           0   3   0   6   0   8   0   0  13   0   0  69   0  36   0   0\n",
       "Gca             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifih1           0   0   0   0   0   0   0   0   0   0   0   2   0   0   0   0\n",
       "Itgb6           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Il1rn           0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0\n",
       "Dusp2           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Il1b            0   0   0   0   0   0   0  98   0   0   0   0   0 229   2   0\n",
       "Il1a            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Car2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Src             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Slc7a11         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Vcam1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Tlr2            0   0   0   0   5   1   0   0   0   0   0   0   0   0 272   0\n",
       "Gem             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Gbp2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ppap2b          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Pde4b           0   0   0   0   0   9   0   0   0   0   0   0   0   0  59   0\n",
       "Tnfrsf1b        1  45   0   9   0  18  81  65  21   1   1  46   4   3 109   0\n",
       "Hivep3          4   0   0   0   0   1   0   2  56 152   0  30  92   0   1   4\n",
       "Rasgef1b        2   0   3  26   1   2   1   4   7   1   7   2  32   5  55   0\n",
       "Cxcl1           8   0   0  18   0   0   0   0   7   1   0   0   0   0   0   0\n",
       "Cxcl9           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Oas1b           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Clec4e          8   0   0   0   0   0   0   0  44   4  27 364   8   0 321   0\n",
       "Cd69            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Clec2d          0   0   0   0   0   0   0   0   9   0   0   2  14   0   1   0\n",
       "Olr1            0   0  28   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Dusp16          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Vasp            0   1   0   0   0   0 188   0   0   0   0   0   0   0   0   0\n",
       "Nfkbib          0   0   0   0   0  34   4   0   0   0   0   0   0   0   0   0\n",
       "Pim2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Nox1            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Rab11fip1       0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   6\n",
       "Vegfc           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Slc7a2          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Herpud1         0   0  55   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Cx3cl1          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Phxr4           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Birc3          31   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0\n",
       "Pde4a           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ptgs2           0   0   0   0   0   0   0   1   0   0   0   0   1   0   0   0\n",
       "Cish            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Malt1           0   0   0   1 151 194   0   0  55   0   0  24   0   0  49  18\n",
       "Alas1           0   0   0   0   0   0   2   0   8   0   0   0   0   0   0   0\n",
       "Batf            0   0   0   0   0   0   0   0   0   0   0   0   0   0  11   0\n",
       "Ifit1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Cxcl10          0   0   0   0   0   0   0   0   0   0   0   0   0   8   0   0\n",
       "Igsf6           5   0   0   1   0   0   0   2   0  20   0   6   0   0  41   0\n",
       "Ccl5           34   0   0   0   0   0   0   0   9   0   0   0   9   2   0   0\n",
       "Tpbg            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Nfkbiz          0  22   0   2   4  26   0   0   0  22   0   3   0   0 137   1\n",
       "Ccl7            0   0   0   0   0   0   0   0   0   0   0   0   0   0   4  10\n",
       "Sh3bp4          0   0   0   0   0   0   0   0   0   4   0   0   0   0   0   1\n",
       "Igfbp7          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Gadd45a         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ppfibp2         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Spry1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Chst7           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Icam1           0   2  16   0   0   0   0  12   0   1   0  22   0   6   1   9\n",
       "Serpine1        0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ppm1k           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Egr2            0  12   0   0   0   0   2  31  21   0   0   0   0   0   2   0\n",
       "Snn             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Socs1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Prdm1           0   0   0   0   0   0   0   0   0   0   0   0   0   0  51   0\n",
       "Slamf7          0   0   0   1   0 121   0   6   0   0   0   5   0   5  16  17\n",
       "Gcnt1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Itpkb           1   2   2   0   0   1   0   0   7   1   0   1   5   0   2   0\n",
       "Stx11           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Tnfsf10         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Fnbp1l          0   0   0   0   0   0  13   4   1   0   0   7   0  59 285   0\n",
       "Fgl2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Saa3            0   0   9   3   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Gpr34           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Bcor            0   0   0   0   0   0   0   2   4   0   0   0   0   0   1   0\n",
       "Ripk2           0   0   0   1   0   0   0   9   0   0   0   0   0  94   0   0\n",
       "Oasl1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Rassf4          0   0  22   0   0   0   0  57   4  21   0   8  10  22  70   0\n",
       "Ikbke           0  26   0  82   8   6   0  45   0   0   3   1   0   0   6   5\n",
       "Maff            0   0   0   0   0   0   0   0   0   0   0   0   0   0  15   0\n",
       "Ccrl2           0   0   0   1   0   0   0   0   0  20   2  85   0   0   0   0\n",
       "C130050O18Rik   0   0   0  27   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Fzd1            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Il27            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Tlr1            0   0   0   0   0   3   0   0   0   0   0   0  11   0   0   0\n",
       "Cdc42ep2        0   0   0   0   0   0   0   0   0   0   0   0   0   0  13   0\n",
       "Adrb2           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifit2           0   0   0   0   0   0   0   0   0   0   3   0   0   1   0   0\n",
       "Gpr85           1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Phldb1          0   0   0   3   8   0   0   0   0   0   1   0   0   0   0   0\n",
       "P2ry6           0   0   0   0   0   1   3   0   1   0   0   0  44   0  33   0\n",
       "Ifnb1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Clec4a1         0   0 117   0   0   0   3   0   0   0   0 282   0   0   0   0\n",
       "Gpr18           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Mmp13           0   0   0   0   0   0   0   0   0   0   2   0   0   0   0   3\n",
       "Flrt3           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Plagl2          0  81   1   0   0   2   0   2   0   0   0   0   0   0   0  25\n",
       "Tagap1          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Junb            0   1   0   0   0   0   0   0   4   0   0   0  19   0   0   0\n",
       "Socs3           0   0   2   0   0   0   0 139   1   1   0   0   0   0   7  86\n",
       "Bcl3            0   0   0   0   0   0   5   0   0   0   0   0   0   0  12   0\n",
       "Slamf8          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Bcl2a1c         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Lipg            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Iigp1           0   0   0   0   0   0   0   0   0   0   1   0   0   0   0   0\n",
       "Ifi205          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Cebpb           0   0   0   0   0   0   0   2   0   0   0   0   0   0   0   0\n",
       "Cxcl2           2   0 160  18   1   1   0  49   5 328   1 141  29   0 165   2\n",
       "Cxcl11          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Gpr84           0   0   0   0   0   0   0   3   0   0   0   2   3   0   0   0\n",
       "Slc31a2         0  56   0   0   0   0   1   0   0   0   0   0   0   0   0   0\n",
       "2010002M12Rik   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Marcks          1   6  15   0  76   0   0   0   2  19   1   0   3  48   9   2\n",
       "1700086L19Rik   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Zhx2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "4930599N23Rik   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Olfr661         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Il4i1           0   0   0   0   0   0   0   0   0   0   0   0   0   8  40   0\n",
       "Mafb            0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0\n",
       "Thbd            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Arrdc3          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifit3           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Sox4            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifna2           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Igtp            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifi47           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Gbp4            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "E130102H24Rik   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Bcl2a1b         0   0   5   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Pou3f1          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Hspa1b          0   0   0   0   0   0   0   5   0   0   0   0   0   0  66   0\n",
       "Ifna1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifna14          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifna5           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifnz            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Gdap10          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "E230013L22Rik   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Bcl2a1d         0   0  21   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifnab           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Bcl2a1a         0   0   2   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Gbp5            0   0   0   0   0   0   0   0   0   0   0   0   0  41   0   0\n",
       "              C17 C18 C19 C20 C21 C22 C23 C24 C25  C26 C27 C28 C29 C30 C31 C32\n",
       "Mx1             0   0   0   0   0   0   0   0   0    0   0   0   4   0   0   0\n",
       "Pdgfb           0   0   0   0   0   6   0   0   0    0   0   0   0   0   0   0\n",
       "Mmp14           0   1   0   0   0   0   0   0   0    0   3   0   0   0   0   0\n",
       "Ccl3            2 159   3 256   5  61 103   2  21   44   6  61 183  15   0  93\n",
       "Ckb             0   1   2   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Relb            0   0   0   0   0   0   0   0   2    0   0   0   0   0   0   0\n",
       "Ier3            0   0   0   6   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Il12b           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Adamts4         0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Sod2            0   0   0   0   0   0   0   0   0    0   3   1   0   0   0  14\n",
       "Hap1            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Cav1            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Pou2f2          0   0   0   0   1   0   0   0   0    0   1   0   0   0   0   0\n",
       "St3gal1         0   5   0   0   0  15   0   0   0    0   8   3   0   0   0   5\n",
       "Mertk           0   1   0   0   0   0   0   0   0    0   0  18   0   0   0   0\n",
       "Csf1            0   0   0  32   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Gadd45b         0   0   0   0   3   0   0   0  51    7   0   0  55   0   0   0\n",
       "Cd83            0   5   2  60  39   0  68   0  13   10  11  16  48   0   7   0\n",
       "Sqstm1          0   0   0   3   4   0  28   0   0   14  12   0   1   0   0   3\n",
       "Cd274           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Il10            0   0   0   0   0   0   0   0   0   13   0   0   0   0   0   0\n",
       "Slpi            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0  10\n",
       "Mafk            0   0   0  70   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Slc13a3         0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Adora2b         0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Irf1            0  34   0   0   0   7   0   0   4    0   0  20   5   8   0   0\n",
       "Csf2            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Cxcl16          1   0   4  34 252   0  37   1   0 1521   0   8   0  49   1  23\n",
       "Ccl4            0 362  43 238  49  91   0  10   0   53 167  60   0  17   1  10\n",
       "Scn1b           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Tnfaip3         0   4  77   2   0  34   0   0   0    7 107  12   0   8  22   0\n",
       "Dusp6           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Peli1           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Rel             0  23   0   0   0   9  58   1   0    0  19 156  19   3  57   4\n",
       "Tnip1           0   0   0   9   0   0   1   1   3    0   0   1   0   4   0 141\n",
       "Rsad2           0 215   0 137   0 120   0   0   0    0   0   0   0   0   0   0\n",
       "Nfkbia          6   2   5   7   4 185   0   0   0    0  14   0   7   5   0   4\n",
       "Arg2            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Tnfaip2        13   1   0   0   1  53   5   0  20   25  12   9  45   2   0 166\n",
       "Nedd9           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Edn1            0   0   0 170   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Arhgef3         0   0   0   0   0   0   0   0   0    0  56   0   0   0   0   0\n",
       "Wnt5a           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Ednrb           0   0   0   0   0   0   0   0   0    0   0   0   0   0   1   0\n",
       "Irg1            1  48  14  85   2  93   1   0   0    0   3 199  37   0   0 346\n",
       "Zfp263          0   8   0   1  46   0   0   1   0    0   8   0   1   0   0   0\n",
       "Ets2            0   0   1   5   0   0  10   0   0    0   0   3   0   0   0   0\n",
       "Denr            0   0   0   1   0   0   0   0   0    0   0   0   0   0   0  63\n",
       "Il15ra          0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Mx2             0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Nfkbie          0   0   0   6   0   0   0   0   0    0   0   1   0   0   0   0\n",
       "Enpp4           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Pim1            0   0   0  14   0   0  94   2   0    2  11   1   0   0   0   0\n",
       "C3              0   5   0   0   0   0   0   0   0    0   0   0  16   0   0   0\n",
       "Map3k8          0   1   0   0   0   0   0   0   0    0   1   6   0   0   0   0\n",
       "Tnf             2   8   4 494  39  88  69   1 161    7 196 139  36  73   0  10\n",
       "Lox             0   0   0   0   0   0   0   0   0    2   0   0   0   0   0   0\n",
       "Ehd1            0   0   0   2   0   0   1   0   0    0  25  80 109   0   0  11\n",
       "Fas             0   0   0   0   0   0   0   0   0    0   2   0   0   0   0   0\n",
       "Jak2            0   2   0   0   0   1   0   0   0    0   8 132   0   0   0   0\n",
       "Nfkb2           0   0   0   0  10   0 169   0   0    0   0 142   3   0   0  50\n",
       "Il6             0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Slco3a1         0   0   0   0   0  22   0   0   0    0   0   0   0   0   0   0\n",
       "Cflar           4   0   1  63   1   0   0   5   0    0  27   6   1   0   0   0\n",
       "Rgs2            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Marco           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Fcgr2b         14   2   0   0   0   0  62   0   0    0   9  10   0   0   0   0\n",
       "Lcn2            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Traf1           0   0   0  15   0   0   0   0   5    0   0   0   6  80   0   7\n",
       "Gca             0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Ifih1           0   0   0  11   0   0  44   0   0    0   0   0   0   0   0 108\n",
       "Itgb6           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Il1rn           0   2   0  30   0 152  63   0   0   10   0  58   0   0   0   0\n",
       "Dusp2           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Il1b          100   0   0   0   0   5   0   0   0    0   0  75   7   0   0 221\n",
       "Il1a            0   0   0   0   0  13   0   0   0    0 344   0   0   0   0   0\n",
       "Car2            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Src             0   0  41   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Slc7a11         0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Vcam1           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Tlr2            0   0   0   0   0   0   0   0   0    0  42 135   0   2   0   3\n",
       "Gem             0   0   0   6   1   2   0   0   0    0   5   1   0   0   0   0\n",
       "Gbp2            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Ppap2b          0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Pde4b           0 129   1 102   3   7  14   0   0    0   1  80 144   0   1 112\n",
       "Tnfrsf1b        0 281  36   0   1   0 387  15   0   11  17  42   2   0  12   0\n",
       "Hivep3          0   0  12   7  57   0 127   0 138    6   8  23   2   1   0  12\n",
       "Rasgef1b       30 241   6  61  10  12   9   1   1   19  22  21   1   4   8   0\n",
       "Cxcl1           0   0   0   6   2  16   0   0   0   24   0   0  71   0   0   0\n",
       "Cxcl9           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Oas1b           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Clec4e          0   0   0  90   0  13   0  28   1   16  55  71   5   0   1 205\n",
       "Cd69            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Clec2d          0   0   0   7   0   0   0   0  39   46   0   0   0   0   0   0\n",
       "Olr1            0   0   0   0   0   0   0   0   0    2   0   0   0   0   0   0\n",
       "Dusp16          0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Vasp            0   0   0   0   0   0   0   0   0    0   1   0   0   0   0   0\n",
       "Nfkbib          0   0   0   0   0   0   0   1   0    0  39   0   0   0   0   0\n",
       "Pim2            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Nox1            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Rab11fip1       0  23   1   0   0   1   0   0   0    0   0   0   0   0   0   0\n",
       "Vegfc           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Slc7a2          0   0   0   0   0   0   6   0   0    0   0   0   0   0   0   0\n",
       "Herpud1         0   0   0   0   0   0   0   0   0    0   1   1  54   0   0  64\n",
       "Cx3cl1          0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Phxr4           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Birc3           0   0   0   0   0   0   0   0   1    3   0   0   3   2   0   0\n",
       "Pde4a           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Ptgs2           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0  64\n",
       "Cish            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Malt1           0   0   0   0   2   0  14   0  31    0   0   0   0   0   0  21\n",
       "Alas1           0   0   0   0   0   0   0   0   0    0   1   0   0   0   0   1\n",
       "Batf            0   0   0   0   0   0   0   0   0    0  50   0   0   0   0   0\n",
       "Ifit1           0   2   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Cxcl10          0  10   0   0   0   0   0   0   0    0   0  25   0   0   0   0\n",
       "Igsf6           0   0   3  51   0   0   0   0   0    0   3   0   0   0   0   0\n",
       "Ccl5            0   0   0  31   6   0   0   0  15    0   0   1   0   0   0 107\n",
       "Tpbg            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Nfkbiz          0   4   0   0   0 226  59   1   4    0  81   0   1  54   0 242\n",
       "Ccl7            0   0   0 118   0   0   3   0   0    0   0   0   0   0   1   0\n",
       "Sh3bp4          0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Igfbp7          0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Gadd45a         0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Ppfibp2         0   0   0   0 237   0   0   0   0    0   0   0   0   0   0   0\n",
       "Spry1           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   9\n",
       "Chst7           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Icam1           0   1   0  12   0   3 125   0   0    3  17   1  34   2   0   0\n",
       "Serpine1        0   0   0   0   0   0   0   0   0    0  13   0   0   0   0   0\n",
       "Ppm1k           0   0   0   0   0   0   0   0   0    0   0  17   0   0   0   0\n",
       "Egr2            0   1   0   0   0   2  14   0  10   13  44   0   0   0   0 162\n",
       "Snn             0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Socs1           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Prdm1           0   0   0   0   0   0   0   0   0    4   3   8   0   0   0   0\n",
       "Slamf7          1   1   0  17   4   2   0   0   1    1   2   1   0   2   0 171\n",
       "Gcnt1           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Itpkb           0  57   0   0  16   1   6   0  11    1   0   2  32   0   2   1\n",
       "Stx11           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Tnfsf10         0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Fnbp1l          0   0   0   0   0   0   9   0   0   12   0   0   0   0   0   0\n",
       "Fgl2            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Saa3            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Gpr34           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Bcor            0   0   0   0   0   0   2   0  26    1   0   2   0   0   0   7\n",
       "Ripk2           0   5   0  55   0   0   1   0   0    0   0  10   0   0   0   0\n",
       "Oasl1           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Rassf4         40  68   4   1   0   0  19   0  72    1  36   5 107   0   2  26\n",
       "Ikbke           0  20   0   2   3  25   0   0   0    0   2  11   0   0   0   0\n",
       "Maff            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Ccrl2           0   0   0  53   0   0   0   0   0    7  12  92   0   0   0   0\n",
       "C130050O18Rik   0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Fzd1            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Il27            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Tlr1            0   0   0   0   0   0   0   0   1    4 302   6   0   0   0  10\n",
       "Cdc42ep2        0   0   0   1   0   0   0   0   0    1   0   1   0   0   0   0\n",
       "Adrb2           0   0   0   0   0   1   0   0   0    0   0   0   0   0   0   0\n",
       "Ifit2           0   8   0   0   0   0   0   0   2    0   0   0   0   0   0   0\n",
       "Gpr85           0   0   0   0   0   0  14   0   0    0   0   0   0   0   0   0\n",
       "Phldb1          0   0   2   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "P2ry6           0   0   0   4   0   1   0   0   0    0   0   0   0   0   0   0\n",
       "Ifnb1           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Clec4a1         0   5   0   0  32   0   0   0   0    0   0 205   1   0   0 114\n",
       "Gpr18           0   0   0   0   0   0   0   0   0    0   0   0   0 247   0   0\n",
       "Mmp13           0   0   0   0  10   0   0   0   0   12   0   0   0   1   8   0\n",
       "Flrt3           0   0   0   0   1   0   0   0   0    0   0   0   0   0   0   0\n",
       "Plagl2          0  23   0   6   1   0   0   0   0    0   0  25   0   0   5   1\n",
       "Tagap1          0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Junb            0   3   0  18   0  17   1   1   1    0   0   0   0  32   0   0\n",
       "Socs3           0   0   0   1   1   0   1   0   2   11   0   0   0   0   0   0\n",
       "Bcl3            0   0   0   0   0   0   0   0   0    0   0   0   0  30   1   0\n",
       "Slamf8          0   0   0   0   0  12   0   0   0    0   0   0   0   4   0   0\n",
       "Bcl2a1c         0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Lipg            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Iigp1           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Ifi205          0   0   0   0   0   0   0   0   4    0   0   0   0   0   0   0\n",
       "Cebpb           0   1   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Cxcl2           1  39  11 190  55 170   1   9  44    5 140 175 305   1   7 110\n",
       "Cxcl11          0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Gpr84           0   0   0  54   0   0  72   0   0    0   0  57   0   0   0   0\n",
       "Slc31a2         0   0   0  93   0   0   0   0   0    0   0   0  11   0   0   0\n",
       "2010002M12Rik   0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Marcks          0 133  10  40   3   0  50  25  51    0   6  10   2  23   0  12\n",
       "1700086L19Rik   0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Zhx2            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "4930599N23Rik   0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Olfr661         0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Il4i1           0   1   0   0   0   0   3   0   0    0   0   0   0   0   0   0\n",
       "Mafb            0  40   0   1   0   0   0   0   0    0   2   0   0   0   0   0\n",
       "Thbd            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Arrdc3          0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Ifit3           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Sox4            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Ifna2           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Igtp            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Ifi47           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Gbp4            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "E130102H24Rik   0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Bcl2a1b         0   0   0   0   0   0   0   0   0    0   0   0   0   0   6   0\n",
       "Pou3f1          0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Hspa1b          0   0   0   0   0   0   0   0   0    0   0   1   0   0   0   0\n",
       "Ifna1           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Ifna14          0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Ifna5           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Ifnz            0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Gdap10          0   0   0   0   0   0   0   0   0    0   1   0   0   0   8   0\n",
       "E230013L22Rik   0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Bcl2a1d         0   0   0   0   0   0   0   0   1    0   0   0   0   0   0   0\n",
       "Ifnab           0   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0\n",
       "Bcl2a1a         0   0   0   0   0   0   0   0   0    0   0   0   0   0   6   0\n",
       "Gbp5            0   0   1   0   0   0   0   0   0    0   0   8   0 195   0   0\n",
       "              C33 C34 C35 C36 C37 C38 C39 C40  C41 C42 C43 C44 C45 C46 C47 C48\n",
       "Mx1             0   0   1   0   0   1   0   0    0   8   0   0  50   0   0   0\n",
       "Pdgfb           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Mmp14           0   0   0   0   0   0   0   0   11   1   0   0  24   0   0   0\n",
       "Ccl3            9   1  97   3   2 289 923 244  100   4   7   0 560   0  67   9\n",
       "Ckb             0   0   0   0   0 168   0   0    2   0   0   0   0   0   0   0\n",
       "Relb            0   3   1  10   0  88   0   0    2  54   0   0   0   0   0   0\n",
       "Ier3            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Il12b           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Adamts4         0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Sod2            0   0   0   0  29   0   0   0    0   0   0   0  11   5   0   1\n",
       "Hap1            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Cav1            0   0   0   0   0   0   0   0    0   0   1   0   0   0   0   0\n",
       "Pou2f2         96   0   0   0  31   0   0   0    0   2   0   0   0   0   0   0\n",
       "St3gal1         1  34  19   0   0   1   0   0    4   1   0   0   0  25   1   0\n",
       "Mertk           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Csf1            0   0 208   0   0   0   0   0    0   0  40   0   0   0   5   1\n",
       "Gadd45b         5   0   0   0   0   4   0   0    0   0   0   0   0   0   5   0\n",
       "Cd83            0   0   0   1   0  20 111  35    1  65   2  14 263  87   0  57\n",
       "Sqstm1         59   0 340  89  83 244  10   2    0   0   2 111   0   8   1   2\n",
       "Cd274           0   0   1   0   0   0   0   0    0   4   0   0   0   0   0   1\n",
       "Il10            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Slpi            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Mafk            0   0   0   0   0   0   0   0    0   0   1   0   0   0   0   2\n",
       "Slc13a3         0   0   0   0   7   0   0   0    2   0   0   0   0   0  45   0\n",
       "Adora2b         0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Irf1            0   0   0   0   1   2   2   0  159   0   0   0   0  47   0   1\n",
       "Csf2            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Cxcl16          1   1   1   0   7   0   0   1   39  28  13   4  17   0  26  24\n",
       "Ccl4            0   0  33   0  24   0   0  47   94   2  52   0  40  73 268   0\n",
       "Scn1b           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Tnfaip3         0   0   0 132 732   1   1  36    5   4  84   0   4  16   0   7\n",
       "Dusp6           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Peli1           0   0   0   0   0   0   0  17    1   0   0   0   0   0   0   0\n",
       "Rel             2   0   0   2   4  56  15   0    0  28  35  23  15  37  14   0\n",
       "Tnip1           0   0  15   0  40  72   5   0    0   0   2   0   0   2   0   0\n",
       "Rsad2           0   0   0   0 135   0   0 148   31   0   0   0 526  12   0   5\n",
       "Nfkbia         14   1   3   2  94  47 115 315    0  49   6   1 147   4   1 342\n",
       "Arg2            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Tnfaip2        58   1   0 516   6   9   9  11    0   3 188   0 154  13  95  10\n",
       "Nedd9           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Edn1            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Arhgef3         0   0   0   0   2   0   0   0    0   0   0   0   0   0   0   0\n",
       "Wnt5a           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Ednrb           0   0   0   0   0   1   0   0    6   0   0   0   0   7   0   0\n",
       "Irg1            2   0   3 225  44   1 133   0    0  68   2   0   0 119   0   1\n",
       "Zfp263          0   0   0   0   5  52   1   0  377   0  26   0   0   2   2   0\n",
       "Ets2            0   0   0   0 404   0   0   0    0  42   0   0   0   0  33   0\n",
       "Denr            0   0   0   0   0   0 116   0    0   4   0   0   0   0   0   0\n",
       "Il15ra          0   0   0   0   3   0   0   0    0   0   0   0   0   0   0   0\n",
       "Mx2             0   0   3   0   0   0   0   0    0   0  13   0   0   0   0   0\n",
       "Nfkbie          0   0   6   0   3  12   0   0    0   2  71   0   0   0   0   1\n",
       "Enpp4           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Pim1            0   0   0   0   5   0   0   0    0   1   0   1   0   0   2   2\n",
       "C3              0   0   4   7   0   0   2   0    0   0  59   0   0   2   0   0\n",
       "Map3k8          0   0   0   0   0   8   0   0    0   0   0   0   0   0   0   0\n",
       "Tnf            18  38 114   5  11 176 209  21   84 297  60   0   1  27  25  14\n",
       "Lox             0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Ehd1            0   1  25   5   7 265   0   4    0   3  67   0   0   6   0   1\n",
       "Fas             0   0   0   0   0   0   0   3    0   0   0   0   0   0   0   0\n",
       "Jak2            0   0   0  11  18   0   0   0   96   0   0   0   0   0   0   0\n",
       "Nfkb2           0   1   0  22   0   0   0   0    0   0  15   0   1   0   2   3\n",
       "Il6             0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Slco3a1         0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Cflar           0   0   0  11  29  25   0   0    2   1   0   0  43  36   0   2\n",
       "Rgs2            0   0   0   0   0   3   0   1    0   0   0   0   0   0   0   0\n",
       "Marco           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   1\n",
       "Fcgr2b          0   0   0   0   2   0   0   0    0  43   0   0   0   5   0   0\n",
       "Lcn2            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Traf1           1   0   6   0   3   0   0   0    0   0   0   0   0  34   0  92\n",
       "Gca             0   0   0   0   0   0   0   0  102   0   0   0   0   0   0   0\n",
       "Ifih1           0   0   0   0   0   0   0   0   30   0   0   0   0   0  24   5\n",
       "Itgb6           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Il1rn           0   0   5   0   0   0   0   0    0   0   1   0  12   0   0   0\n",
       "Dusp2           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Il1b            0   0   0   0   0 537   0   0 1309 102   0   0   0   4   0   1\n",
       "Il1a            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Car2            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Src             0   0   1   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Slc7a11         0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Vcam1           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Tlr2            0   0   0   0   0   0  20   0    0   0   0   0   0  14   0   0\n",
       "Gem             0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Gbp2            0   0   1   0   0   0 164   0    0   0   0   0   0   0   0   0\n",
       "Ppap2b          0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Pde4b           1   0   0   0   4   0   3   0    0   0   1   0  22   0   0   1\n",
       "Tnfrsf1b        2   0   0  14  77 192   1   4    2  13   9   2   1   9  11   0\n",
       "Hivep3         51   0   0   1  30   0   0   0   11 115   0   0   0 158  12  37\n",
       "Rasgef1b        0   0  55   0   6  10   7   7   74  63  23   0   0  41   3   7\n",
       "Cxcl1           0   0   0   0 101   3   0   0    0   0   1   1   0   6   0   0\n",
       "Cxcl9           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Oas1b           0   0   0   0   0   0   0   0    0   0   0   0   0   1   0   0\n",
       "Clec4e          0   0  62   0 139   0   2   0   54  11  33   0  43  66   0 107\n",
       "Cd69            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Clec2d          0   0   1   0   0   0   5   0    0   0   0   0   0   0   0  10\n",
       "Olr1            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   1\n",
       "Dusp16          0   0   0   2   0   0   0   0    0   0   0   0   0 263   6   0\n",
       "Vasp            0   3   0   5   2   0   0   0   12 185   0   0   0   0   1   0\n",
       "Nfkbib          0   0   0   0   0 199   0   0    0   0   4   0   0   0  18   0\n",
       "Pim2            0   0   0   0   0   0   0   0    0   0   0   0   5   0   0   0\n",
       "Nox1            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Rab11fip1       0   0   0   0   0   0   0   0    0   2   0   0   0   0   0   0\n",
       "Vegfc           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Slc7a2          0   0   0   0   0   0   0   0    0   0   0   0   3   0   0   0\n",
       "Herpud1         0   0   0   0 106   0   0   0    0   0   0   0   0   0   0   0\n",
       "Cx3cl1          0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Phxr4           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Birc3           0   0   0   0   0   0   2   0    0   0   0   0   1   0   0   0\n",
       "Pde4a           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Ptgs2           0   0   0   0   5   0   0   0    0   1   0   0   2   0 247   0\n",
       "Cish            0   0   0   0   0   0   0   8    0   0   0   0   0   0   0   0\n",
       "Malt1           0   0   0  15  70   7   0   8    1   2  95   0   1   5   0   1\n",
       "Alas1           0   0   0   0   0 163   0  14    0   0   0   0   0   0   0   0\n",
       "Batf            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   1\n",
       "Ifit1           0   0   2   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Cxcl10          1   0   5   0   0 369   0   0    0   1   0   0   0   0   0   0\n",
       "Igsf6           0   0   0   0   0  33   0   0    0   0  46   1   0   0   0   0\n",
       "Ccl5            0   0  18   0   0   0   0   0   31   0   0   0   0   1   0   0\n",
       "Tpbg            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Nfkbiz         57 139   0   0   0   0   2   1  203   0 112   0  10   0   0   4\n",
       "Ccl7            0   0   0   0   0   0   0   0   16   0   0   0   0   0   0   0\n",
       "Sh3bp4          0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   4\n",
       "Igfbp7          0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Gadd45a         0   0   0   0   7   0   0   0    0   0   0   0   0   0   0   0\n",
       "Ppfibp2         0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   2\n",
       "Spry1           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Chst7           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Icam1           0   0   0   0   6   7  19   2    1   4 387   0   0   2  91   4\n",
       "Serpine1        0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Ppm1k           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Egr2            0   0   0   0  33  92   0   1    5   0   8   0   0   2 177   0\n",
       "Snn             0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Socs1           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Prdm1           0   1   0   0   0  59   0   0    0  12   0   0   0   0   0   0\n",
       "Slamf7          2   0   0   0  55  18   0   0   12   2  22  10   6   0   0   0\n",
       "Gcnt1           0   0   0   0  41   3   0   0    0   0   0   0   0   0   0   0\n",
       "Itpkb           0   0   1   0  27  12   0   0    1   7   0   0  78   8  52   3\n",
       "Stx11           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Tnfsf10         0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Fnbp1l          0   1   7   1  22   0   1   0    0   2  21   0   0   0  57   1\n",
       "Fgl2            0   0   0   0   0 194   0   0    0   0   0   0   9   0   0   0\n",
       "Saa3            0   0   0   0   0  16   0   0    0   0   0   0   1   0   0   0\n",
       "Gpr34           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Bcor            0   0   0   0  11   7   0   0    1   1   1   0   0   1   0   0\n",
       "Ripk2           0   0   0   0   0   0   0   0    0   0   2   0   0   0   1   0\n",
       "Oasl1           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Rassf4          0   0   0   7  76 774   0  11   41  40  99   0   0  15   8   0\n",
       "Ikbke           0  25   1  75   0   6   0   0   21   6  17   0   0   0 431   9\n",
       "Maff            0   0   0   0   0   0   0   1    0   0   0   0   0   0   0   0\n",
       "Ccrl2           0   0   0   0   0   0   0   1    0   0   0   0   0   0   0   0\n",
       "C130050O18Rik   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Fzd1            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Il27            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Tlr1            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   2\n",
       "Cdc42ep2        0   0   0   0   0 451   0   4    0   0   0   0   0   0   0   0\n",
       "Adrb2           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Ifit2           0   9   0   0   0   1   0   0    0   1  50   0   4   0   0   0\n",
       "Gpr85          23   0   0   0   3   0   0   0    0   0   0   0   0   0   1   0\n",
       "Phldb1          0   0   0   3   3   0   0   1    2   0  23   0   0  81  46  14\n",
       "P2ry6           0   0   5   1   1   0   0   0    0   1   0   0   6   0   0   1\n",
       "Ifnb1           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Clec4a1         0   0   0   0 112   0   0   0    0   0  32   0   0   0 127   0\n",
       "Gpr18           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Mmp13           0   0   0   0   0   1   0   0    0   0   0   0   0   3   0   3\n",
       "Flrt3           0   0   0   0  10   0   0   0    0   0   0   0   0   0   0   0\n",
       "Plagl2          0   1   0  12 181   0   0   0    0   4   0  13   0   4   0   0\n",
       "Tagap1          0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Junb            0   0   4   2   4  42   0   0    0   2   0   0  17   0   0   0\n",
       "Socs3           0   1 102   0   1   2  43   0    0   1   0   0  31   0   0   0\n",
       "Bcl3            1   0   1   0   0   0   0   0    0   0   1   0   6   0   0   5\n",
       "Slamf8          0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   1\n",
       "Bcl2a1c         0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Lipg            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Iigp1           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Ifi205          0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Cebpb           0   1   5   0   0   0   0   0    0   0   0   1   0   0   0   0\n",
       "Cxcl2          94  71   0  68  59  47   2 231  570  83 409   2 204  11   5   0\n",
       "Cxcl11          0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Gpr84           0   0   0   0   0   0   0   0    0   2   0   0   0   0   0   0\n",
       "Slc31a2         0   0   0   0   0 104   0   0    0   0   0   0   0   0   0   1\n",
       "2010002M12Rik   0   0   1   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Marcks          9   0 120   6   0  26  16   0   66   2  70   4  18   0  14   2\n",
       "1700086L19Rik   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Zhx2            0   0   0   0   0   0   0  31    0   0   0   0   0   0   0   0\n",
       "4930599N23Rik   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Olfr661         0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Il4i1           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Mafb            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Thbd            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Arrdc3          0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Ifit3           0   0   0   0   0   0   4   0    0   0   0   0   0   0   0   0\n",
       "Sox4            0   0   0   0   0   0   0   1    0   0   0   0   0   0   0   0\n",
       "Ifna2           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Igtp            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Ifi47           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Gbp4            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "E130102H24Rik   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Bcl2a1b         0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Pou3f1          0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Hspa1b          0   0   0   0   0   9   0   0    0   0   1   0   0   0   0   0\n",
       "Ifna1           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Ifna14          0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Ifna5           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Ifnz            0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Gdap10          0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   4\n",
       "E230013L22Rik   0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Bcl2a1d         0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Ifnab           0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Bcl2a1a         0   0   0   0   0   0   0   0    0   0   0   0   0   0   0   0\n",
       "Gbp5            0   0   0   0   4   0   0   0    0   0   0   0   1   0   0   0\n",
       "              C49 C50 C51 C52 C53 C54 C55 C56 C57 C58 C59 C60 C61  C62 C63  C64\n",
       "Mx1             0   0   0   0   0   0   0   0   0   0   0   0   0    0   1    3\n",
       "Pdgfb         195   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Mmp14           0   0   0   1   0   0   0   0   0   0   0   0   0    0   0    3\n",
       "Ccl3            2   0  21   0  15   0   4 264  51   4 101   1   0   16   6  124\n",
       "Ckb             0   0   0   0   0   0   1   0   0   1   0   0   0    0   0    0\n",
       "Relb            0   0   0  10   0   0   0   0   0   2  14   1   0    0   0   13\n",
       "Ier3            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Il12b           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Adamts4         0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Sod2            6   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Hap1            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Cav1            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Pou2f2          5   0   0   0   0   0 169   0   0   2   0   0   0   56   0    0\n",
       "St3gal1        29   7   0   2   0   0  28   0   0   0   3   0   0    3   0    2\n",
       "Mertk           0   0  21   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Csf1            0   0  29   0   0   0   6   0   0   0   0   0   0   12   0    0\n",
       "Gadd45b         0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Cd83           60   0   1   0   0   0   1   0 108 205   8   0   0    0  34   48\n",
       "Sqstm1          1   0  10 165   0   0   0   3   1   0   3   0   2    0   0  345\n",
       "Cd274           0   0   0   0   0   0   0   0   0   0   1   0   0    0   1    6\n",
       "Il10            0   0   0   0   0   0   0   0   0 106   0   0   0    0   0    0\n",
       "Slpi            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Mafk            0   0   2   0   0   0   0   0   0   0   1   0   0    0   0   88\n",
       "Slc13a3         0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Adora2b         0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Irf1            9   0   0   0   0   0   0   0   0   0   0   1   0    6   1   87\n",
       "Csf2            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Cxcl16          2   0  18   0   0  40  38 203   0 120   1   0   0    0   0  463\n",
       "Ccl4            3   0  50   0  43   0   0   0   1 379   2   0   4  139  85  209\n",
       "Scn1b           0   0   0   0   0   0   0   0   0   3   0   0   0    0   0    0\n",
       "Tnfaip3        10 118   5   0   2   0   0   0   2   1  41   2   8    0   1   81\n",
       "Dusp6           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Peli1           0   0  35   0   0   0   0   0   0   0   0   0   0    0   0    7\n",
       "Rel            41   2   2   1   0   0  32  51   0  22   3   6  13 1221  47   47\n",
       "Tnip1          11   0   0   9   0   0   0   0   0   2   0   0   0    0   0   20\n",
       "Rsad2           0   0   0   0   0   0  82   0   0   0 141   0   0    0  14   20\n",
       "Nfkbia         64  15   0  61   1   2   0  54   0   1   0   0  15    0   0    2\n",
       "Arg2            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Tnfaip2         3   0  61   6   1   1   9   5   0  13   0   0   1   11  13   32\n",
       "Nedd9           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Edn1            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0   30\n",
       "Arhgef3        34   0   0   0   5   0   0   2   0   0   0   0   0    0   0   41\n",
       "Wnt5a           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Ednrb           0   0   0   0   0   0   0  71   0   0   0   0   0    0   0    0\n",
       "Irg1            4 102  37   0   5   4   0  38   4  19  42   0   2    4  11  162\n",
       "Zfp263          0   0   0   0   0   0   0  48   5   0  20   0   1    6   0    0\n",
       "Ets2            0   0   0   0   0   0   0   0   0   1   1   2   0    0   0    3\n",
       "Denr            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Il15ra          0   0   0   0   0   0   0   1   0   0   0   0   0    0   0   71\n",
       "Mx2             0   0   0   0   0   0   0   0   0   0  20   0   0    0   8    0\n",
       "Nfkbie          9   0   0   0   0   1   0  39   0   0   0   0   0    0   0    0\n",
       "Enpp4           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Pim1            0   0   1   0   3   0   1   2   0   8   0   0   9    0   0    8\n",
       "C3              1   0   7   0   0   0   0   0   2   0   0   0   0  102   0    0\n",
       "Map3k8          0   0   0   0   1   0   0   0   0   0   0   0   0    0   0    0\n",
       "Tnf            53 122  49  59  35  10  20 112   2  38   1   0  48    2  95 1056\n",
       "Lox             0   0   0   0   0   0   0   0   0   1   0   0   0    0  18    0\n",
       "Ehd1            4   0  34   0   0   2   2  13   0   4   0   0   0   52   6   66\n",
       "Fas             0   0   0   0   0   0   2   0   0   0   0   0   0    0   0    0\n",
       "Jak2            0   0   0   0   0   0   0   0   0   0  19   0   0    0   0    0\n",
       "Nfkb2           1   0   0   0   0   0   0   8   0   0   0   0   0    0   0    0\n",
       "Il6             0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Slco3a1         0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Cflar          19  12  18  39   1   6   0   0   0  24   2   0   0    0   0    0\n",
       "Rgs2            0   0   0   0   0   0   0   0   0   0   0   0   0  101   0    0\n",
       "Marco           0   0   0   0   0   0   0   0   0   3   0   0   0    0   0    0\n",
       "Fcgr2b        130   0   0   0   0   0   0   0   0   0   0   0   1    0   0    0\n",
       "Lcn2            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Traf1          19   0   0   2   0   0   0   0   0   0   1   0   0    5   0  336\n",
       "Gca             0   0   0   0   0   0   0   0   0   0   0   0   0   20   0    0\n",
       "Ifih1           0   0   0   0   0   0   0   0   0   0  53   0   0    0  28   55\n",
       "Itgb6           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Il1rn           0   0   7   0   0   0   0  17   0   0  90   0   0    0   0    0\n",
       "Dusp2           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Il1b           33   0   3 437   0   0  97   0   0   0   0   0 674   98   0    0\n",
       "Il1a            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Car2            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Src             0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Slc7a11         0 104   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Vcam1           0   0   0   0   0   0   0   0   0   0   0   2   0    0   0    0\n",
       "Tlr2            0   0   2   0   0   0   0   5   0   0   0   0  36    0   0  467\n",
       "Gem             0   0   0   0   0   0   0   1   0   3   0   0   0    1   0   16\n",
       "Gbp2           27   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Ppap2b          0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Pde4b           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0  160\n",
       "Tnfrsf1b       13   6   1  12   6   2   0  81   2  88   2   0  19    0   5   79\n",
       "Hivep3          0   0   1   0   0 107  24  73   1  51  13   0   1   71   0   40\n",
       "Rasgef1b        0   2  12  22   0   0   0  20  20 561  16   0   0    6  30   15\n",
       "Cxcl1           0   0   0   0   0   0   0   0   0   0   0   5   1    0   0    1\n",
       "Cxcl9           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Oas1b           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Clec4e         24 405   0   0   1   0  82   0   0 101   0   0  17    0  33   47\n",
       "Cd69            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Clec2d          1   0   0   0   0   0  12   2   0   0   0   0   0    0   0    0\n",
       "Olr1            0   0   0   0   0   0   0   0   0   0   1   0   0    0   0    0\n",
       "Dusp16          0   0   0   0   0   0   0   0   0   0   1   0   0    0   0    0\n",
       "Vasp            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Nfkbib          0   0   0   0   0   0   0   3   0   0   0   0   0    0   0    1\n",
       "Pim2            1   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Nox1            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Rab11fip1       0   0   0   0   0   0   1   0   0   0   2   0   0    0   1   97\n",
       "Vegfc           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Slc7a2          0   0   0   0   0   0   0   0   0   0   0   0   0   25   0    0\n",
       "Herpud1        21   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Cx3cl1          0   0   0   0   0   0   0   0   0   0   0   0   0    0   0  149\n",
       "Phxr4           0   0   0   0   0   0   0   0   0   0   8   0   0    0   0    0\n",
       "Birc3           0   1   1  25   1   0   1   9   0   0   0   0   0    0   0    0\n",
       "Pde4a           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Ptgs2           0   0   0   0   0   0   0   0   0   0   0   0  35    0   0    0\n",
       "Cish            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Malt1          97   0   0   0   0   1   0   4   5   1   3   0   4   17   0    0\n",
       "Alas1           0   0   0   0   0   0   0   0   0   2   0   0   0    9   0    6\n",
       "Batf            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Ifit1           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Cxcl10          0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    1\n",
       "Igsf6           0   0   0  24   2   0   0  33  13   2   6   0   1    5   0    0\n",
       "Ccl5            0   0   0   0   0   0   0  16   0   3  61   0   0    9   0    5\n",
       "Tpbg            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Nfkbiz         34   0  37   6   0   0  21   0   9   5   0   0   0   13  30  206\n",
       "Ccl7            0   0   2   0   0   0   0   0   0   2   0   0   0    0   0    1\n",
       "Sh3bp4          0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Igfbp7          0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Gadd45a         0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Ppfibp2         0   0   0   0   0   0   0   0   0   0   4   0   0    0   0    0\n",
       "Spry1           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Chst7           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Icam1           0   1  14   0   0   5   0   6   0  45  34   0   2    1   3   51\n",
       "Serpine1        0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Ppm1k           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Egr2            0   0   0   0   2   0   0  12   0   0  10   1  11    0   0    0\n",
       "Snn             0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Socs1           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Prdm1           0   0   0   0   0   0   0  21   0 817   0   0   0    0   0    0\n",
       "Slamf7          1   0   0   0   8   0   0  99   0  39   0   0   2    0   0   33\n",
       "Gcnt1           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Itpkb          36   0   6   3   0   0  62   0   0  51   4   0   0    2   0    0\n",
       "Stx11           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Tnfsf10         0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Fnbp1l          0   0   1   0   0   4   2   0   0  27   0   0   0    8   0    0\n",
       "Fgl2            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Saa3            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Gpr34           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Bcor            0   0   0   0   0   1   9   0   0   0  17   0   0    0   0    0\n",
       "Ripk2           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Oasl1           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Rassf4         10   0   8   0   0   0  72  13   0   7   5   0   0   15   1   25\n",
       "Ikbke           2   4   0   0   2  17  24   2   0   0   0   0   0    4   0    1\n",
       "Maff            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Ccrl2           0  12   2   0   0   0   0   0   0   0  24   0   0    9   0 1136\n",
       "C130050O18Rik   0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Fzd1            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Il27            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    2\n",
       "Tlr1            0   0   0   0   0   0   0   0   0   7   0   0   0    0   0    0\n",
       "Cdc42ep2        0   0   0   0   0   0   0   0   0   0   0   0   0    0  38    0\n",
       "Adrb2           0   0   0   0   0   0   0   2   0   0   0   0   0    0   0    0\n",
       "Ifit2           0   0   0   0   2   0   0   0   0   0   0   0   0    0   0    0\n",
       "Gpr85           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Phldb1          0   0   0   0   0   0   0   0   8   4   0   0  33    0   0    0\n",
       "P2ry6           0   0  13   2   0   0   0  19   0   0   0   0   0    0   0    3\n",
       "Ifnb1           0   0   0   0   0   0   0   1   0   0   0   0   0    0   0    0\n",
       "Clec4a1         3   0   1  31   0   0   0   0   0   0   0   0   0    2   0    0\n",
       "Gpr18           0   0   0   0   0   0   0   0   0   0   1   0   0    0   0    0\n",
       "Mmp13           0   0   0   0   0   0   0   1   0   0   0   0   0    0   0    1\n",
       "Flrt3           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Plagl2         11   0   0   0   0   0 102  57   0   0   0   0   0    0   0    0\n",
       "Tagap1          0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Junb            0   0   0   8   0   0   0  17   0   0  36   0   0    0   0  221\n",
       "Socs3           0   0   1   6   0   0   0   0   0   1  23   0   0    1   1    0\n",
       "Bcl3            0   0   0   0   0   0   0   0   0   0   1   0   0    0   0    0\n",
       "Slamf8          0   0   0   0   0   0   0  45   0   0   0   0   0    0   0    0\n",
       "Bcl2a1c         0   2   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Lipg            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Iigp1           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Ifi205          9   0   0   0   0   0   0   0   0   0   0   0   0    0   0   42\n",
       "Cebpb           0   0   0   0   0   0   0   0   0   1   0   0   0    0   0    0\n",
       "Cxcl2          28   0  34   6   0   0   0   2   0  11  68   0   1    1  38  573\n",
       "Cxcl11          0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Gpr84           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Slc31a2         5   0   0   0   0   0   0   0   0   0  11   0   0   17   0    0\n",
       "2010002M12Rik   0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Marcks          0  19   0  34  48   0  26  54   0   0   0   0   0    5   0   20\n",
       "1700086L19Rik   0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Zhx2            0   0   0   0   0   0   0  79   0   0   0   0   0    0   0    3\n",
       "4930599N23Rik   0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Olfr661         0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Il4i1           0   0   0   0   0   0   0   0   0   0  27   0   0    0   0    0\n",
       "Mafb            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Thbd            0   0   0   0   0  11   0   0   0   0   0   0   0    0   0    0\n",
       "Arrdc3          0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Ifit3           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    1\n",
       "Sox4            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Ifna2           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Igtp            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Ifi47           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Gbp4            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "E130102H24Rik   0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Bcl2a1b         0   5   0   0   0   0   0   2   0   0   0   0   0    0   0    0\n",
       "Pou3f1          0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Hspa1b          0   0   0   0   0   0   0   0   0   0   5   0   0    0   0    0\n",
       "Ifna1           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Ifna14          0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Ifna5           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Ifnz            0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Gdap10          0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "E230013L22Rik   0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Bcl2a1d         0  11   0   0   0   0   0  24   0   0   0   0   0    0   0    0\n",
       "Ifnab           0   0   0   0   0   0   0   0   0   0   0   0   0    0   0    0\n",
       "Bcl2a1a         0   2   0   0   0   0   0  19   0   0   0   0   0    0   0    0\n",
       "Gbp5            0   0   0   0   0   2   0   0   0   0   0   0   0    0   4   50\n",
       "              C65 C66 C67 C68 C69 C70 C71 C72 C73 C74 C75 C76 C77 C78 C79 C80\n",
       "Mx1             0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Pdgfb           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Mmp14           0   0   4   0   0   0   0   0   1   0   0   0   0   0   0   0\n",
       "Ccl3          135   1  78   1   0   0   2   0   0   0  20 948   3   0   0  17\n",
       "Ckb             1   0   0   0   0   0   0   0   0   0   0   0   0   0  46   0\n",
       "Relb           48   0 129   0   0   3   0   0   1   0   0   0   0   1   0   0\n",
       "Ier3            0   0   1   0   0   0   0   0   0   0   1   0   0   0   0   0\n",
       "Il12b           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Adamts4         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Sod2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   7   0\n",
       "Hap1            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Cav1            0   0   0   0   0   0   0   0   0   0   0   0   0   2 158   0\n",
       "Pou2f2          0   0   0   0   0   0   0   0   0   0   0   0   0   1  21   0\n",
       "St3gal1         0   0   0  14   1   4   1   0   0   0   0   0   0   0   0   0\n",
       "Mertk           0   0   8   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Csf1           11   0   1  89   0   0   0   0   0   0   0   0   9   1   0   0\n",
       "Gadd45b         0   7   0   0   0   0   0   0   0   1   0   1   0   0   0   0\n",
       "Cd83            0   0   2   4  37   0   0   0   0   0   0  17  19   0   0   0\n",
       "Sqstm1         40   0   4  10   0   0  23   0  13   0   0   1  10   0   4   0\n",
       "Cd274           0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0\n",
       "Il10            0   0   0   0   0   0   0   0   0   0   0   4   0   0   0   0\n",
       "Slpi            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Mafk            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Slc13a3         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Adora2b         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Irf1           51   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Csf2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Cxcl16          1   0  24   0   0  16   5   0   0  47   0 283   1   0   0  45\n",
       "Ccl4          139   2   0   1   8   0   0   0  22   0   8 329  26   0   0   0\n",
       "Scn1b           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Tnfaip3        78   4   0   1   0   0   0   0   0   0   0   0   0   0  29   1\n",
       "Dusp6           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Peli1           0   0   0   0   3   0   0   0   0   0   0   0   0   0   0   0\n",
       "Rel             0   0 145   0   0   0   0   0   0   2   0   0   3   0  12   0\n",
       "Tnip1           0   1   9   0   0   0   0   0  24   0   0   0   0   2 152   0\n",
       "Rsad2           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Nfkbia          0   0   5   0   0   1   0   2   0   1   0   0   0   0   0   0\n",
       "Arg2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Tnfaip2        22   2  68   0   0   0   4   0   4   0   3   0  54   2   0   1\n",
       "Nedd9           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Edn1            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Arhgef3         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Wnt5a           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ednrb           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Irg1           19   0 125   6   0   1   0   0   0   0   1   0  14   8 256   0\n",
       "Zfp263          1   0   0   0   9   2   0   0   0   0   0  44   0  12   2   0\n",
       "Ets2            0   0   0   0   0   0   0   0   8   0   0   2   0   0   0   4\n",
       "Denr            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Il15ra          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Mx2             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Nfkbie          0   0   0   0   0   0  19   0   0   0   0   0   0   0   0   2\n",
       "Enpp4           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Pim1            0   0  44   0   0   0   0   0   0  37   0  11   0   2   3   0\n",
       "C3              0   0   0   0  23   0   0   0   2   0   0   0   0   1  56   0\n",
       "Map3k8          0   0   0   0   0   0   0   0  77   0   0   0   0  32   0   0\n",
       "Tnf             9   7 536   0  48  42   0   0  12   0   0 118   0   0  50   3\n",
       "Lox             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ehd1            4   0  36   0  22   0   0   0   0   0   0  24   0   0  68   0\n",
       "Fas             0   0   0   0   0   0   0   0   0   0   0   1   0   0   1   0\n",
       "Jak2            0   2   0   0   0   0   0   0   0   0   0   0   0 149 120   0\n",
       "Nfkb2           0   0   5   0   0   0   0   0   0   0   0   0   0   0  75   0\n",
       "Il6             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Slco3a1         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Cflar           0   3  59   0   0   0   0   0   0   7   0   0   0   0   1   0\n",
       "Rgs2            0   0  35   0   0   0   0   0   0   0   0   0   0   1   0   0\n",
       "Marco           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Fcgr2b          0   0   0   0   0   0   0   0   0   0   0   3   0   0   0   0\n",
       "Lcn2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Traf1           0   0  64   0   0  10   2   0   0  17   0   1   0   0   0  12\n",
       "Gca             0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0\n",
       "Ifih1           4   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Itgb6           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Il1rn           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Dusp2           0   0   0   0   0   0   0   0   0   0   0   0   0   3   0   0\n",
       "Il1b            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Il1a            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Car2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Src             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Slc7a11         0   0   0   0   0   2   0   0   0   0   0   0   0   0   0   0\n",
       "Vcam1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Tlr2            1   0   0   0   0   0   0   0   0   0   0   2   0   0  13   0\n",
       "Gem             0   0   0   0   0   0   0   0   0   0   0   5   0   0   0   0\n",
       "Gbp2            0   0   0   0   0  25   0   0   0   0   0   0   0   0   0   0\n",
       "Ppap2b          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Pde4b           0   0   0   5   2   0   0   0   0   0   2  49   0   0   0   0\n",
       "Tnfrsf1b        4  27   0   0  59  23  12   0   0   1   0   0   0  15 159   0\n",
       "Hivep3          2   0   4  20  16   2   2   0   0   1   0  11   0  77   0   0\n",
       "Rasgef1b        4   6  56   2  13   9   2   0   3   0   0  17   0  27   0  15\n",
       "Cxcl1           0   0   0   0   0   0   1   0   0   0   0   0   0   2   0   0\n",
       "Cxcl9           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Oas1b           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Clec4e          0   7  35   0   1  48   1   1   0   0   0  30   0  25  34   1\n",
       "Cd69            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Clec2d          0   0   0   0   1   0   1   0   0   0   0   0   0   0   0   0\n",
       "Olr1            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Dusp16          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Vasp            3   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Nfkbib          0   0  57   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Pim2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Nox1            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Rab11fip1       0   0  11   0   0   0   0   0   0   0   0   0   0   9   0   0\n",
       "Vegfc           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Slc7a2          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Herpud1         0   0   2   0   0   0   3   0   0   0   0   0   0   0   0   0\n",
       "Cx3cl1          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Phxr4           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Birc3           0   0   5   0   0   0   0   0   0   0   0   0   0   8   0   0\n",
       "Pde4a           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ptgs2           0   5   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Cish            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Malt1           5   6   1   0   3   2   0   0   0   0   0   0   0   1   0   0\n",
       "Alas1           0   0   0   0   0   0   0   0   0   0   0   0   0   0  14   0\n",
       "Batf            7   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifit1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Cxcl10          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Igsf6           0   0  23   0   0   0   0   0   1   0   0   0   0   1   0   0\n",
       "Ccl5            8   0   0   2  84   0   0   0   0   0   0   0   0   1   0   0\n",
       "Tpbg            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Nfkbiz         40   0 264   1  19   2  13   0   2   0   0 139   0   0 345  80\n",
       "Ccl7            3   0   0   0   0   0   0   0   0   0   0  73   0   0   0   0\n",
       "Sh3bp4          3   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Igfbp7          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Gadd45a         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ppfibp2         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Spry1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Chst7           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Icam1           2   0  28   0  17   3   0   0   1   0   0  81   0  44  36   3\n",
       "Serpine1        0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ppm1k           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Egr2            1   0   1   0  61   1   0   0   0   0   0  23   1   2   0   0\n",
       "Snn             0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Socs1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Prdm1           0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Slamf7          0   0  26   0   0   0   1   0   1   0   0  51   0   1   1   0\n",
       "Gcnt1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Itpkb           0   0  64   0   0   0   2   0   0   0   0   5   0   0   0   1\n",
       "Stx11           0   0   2   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Tnfsf10         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Fnbp1l          0   0   1   0  52   0   0   0   0   1   0   0   4   0   0 165\n",
       "Fgl2            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Saa3            0   0  10   0 128   0   0   0   0   0   0   0   0   0 580   0\n",
       "Gpr34           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Bcor            0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0\n",
       "Ripk2           0   0   0  30   0   0   0   0   0   0   0   0   0   0   0   2\n",
       "Oasl1           0   0   0   0   0   0   0   0   0   0   0   0   0   0 625   0\n",
       "Rassf4        145   0 123   0  23   0   0   0 207   1   0   0   0   0  19   4\n",
       "Ikbke           1   0 111   0  29   0   0   0   0   0   0   9   1  27   1  75\n",
       "Maff            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ccrl2           0   0   0   0   4   0   0   0   0   0   0   8   0   0   0   0\n",
       "C130050O18Rik   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Fzd1            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Il27            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Tlr1            0   0   0   0  18   0   0   0   0   0   0   0   0  43  40   0\n",
       "Cdc42ep2        0   0   4   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Adrb2           0   0   9   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifit2           0   0   6   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Gpr85           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Phldb1          0   0   0   0   0   0   0   0   0   0   0   0   0 194   0  27\n",
       "P2ry6           0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifnb1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Clec4a1         0   0  19   0   0   0   0   0   0   0   0   0 635   0   0   0\n",
       "Gpr18           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Mmp13           3   0   6   0   1   0   0   0   0   0   0   0   0   0   0   0\n",
       "Flrt3           0   0   0   2   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Plagl2          2   0   0   0   0   0  10   0   0   1   0   0   0  22   6   0\n",
       "Tagap1          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Junb            0   0  23   0   0   0   0   0   0   0   0  57  80   0   0   0\n",
       "Socs3           0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0\n",
       "Bcl3            3   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Slamf8          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Bcl2a1c         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Lipg            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Iigp1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifi205          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Cebpb           1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Cxcl2           4   1  30   6  19  51  11   0  22   1   0 142  26 221  63 403\n",
       "Cxcl11          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Gpr84           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Slc31a2         0   0   0   0   0   0   0   0   0   0   0   0   0   0   8   0\n",
       "2010002M12Rik   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Marcks         44   0  34   0   0   0   1   0   0   0   0   0   0   0   0   0\n",
       "1700086L19Rik   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Zhx2            0   0   0   0   0   0   0   0   0   0   0   4   0   0   0   0\n",
       "4930599N23Rik   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Olfr661         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Il4i1           0   0   0   0  26   0   0   0   0   0   0   0   0   0   0   0\n",
       "Mafb            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Thbd            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Arrdc3          0   0   7   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifit3           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Sox4            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifna2           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Igtp            0   0   2   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifi47           2   0   0   0   0   0   0   0   0   0   0   0   0   0   6   0\n",
       "Gbp4            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "E130102H24Rik   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Bcl2a1b         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Pou3f1          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Hspa1b          0   0   0   0   0   1   0   0   0   0   0   0   0   0   3   0\n",
       "Ifna1           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifna14          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifna5           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifnz            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Gdap10          0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "E230013L22Rik   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Bcl2a1d         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Ifnab           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Bcl2a1a         0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
       "Gbp5            0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0\n",
       "              C81 C82 C83  C84  C85 C86 C87  C88  C89 C90 C91 C92  C93 C94 C95\n",
       "Mx1             0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Pdgfb           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Mmp14           0   0   0    0    0   0   0    0    8   0   0   0    0   0   0\n",
       "Ccl3            8   0 795   10    5   7   5  745  483  18   0 113  194   1   0\n",
       "Ckb             0   0   0   39    0   0   0    0   92   1   0   0    0   0   0\n",
       "Relb            0   0   0    0    0   1  40    0    2   0   0   0    0  78   0\n",
       "Ier3            0   0   0    0    0   0   0    0    0  29   0   0    0   0   0\n",
       "Il12b           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Adamts4         0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Sod2            0   0   0    0    3   0   0    0    2   0   0   0    5   0   0\n",
       "Hap1            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Cav1            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Pou2f2          0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "St3gal1       123   0   0    0    0   1   0    0   19   0  32   0    0   0   5\n",
       "Mertk           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Csf1            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Gadd45b         0   0   0    0    0   0   0    0    2   0   0   0    0   0  14\n",
       "Cd83            4   0   0  663    4   4   0  354    7  86   0   0    0 204   0\n",
       "Sqstm1          7   0   0    0    0   0   4   13    0   8   0   2    0   0   0\n",
       "Cd274           0   0   0    0    3   0   0    0    0   4   0   0    0   0   0\n",
       "Il10            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Slpi            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Mafk            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Slc13a3         0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Adora2b         0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Irf1            0   0   0    1    0  16   0   44    0   0   0   0    0   0   0\n",
       "Csf2            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Cxcl16         16  23   0    0    0  14   0   41  678   0   0   0    0 318   0\n",
       "Ccl4            0   2   0    0  175   0   0 1166  134 236   0   0  683 127   0\n",
       "Scn1b           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Tnfaip3        13   0   0   73    2   0  11  179   14   1   0   8    0  87   4\n",
       "Dusp6           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Peli1           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Rel            45   0   0    0    0   1   0    9    0   0   1   0    0  73   0\n",
       "Tnip1           0   0   0    1    0   0   0    0  198   6   0  25    0   0   0\n",
       "Rsad2           0   0   0    0    0   0   0  366    9 121   0   0    0   0   0\n",
       "Nfkbia          0   0   0   28   73   2   0   89   90   5   0   0    0   1  16\n",
       "Arg2            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Tnfaip2         0   0   0   33    3   0  58    0   10  15   0   0    4  10   0\n",
       "Nedd9           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Edn1            0   0   0    0    0   0   0    2   20   0   0   0    0   0   0\n",
       "Arhgef3         0   0   0    0    0   0   0    0    0   0   0   0    0   0   1\n",
       "Wnt5a           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Ednrb           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Irg1           20   0  28   62    0   0   0   38 2060   0   0   0  731   0   0\n",
       "Zfp263          0   0  26    4    0   0   0   12   27   0   0   0    0  19   0\n",
       "Ets2            0   0   0    0    0   0   0    1   23   0   0   0    0   0   0\n",
       "Denr            0   0   0    0    0   0   0    0    2   0   0   0    0   0   0\n",
       "Il15ra          0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Mx2             0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Nfkbie         39   2   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Enpp4           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Pim1            0   0   0  180    0   0   0    2   36   1   0   1    0   0   3\n",
       "C3              0   0   0    6    0   0   0    0    0   6   0   0    0   0   0\n",
       "Map3k8          0   0   0    0    0   0   0    0    0   0   0   0    0  70   0\n",
       "Tnf            21  25   0 1041    0   3   2 1042  197 974   0   2   24 561   4\n",
       "Lox             0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Ehd1          130   0  19   45    0   0   0    0    0   1   0   0    0 111   0\n",
       "Fas             0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Jak2            0   0   0    0    0   3   0    0    0   0   0   0    3   3   0\n",
       "Nfkb2           3   0   0    0    0   0   0    0   13  22   0   0    0   0   1\n",
       "Il6             0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Slco3a1         0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Cflar           0   0   0    0    3  98   0    0  105  41   0   0    0  31   0\n",
       "Rgs2            0   0   0    0    8   0   0    0    0   0   0   0    0 110   0\n",
       "Marco           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Fcgr2b          0   0  43    0    0   0   0    0   41   0   0   0    0   0   0\n",
       "Lcn2            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Traf1           0   0   0    0    1   0   0    3   11   2   0   0    0   0   0\n",
       "Gca             0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Ifih1           0   0   0   57    0   0   0   31   12 397   0   0    0   3   0\n",
       "Itgb6           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Il1rn           5  83   0    0   91   0   0    6   13 370   0   0    0 472   0\n",
       "Dusp2           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Il1b            0   0   1    0    0   0   1    0   21  95   0   4    0   0   0\n",
       "Il1a            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Car2            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Src             0   0   0    7    0   0   0    0    0   0   0   0    0   0   0\n",
       "Slc7a11         0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Vcam1           0   0   0    4    0   0   0    0    0   0   0   0    0   0   0\n",
       "Tlr2            0   0   0   30    6   0   0   70   16   0   1   0    0   0   0\n",
       "Gem             0   0   0    0    0   0   0   61    0   0   0   0    0   0   0\n",
       "Gbp2            0   0   0    0    0   0   0    0    0   0   0   0   55   0   0\n",
       "Ppap2b          0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Pde4b           0   1   0   30   15   0  33    0   64   0   0   0   37   0   0\n",
       "Tnfrsf1b       32   2   0    1    0  29   0   11   84  83  46   2    8   0   0\n",
       "Hivep3         21  11   9    0   10  33   0    2    0  46   0   0    1  10   0\n",
       "Rasgef1b        1   7   0   15  123   2   5  108   31 135   4   9   20   2   0\n",
       "Cxcl1           0   0   0    0    0   0   0   11  190   0   0   0    2 336   0\n",
       "Cxcl9           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Oas1b           0   0   0    0    0   0   0    0    0   0   1   0    0   0   0\n",
       "Clec4e          0   5  88  528    0   9   0  214    0 160   8   1    2  14   0\n",
       "Cd69            0   0   0    0    0   0   0   45    0   0   0   0    0   0   0\n",
       "Clec2d        323   0   0    0    0  21   0    0    3   0   0   0    0   0   0\n",
       "Olr1            0   0   0    0    0   0   0    0    0   0   0   0    0   1   0\n",
       "Dusp16          0   0   0    1    1   0   0    0    0   0   0   0    0   0   0\n",
       "Vasp            0   0   0    0    0   0   0    1    5   0   0   0    0   2   0\n",
       "Nfkbib          0   0   0    0    0   0   0    0    0   0   0   0    2   0   0\n",
       "Pim2            0   0   0    0  108   0   0    0    0   0   0   0    0   0   0\n",
       "Nox1            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Rab11fip1       0   0   0    0    0   0   0    0    0   0   0   0    0   1   7\n",
       "Vegfc           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Slc7a2          0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Herpud1         0   0   0    3  160   0   0    0    1   0   0   0    0   0   0\n",
       "Cx3cl1          0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Phxr4           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Birc3           0   0   0    0    0   0   0    0    1   7   0   0    0   0   0\n",
       "Pde4a           0   0   0    0    0   0   0    0    0   0   0   0    8   0   0\n",
       "Ptgs2           0   1   0    0    0   0   0    6    0  99   0   0    0   9   0\n",
       "Cish            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Malt1          14   1   0    0    0   0   0    5    1   0   0   0    1   0   0\n",
       "Alas1           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Batf            0   0   0  195    0   0   0    0    1   0   0   0    0   0   0\n",
       "Ifit1           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Cxcl10          0   0   0    0    0   0   0   16    0  10   0   0    0   0   0\n",
       "Igsf6           0   0   0    0  799   0   0    0    0   0   0   0   47   0   4\n",
       "Ccl5           16   0   0    0    0   0   1  142    0  73   0   0    7   0   0\n",
       "Tpbg            0   0   0    0    0   0   0    0    1   0   0   0    0   0   0\n",
       "Nfkbiz         84   0 214 1106    0   0   0  327  271   0   0   0    0   0   0\n",
       "Ccl7          178   0   0    0    0   0   0    0    3   0   0   0    0   0   0\n",
       "Sh3bp4          0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Igfbp7          0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Gadd45a         0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Ppfibp2         0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Spry1           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Chst7           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Icam1           1   2   0    0    5   0   0    0    0   6   0   0   55   0   0\n",
       "Serpine1        0   0   0    0    0   0   0    0  141   0   0   0    0   0   0\n",
       "Ppm1k           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Egr2            0   0   0   18    0   0   0   18    3   0   0   0    7   0   0\n",
       "Snn             0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Socs1           0   0   0    0    0   0   0    1    0   0   0   0    0   0   0\n",
       "Prdm1           0   0   0   73    0   0   0    0    0   0   0   0    0   0   0\n",
       "Slamf7          0   0   0    0    0  63   0    1  109  77   0   0    0   2   1\n",
       "Gcnt1           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Itpkb           0   3   0    0   12   1   0    0  202   0   0   0    0   0   0\n",
       "Stx11           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Tnfsf10         0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Fnbp1l         15   0   0   36    0   0   0    0    0   0   0   0    0   0   0\n",
       "Fgl2            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Saa3            0   0   0    0    0   0   0    0    0   0  48   0    0   0   0\n",
       "Gpr34           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Bcor            0   0   0   16    0  60   0    0    0   0   0   0    0   0   0\n",
       "Ripk2           0   0   0    1    0   0   0    0  449   2   0   0    0   0   0\n",
       "Oasl1           0   0   0    0    0   0   0 1129    7   0   0   0    0   0   0\n",
       "Rassf4         17 236   0  514   11   0   0    4   26  73   9   0   16 205   0\n",
       "Ikbke           1   0   0    3    0   1   0   38    0  76   0   0    0   0   0\n",
       "Maff            6   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Ccrl2           0   0   0    0    0   0   0   38    0 134   0   0   11   0   0\n",
       "C130050O18Rik   0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Fzd1            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Il27            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Tlr1           19   0   0    0    4   0   0    0    0   0  30   0    0   0   0\n",
       "Cdc42ep2        0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Adrb2           0   0   0    0   48   0   0    0    0   0   0   0    0   0   0\n",
       "Ifit2           0   0   0    0    0   0   0  256    0   0   0   0    0   0   0\n",
       "Gpr85           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Phldb1          3   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "P2ry6           1   0   0   29    0   0   0    1    0   0   0 380    0   1   0\n",
       "Ifnb1           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Clec4a1         0   0   0    0 1506   0   0  283    0  30   0   0    0   0   0\n",
       "Gpr18           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Mmp13           0   0   0    0    0   0   1  145    0  15   0   0  147   0   0\n",
       "Flrt3           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Plagl2          0   0   0    0    0   3   0    0  139   0   0   0    0   0   0\n",
       "Tagap1          0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Junb            1   0   0   78    0   0   1    0   42  20   0   0    0   0   0\n",
       "Socs3           1   0   0    4    0   0   0  120   62   0   0  16    1   0   0\n",
       "Bcl3            0   0   0    0    0   0   0    0    6   0   0   0    3   0   0\n",
       "Slamf8          0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Bcl2a1c         0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Lipg            0   0   0    2    0   0   0    0    0   0   0   0    0   0   0\n",
       "Iigp1           0   0   0    0    0   0   0    8    0   0   0   0    0   0   0\n",
       "Ifi205          0   0   0    0    0   0   0    0   28   0   0   0    0   0   0\n",
       "Cebpb           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Cxcl2          26   1   1  754    1   2 148 1080  795  85   0   1 1546   1   2\n",
       "Cxcl11          0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Gpr84           0   0   0  104  107   0   0    0    0   0   0   0    0   0   0\n",
       "Slc31a2        30   0   0  233    3   0   0    0   13   0   0   0    0   0   6\n",
       "2010002M12Rik   0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Marcks         76  11   0   64   50   0   0   51  106   1   0   0   86   8   6\n",
       "1700086L19Rik   0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Zhx2            0   0   0    0    0  42   0    0    0   0   0   0    0   0   2\n",
       "4930599N23Rik   0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Olfr661         0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Il4i1           0   0   0   11    0   0   0    0    0   0   0   0    0   0   0\n",
       "Mafb            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Thbd            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Arrdc3          0   0   0    0    0   0   2    0    0   0   0   0    0   0   0\n",
       "Ifit3           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Sox4            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Ifna2           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Igtp            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Ifi47           0   0   0    0    0   0   0    1    0   1   0   0    0   0   0\n",
       "Gbp4            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "E130102H24Rik   0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Bcl2a1b         0   0   0    0    0   0   0   11    0   0   0   0    0   0   0\n",
       "Pou3f1          0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Hspa1b          0   0   0   36    0   0   0    0    0   0   0   0    0   0   0\n",
       "Ifna1           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Ifna14          0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Ifna5           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Ifnz            0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Gdap10          0   0   0    0    0   0   0    0    0  35   0   0    0   0   0\n",
       "E230013L22Rik   0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Bcl2a1d         0   0   0    1    0   0   0   77    0   0   0   0    0   0   0\n",
       "Ifnab           0   0   0    0    0   0   0    0    0   0   0   0    0   0   0\n",
       "Bcl2a1a         0   0   0    2    0   0   0   74    0   0   0   0    0   0   0\n",
       "Gbp5            0   1   0    0    0   0   0    1    0   0   0   0    0   0   0\n",
       "              C96\n",
       "Mx1             0\n",
       "Pdgfb           0\n",
       "Mmp14           0\n",
       "Ccl3           65\n",
       "Ckb             0\n",
       "Relb           10\n",
       "Ier3            0\n",
       "Il12b           0\n",
       "Adamts4         0\n",
       "Sod2            0\n",
       "Hap1            0\n",
       "Cav1            0\n",
       "Pou2f2          0\n",
       "St3gal1         0\n",
       "Mertk           0\n",
       "Csf1            0\n",
       "Gadd45b        19\n",
       "Cd83            0\n",
       "Sqstm1        241\n",
       "Cd274           0\n",
       "Il10            0\n",
       "Slpi            0\n",
       "Mafk            0\n",
       "Slc13a3         0\n",
       "Adora2b         0\n",
       "Irf1            0\n",
       "Csf2            0\n",
       "Cxcl16          0\n",
       "Ccl4            0\n",
       "Scn1b           0\n",
       "Tnfaip3         0\n",
       "Dusp6           0\n",
       "Peli1           0\n",
       "Rel            70\n",
       "Tnip1           0\n",
       "Rsad2           0\n",
       "Nfkbia          0\n",
       "Arg2            0\n",
       "Tnfaip2        83\n",
       "Nedd9           0\n",
       "Edn1            0\n",
       "Arhgef3         0\n",
       "Wnt5a           0\n",
       "Ednrb           0\n",
       "Irg1           17\n",
       "Zfp263          0\n",
       "Ets2            0\n",
       "Denr            0\n",
       "Il15ra          0\n",
       "Mx2             0\n",
       "Nfkbie          1\n",
       "Enpp4           0\n",
       "Pim1            0\n",
       "C3              0\n",
       "Map3k8          0\n",
       "Tnf            22\n",
       "Lox             2\n",
       "Ehd1            0\n",
       "Fas             0\n",
       "Jak2            1\n",
       "Nfkb2           0\n",
       "Il6             0\n",
       "Slco3a1         0\n",
       "Cflar          13\n",
       "Rgs2            0\n",
       "Marco           0\n",
       "Fcgr2b          0\n",
       "Lcn2            0\n",
       "Traf1           1\n",
       "Gca             0\n",
       "Ifih1           0\n",
       "Itgb6           0\n",
       "Il1rn           0\n",
       "Dusp2           0\n",
       "Il1b            0\n",
       "Il1a            0\n",
       "Car2            0\n",
       "Src             0\n",
       "Slc7a11         0\n",
       "Vcam1           0\n",
       "Tlr2            0\n",
       "Gem             0\n",
       "Gbp2            0\n",
       "Ppap2b          0\n",
       "Pde4b           0\n",
       "Tnfrsf1b        0\n",
       "Hivep3          1\n",
       "Rasgef1b        0\n",
       "Cxcl1           0\n",
       "Cxcl9           0\n",
       "Oas1b           0\n",
       "Clec4e          7\n",
       "Cd69            0\n",
       "Clec2d          0\n",
       "Olr1            0\n",
       "Dusp16          0\n",
       "Vasp            0\n",
       "Nfkbib          0\n",
       "Pim2            0\n",
       "Nox1            0\n",
       "Rab11fip1       0\n",
       "Vegfc           0\n",
       "Slc7a2          0\n",
       "Herpud1         0\n",
       "Cx3cl1          0\n",
       "Phxr4           0\n",
       "Birc3           0\n",
       "Pde4a           0\n",
       "Ptgs2           0\n",
       "Cish            0\n",
       "Malt1           1\n",
       "Alas1           0\n",
       "Batf            0\n",
       "Ifit1          15\n",
       "Cxcl10          0\n",
       "Igsf6           0\n",
       "Ccl5           11\n",
       "Tpbg            0\n",
       "Nfkbiz          1\n",
       "Ccl7            0\n",
       "Sh3bp4          0\n",
       "Igfbp7          0\n",
       "Gadd45a         0\n",
       "Ppfibp2         0\n",
       "Spry1           0\n",
       "Chst7           0\n",
       "Icam1           0\n",
       "Serpine1        0\n",
       "Ppm1k           0\n",
       "Egr2            0\n",
       "Snn             0\n",
       "Socs1           0\n",
       "Prdm1           0\n",
       "Slamf7          0\n",
       "Gcnt1           0\n",
       "Itpkb           0\n",
       "Stx11           0\n",
       "Tnfsf10         0\n",
       "Fnbp1l          0\n",
       "Fgl2            0\n",
       "Saa3            0\n",
       "Gpr34           0\n",
       "Bcor            0\n",
       "Ripk2           0\n",
       "Oasl1           0\n",
       "Rassf4          0\n",
       "Ikbke           8\n",
       "Maff            0\n",
       "Ccrl2           0\n",
       "C130050O18Rik   0\n",
       "Fzd1            0\n",
       "Il27            0\n",
       "Tlr1            0\n",
       "Cdc42ep2        0\n",
       "Adrb2           0\n",
       "Ifit2           0\n",
       "Gpr85           0\n",
       "Phldb1         68\n",
       "P2ry6           0\n",
       "Ifnb1           0\n",
       "Clec4a1         0\n",
       "Gpr18           0\n",
       "Mmp13           0\n",
       "Flrt3           0\n",
       "Plagl2          0\n",
       "Tagap1          0\n",
       "Junb            0\n",
       "Socs3           0\n",
       "Bcl3            0\n",
       "Slamf8          0\n",
       "Bcl2a1c         0\n",
       "Lipg            0\n",
       "Iigp1           0\n",
       "Ifi205          0\n",
       "Cebpb           0\n",
       "Cxcl2          42\n",
       "Cxcl11          0\n",
       "Gpr84           0\n",
       "Slc31a2         0\n",
       "2010002M12Rik   0\n",
       "Marcks          0\n",
       "1700086L19Rik   0\n",
       "Zhx2            0\n",
       "4930599N23Rik   0\n",
       "Olfr661         0\n",
       "Il4i1           0\n",
       "Mafb            0\n",
       "Thbd            0\n",
       "Arrdc3          0\n",
       "Ifit3           0\n",
       "Sox4            0\n",
       "Ifna2           0\n",
       "Igtp            0\n",
       "Ifi47          17\n",
       "Gbp4            0\n",
       "E130102H24Rik   0\n",
       "Bcl2a1b         0\n",
       "Pou3f1          0\n",
       "Hspa1b          0\n",
       "Ifna1           0\n",
       "Ifna14          0\n",
       "Ifna5           0\n",
       "Ifnz            0\n",
       "Gdap10          0\n",
       "E230013L22Rik   0\n",
       "Bcl2a1d         0\n",
       "Ifnab           0\n",
       "Bcl2a1a         0\n",
       "Gbp5            0\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "#write.csv(cdBactAnnot,'cdBactAnnotRiken.csv')\n",
    "cdBactAnnot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fitting the BASiCS model\n",
    "\n",
    "The BASiCS model is fitted using a Bayesian approach by means of an Adaptive Metropolis within Gibbs Sampling algorithm[^RobertsRosenthal2009]. We assume prior independence between all model parameters (except within the $\\phi_j$'s, for which its total sum is constrained), using a flat *non informative* prior for the normalised expression rates $\\mu_1,\\ldots,\\mu_{q_0}$ and proper *informative* prior distributions for all other model parameters. The adopted prior is given by\n",
    "\n",
    "* $\\pi(\\mu_i) \\propto \\mu_i^{-1}$ for $i=1, \\ldots, q_0$,\n",
    "\n",
    "* $\\delta_i \\sim \\mbox{Gamma}(a_{\\delta},b_{\\delta})$ for $i=1, \\ldots, q_0$,\n",
    "\n",
    "* $n^{-1} \\phi = n^{-1} (\\phi_1, \\ldots, \\phi_n)'  \\sim \\mbox{Dirichlet}(p)$ with $p = (p_1,\\ldots,p_n)'$ and each $p_j>0$,\n",
    "\n",
    "* $s_j \\sim \\mbox{Gamma}(a_s,b_s)$,\n",
    "\n",
    "* $\\theta \\sim \\mbox{Gamma}(a_{\\theta},b_{\\theta})$, \n",
    "\n",
    "for fixed hyper-parameter values $a_{\\delta}, b_{\\delta}$, $p, a_s, b_s, a_{\\theta}$ and $b_{\\theta}$[^Prior]. As a default option, the sampler assumes all these hyper-parameters are equal to 1 (in case of $p$, we use $p_1 = p_2 = \\cdots = p_n = 1$). However, the user can modify this option. \n",
    "\n",
    "[^Prior]: The analysis of simulated and real datasets suggested that the choice of these hyper-parameters does not have major consequences in posterior inference.\n",
    "\n",
    "[^RobertsRosenthal2009]: Roberts GO, Rosenthal JS (2009). Examples of adaptive MCMC. Journal of Computational and Graphical Statistics 18: 349-367. DOI: 10.1198/jcgs.2009.06134\n",
    "\n",
    "The function `BASiCS_MCMC` runs the sampler. It requires 4 basic arguments:\n",
    "\n",
    "* `Data`: an object of class `BASiCS_Data`\n",
    "\n",
    "* `N`: the total number of iterations for the algorithm\n",
    "\n",
    "* `Thin`: thinning period for the algorithm (i.e. the output will only store parameter draws every `Thin` iterations)\n",
    "\n",
    "* `Burn`: burning period for the algorithm (i.e. the initial `Burn` iterations are discarded)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "# From ERCCData we take only the ERCC for which the input value is given. In the fasta file, downloaded from\n",
    "# https://www-s.nist.gov/srmors/view_datafiles.cfm?srm=2374 has more ERCC than the input values\n",
    "SpikeExp <- ERCCData[rownames(ERCCData) %in% spike_input$Spike, ]\n",
    "SpikeExp <- SpikeExp[, colnames(SpikeExp) %in% colnames(cd)]\n",
    "Tech <- c(rep(FALSE, dim(cd)[1]), rep(TRUE, dim(SpikeExp)[1]))\n",
    "SpikeInput <- spike_input\n",
    "Counts <- as.matrix(rbind(cd, SpikeExp))\n",
    "write.table(Counts, \"/home/baker/my-scratch-share/James_Bagnall_Single_cell_rna_seq_V3/output/BASiCS_OUTPUT/AllCombinedCounts.txt\", quote = FALSE,\n",
    "            sep = \"\\t\", col.names = NA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "basics_data <- newBASiCS_Data(Counts, Tech, SpikeInput)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "store_dir <- \"/home/baker/my-scratch-share/James_Bagnall_Single_cell_rna_seq_V3/output/BASiCS_OUTPUT\"\n",
    "run_name <- \"batch-clean\"\n",
    "if (file.exists(paste0(store_dir, \"/chain_phi_\", run_name, \".txt\"))) {\n",
    "  chain_mu = as.matrix(fread(paste0(store_dir, \"/chain_mu_\", run_name, \".txt\")))\n",
    "  chain_delta = as.matrix(fread(paste0(store_dir, \"/chain_delta_\", run_name, \".txt\")))\n",
    "  chain_phi = as.matrix(fread(paste0(store_dir, \"/chain_phi_\", run_name, \".txt\")))\n",
    "  chain_s = as.matrix(fread(paste0(store_dir, \"/chain_s_\", run_name, \".txt\")))\n",
    "  chain_nu = as.matrix(fread(paste0(store_dir, \"/chain_nu_\", run_name, \".txt\")))\n",
    "  chain_theta = as.matrix(fread(paste0(store_dir, \"/chain_theta_\", run_name, \".txt\")))\n",
    "\n",
    "  mcmc_output <- newBASiCS_Chain(mu = chain_mu, delta = chain_delta,\n",
    "                                 phi = chain_phi, s = chain_s,\n",
    "                                 nu = chain_nu, theta = chain_theta)\n",
    "\n",
    "  time_total <- readRDS(paste0(store_dir, \"/time_total_\", run_name, \".rds\"))\n",
    "} else {\n",
    "  time_start <- Sys.time()\n",
    "  mcmc_output <- BASiCS_MCMC(basics_data, N = 40000, Thin = 10, Burn = 20000,\n",
    "                             PrintProgress = TRUE, StoreChains = TRUE,\n",
    "                             StoreDir = store_dir, RunName = run_name)\n",
    "  time_end <- Sys.time()\n",
    "  time_total <- difftime(time_end, time_start, units = \"hours\")\n",
    "  saveRDS(time_total, paste0(store_dir, \"/time_total_\", run_name, \".rds\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "mcmc_summary <- Summary(mcmc_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "mcmc_summary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cell-specific normalizing constants\n",
    "BASiCS models two cell-specific parameters. \n",
    "\n",
    "* $\\phi_j$: The cell-specific normalizing constants related to differences in mRNA content. For identifiability reasons, BASiCS impose the constrain: $\\sum_{j=1}^{n} \\phi_j = n$ . \n",
    "* $s_j$: Cell specific normalizing constant related to differences in capture efficiency (when analyzing UMI based expression counts) and other amplification biases such as sequencing depth (when analyzing non-UMI based expression counts, as generated by standard scRNA-seq experiments)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "plot(mcmc_summary, Param = \"phi\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "plot(mcmc_summary, Param = \"s\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "phi <- displaySummaryBASiCS(mcmc_summary, Param = \"phi\")\n",
    "s <- displaySummaryBASiCS(mcmc_summary, Param = \"s\")\n",
    "#parameters <- cbind(phi, s)\n",
    "gene_count <- colSums(counts(basics_data, type = \"biological\"))\n",
    "ercc_count <- colSums(counts(basics_data, type = \"technical\"))\n",
    "parameters <- cbind(phi, s, gene_count, ercc_count)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "#rownames(parameters) <- c(1:96)\n",
    "parameters <- data.frame(parameters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "#png(\"phi_vs_Gene_Counts.png\")\n",
    "phi_gene_count <- ggplot(parameters, aes(x = gene_count, y = Phi)) +\n",
    "  geom_point() +\n",
    "  geom_smooth(method = \"lm\") +\n",
    "  labs(x = \"Gene molecule count\",\n",
    "       title = paste0(\"Phi measures gene molecule count differences\\nr = \",\n",
    "                      round(cor(parameters$gene_count, parameters$Phi), 2)))\n",
    "phi_gene_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## S versus ERCC molecule count\n",
    "As $s_j$ is determined from ERCC molecules, they are proportional to that"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "#png(\"S_vs_ERCC_Counts.png\")\n",
    "S_ERCC_count <- ggplot(parameters, aes(x = ercc_count, y = S)) +\n",
    "  geom_point() +\n",
    "  geom_smooth(method = \"lm\") +\n",
    "  labs(x = \"ERCC molecule count\",\n",
    "       title = paste0(\"S measures ERCC molecule count differences\\nr = \",\n",
    "                      round(cor(parameters$ercc_count, parameters$S), 2)))\n",
    "S_ERCC_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## $\\phi$ vs $S$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "#png(\"phi_vs_S.png\")\n",
    "phi_s <- ggplot(parameters, aes(x = S, y = Phi)) +\n",
    "  geom_point() +\n",
    "  geom_smooth(method = \"lm\") +\n",
    "  labs(       title = paste0(\"Phi versus S\\nr = \",\n",
    "                      round(cor(parameters$S, parameters$Phi), 2)))\n",
    "phi_s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Denoised data\n",
    "\n",
    "The function `BASiCS_DenoisedCounts(Data=Data, Chain = MCMC_Output)` produce a matrix of normalized counts for which the effect of technical variation is removed \n",
    "\n",
    "For each gene $i$ and cell $j$ this function returns\n",
    "\n",
    "$$ x^*_{ij} = \\frac{ x_{ij} } {\\hat{\\phi}_j \\hat{\\nu}_j}, $$\n",
    "\n",
    "where $x_{ij}$ is the observed expression count of gene $i$ in cell $j$, $\\hat{\\phi}_j$ denotes the posterior median of $\\phi_j$ and $\\hat{\\nu}_j$ is the posterior median of $\\nu_j$.\n",
    "\n",
    "Remove technical noise (i.e. normalize using the ERCC spike-ins). This takes a long time!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "denoised = BASiCS_DenoisedCounts(Data = basics_data, Chain = mcmc_output)\n",
    "write.table(denoised, \"/home/baker/my-scratch-share/James_Bagnall_Single_cell_rna_seq_V3/output/BASiCS_OUTPUT/basics-denoised.txt\", quote = FALSE,\n",
    "            sep = \"\\t\", col.names = NA)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternativelly, the user can compute the normalised and denoised expression rates underlying the expression of all genes across cells using `BASiCS_DenoisedRates`. The output of this function is given by \n",
    "\n",
    "$$ \\Lambda_{ij} = \\hat{\\mu_i} \\hat{\\rho}_{ij},  $$\n",
    "\n",
    "where $\\hat{\\mu_i}$ represents the posterior median of $\\mu_j$ and $\\hat{\\rho}_{ij}$ is given by its posterior mean (Monte Carlo estimate based on the MCMC sample of all model parameters)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "denoised_rates = BASiCS_DenoisedRates(Data = basics_data, Chain = mcmc_output)\n",
    "write.table(denoised_rates, \"/home/baker/my-scratch-share/James_Bagnall_Single_cell_rna_seq_V3/output/BASiCS_OUTPUT/basics-denoised-rates.txt\", quote = FALSE,\n",
    "            sep = \"\\t\", col.names = NA)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Chekcing for the Convergence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R -w 800 -h 1000\n",
    "par(mfrow=c(3,2))\n",
    "plot(mcmc_output, Param = \"mu\", Gene = 1, log = \"y\")\n",
    "plot(mcmc_output, Param = \"delta\", Gene = 1)\n",
    "plot(mcmc_output, Param = \"phi\", Cell = 1)\n",
    "plot(mcmc_output, Param = \"s\", Cell= 1)\n",
    "plot(mcmc_output, Param = \"nu\", Cell = 1)\n",
    "plot(mcmc_output, Param = \"theta\", Batch = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__PCA Plotting with Normalized Dataset__\n",
    "\n",
    "This plot is generated taking both the gene counts and the ERCC spike in counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R -w 600 -h 600\n",
    "FinalPCAData <- denoised\n",
    "# # Used this function to save the file in pdf format\n",
    "#FinalPCAData <- na.omit(FinalPCAData)\n",
    "#print(dim(FinalPCAData))\n",
    "strain <- factor(c(rep(\"Cells\", CellNumber)), levels=c(\"Cells\"), labels=c(\"Cells\"))\n",
    "PCx=1\n",
    "PCy=2\n",
    "pcaPRComp <- prcomp((FinalPCAData))\n",
    "percentVar <- pcaPRComp$sdev^2/sum(pcaPRComp$sdev^2)\n",
    "df<-data.frame(PCX=pcaPRComp$rotation[,PCx],PCY=pcaPRComp$rotation[,PCy],Cell=strain)\n",
    "p<-ggplot(df, aes_string(x=\"PCX\",y =\"PCY\",color=\"Cell\"))\n",
    "p<-p+ggtitle(\"PCA with Normalized Dataset\")\n",
    "p<-p+geom_point(size = 2)\n",
    "p<-p+xlab(paste(\"PC\",PCx,\": \", round(percentVar[PCx] * 100), \"% variance\"))\n",
    "p<-p    + ylab(paste(\"PC\",PCy,\": \", round(percentVar[PCy] * 100), \"% variance\"))\n",
    "p<-p + geom_text(data=df,aes(label=colnames(cd)),alpha=0.5,size=5, vjust=1,hjust=0.5,angle=45, color=\"black\")\n",
    "print(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__PCA plotting with the raw data__\n",
    "\n",
    "This plot is generated taking both the gene counts and the ERCC spike in counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R -w 600 -h 600\n",
    "FinalPCAData <- Counts\n",
    "#FinalPCAData <- na.omit(FinalPCAData)\n",
    "#print(dim(FinalPCAData))\n",
    "#png(\"PCA with Raw Data.png\")\n",
    "strain <- factor(c(rep(\"Cells\", CellNumber)), levels=c(\"Cells\"), labels=c(\"Cells\"))\n",
    "PCx=1\n",
    "PCy=2\n",
    "pcaPRComp <- prcomp((FinalPCAData))\n",
    "percentVar <- pcaPRComp$sdev^2/sum(pcaPRComp$sdev^2)\n",
    "df<-data.frame(PCX=pcaPRComp$rotation[,PCx],PCY=pcaPRComp$rotation[,PCy],Cell=strain)\n",
    "p<-ggplot(df, aes_string(x=\"PCX\",y =\"PCY\",color=\"Cell\"))\n",
    "p<-p+ggtitle(\"PCA with Raw Dataset\")\n",
    "p<-p+geom_point(size = 2)\n",
    "p<-p+xlab(paste(\"PC\",PCx,\": \", round(percentVar[PCx] * 100), \"% variance\"))\n",
    "p<-p    + ylab(paste(\"PC\",PCy,\": \", round(percentVar[PCy] * 100), \"% variance\"))\n",
    "p<-p + geom_text(data=df,aes(label=colnames(cd)),alpha=0.5,size=5, vjust=1,hjust=0.5,angle=45, color=\"black\")\n",
    "print(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Gene Expression per cell__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R -w 600 -h 600\n",
    "CellGeneExpression <- colSums(denoised_rates)\n",
    "barplot(CellGeneExpression,xlab=\"Cell Number\", ylab=\"Total Gene Expression\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## SCDE for Heterogeneity analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The pagoda routines implemented in the scde resolves multiple, potentially overlapping aspects of transcriptional heterogeneity by identifying known pathways or novel gene sets that show significant excess of coordinated variability among the measured cells. Briefly, cell-specific error models derived from scde are used to estimate residual gene expression variance, and identify pathways and gene sets that exhibit statistically significant excess of coordinated variability (overdispersion). pagoda can be used to effectively recover known subpopulations and discover putative new subpopulations and their corresponding functional characteristics in single-cell samples. For more information, please refer to the original manuscript by [Fan et al](http://www.nature.com/nmeth/journal/vaop/ncurrent/full/nmeth.3734.html)\n",
    "\n",
    "### Fitting error models\n",
    "\n",
    "Next, we'll construct error models for individual cells. Here, we use k-nearest neighbor model fitting procedure implemented by `knn.error.models()` method. Our data is a UMI based data, so we raise the min.count.threshold to 1 (minimum number of reads for the gene to be initially classified as a non-failed measurement), requiring at least 5 non-failed measurements per gene. We're providing a rough guess to the complexity of the population, by fitting the error models based on 1/4 of most similar cells (i.e. guessing there might be ~4 subpopulations).\n",
    "\n",
    "Note this step takes a considerable amount of time unless multiple cores are used. We highly recommend use of multiple cores. You can check the number of available cores available using `parallel::detectCores()`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__We had to use the raw data for this as SCDE does not accept normalized data__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "#cdSCDE <- read.csv('/home/baker/my-scratch-share/James_Bagnall_Single_cell_rna_seq_V3/output/GeneAnnotationIdName.csv',check.names=FALSE,row.names=1)\n",
    "CountData <- read.csv('/home/baker/my-scratch-share/James_Bagnall_Single_cell_rna_seq_V3/output/AllHtseqCounts.csv',check.names=FALSE,row.names=1)\n",
    "CellNumber <- dim(CountData)[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "dim(CountData)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We used the clean.counts() method from SCDE to filter the cells and genes. However, the detault parameter filtered out too much of the cells and so we had to chage it\n",
    "\n",
    "clean.counts(counts, min.lib.size = 1800, min.reads = 10,\n",
    "  min.detected = 5)\n",
    "  \n",
    "Arguments\n",
    "\n",
    "| Parameters | Description |\n",
    "|----------|:-------------|\n",
    "| counts |  read count matrix. The rows correspond to genes, columns correspond to individual cells |\n",
    "| min.lib.size |    Minimum number of genes detected in a cell. Cells with fewer genes will be removed (default: 1.8e3) |\n",
    "| min.reads | Minimum number of reads per gene. Genes with fewer reads will be removed (default: 10) |\n",
    "| min.detected | Minimum number of cells a gene must be seen in. Genes not seen in a sufficient number of cells will be removed (default: 5) |\n",
    "\n",
    "We changed the `min.lib.size` to `1500`. This gives 57 cells. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1] 1978   39\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "cdMod <- cd[,3:98]\n",
    "rownames(cdMod) <- cd[,1]\n",
    "cdMod <- clean.counts(cdMod, min.lib.size=700)\n",
    "dim(cdMod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R -o cd\n",
    "# omit the non-mapped gene information\n",
    "cd = CountData[1:(dim(CountData)[1]-5),]\n",
    "# omit genes that are never detected\n",
    " #cd <- cd[rowSums(cd[,2:dim(cd)[2]])>0, ]\n",
    "# omit cells with very poor coverage\n",
    " #cd <- cd[, colSums(cd[,2:dim(cd)[2]])>1e4]\n",
    "CellNumber <- dim(cd)[2]\n",
    " #GeneNames <- cbind(rownames(cd), as.character(cd[,1]))\n",
    " #colnames(GeneNames) <- c(\"EnsembleID\", \"GeneName\")\n",
    " #cd$gene_name <- NULL\n",
    " #print(head(GeneNames[,1]))\n",
    "cd <- clean.counts(cd, min.lib.size=1500)\n",
    "dim(cd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "print(colnames(cd))\n",
    "colSums(cd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "rownames(cd) <- gsub(\"\\\\..*\",\"\",rownames(cd))\n",
    "head(rownames(cd))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "GeneNames[,1] <- gsub(\"\\\\..*\",\"\",GeneNames[,1])\n",
    "GeneNamesList <- split(GeneNames, rep(1:ncol(GeneNames), each = nrow(GeneNames)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "knn <- knn.error.models(cd, k = ncol(cd)/4, n.cores = 2, min.count.threshold = 1, min.nonfailed = 5, save.model.plots = TRUE, max.model.plots = 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "head(knn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "knnWithERCC <- knn.error.models(Counts, k = ncol(Counts)/4, n.cores = 2, min.count.threshold = 1, min.nonfailed = 5, save.model.plots = TRUE, max.model.plots = 50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The fitting process above wrote out `cell.models.pdf` file in the current directory showing model fits for the first 50 cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalizing variance\n",
    "In order to accurately quantify excess variance or overdispersion, we must normalize out expected levels of technical and intrinsic biological noise. Briefly, variance of the NB/Poisson mixture processes derived from the error modeling step are modeled as a chi-squared distribution using adjusted degrees of freedom and observation weights based on the drop-out probability of a given gene. Here, we normalize variance, trimming 3 most extreme cells and limiting maximum adjusted variance to 5."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R -w 800 -h 400\n",
    "varinfo <- pagoda.varnorm(knn, counts = cd, trim = 3/ncol(cd), max.adj.var = 5, n.cores = 2, plot = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The plot on the left shows coefficient of variance squared (on log10 scale) as a function of expression magnitude (log10 FPM). The red line shows local regression model for the genome-wide average dependency. The plot on the right shows adjusted variance (derived based on chi-squared probability of observed/genomewide expected ratio for each gene, with degrees of freedom adjusted for each gene). The adjusted variance of 1 means that a given gene exhibits as much variance as expected for a gene of such population average expression magnitude. Genes with high adjusted variance are overdispersed within the measured population and most likely show subpopulation-specific expression:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Using ERCC spike ins to fit dependency between expression magnitude and variance__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R -w 800 -h 400\n",
    "ercc.genes <- grep(\"^ERCC-\",rownames(Counts),value=T)\n",
    "varinfo <- pagoda.varnorm(knnWithERCC, counts = Counts, trim = 3/ncol(Counts), fit.genes=ercc.genes,n.cores=2,plot=T,verbose=1,max.adj.var=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Now fitting model with only read counts__\n",
    "\n",
    "As the ERCC spike in data were not conclusive, here I am running the error model fitting from SCDE with only the read counts. The fitting process relies on a subset of robust genes that are detected in multiple cross-cell comparisons. \n",
    "\n",
    "Note this step takes a considerable amount of time unless multiple cores are used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "# calculate models\n",
    "#o.ifm <- scde.error.models(counts = cd, n.cores = 2, threshold.segmentation = TRUE, save.crossfit.plots = FALSE, save.model.plots = TRUE, verbose = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "# list top overdispersed genes\n",
    "sort(varinfo$arv, decreasing = TRUE)[1:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Controlling for sequencing depth\n",
    "\n",
    "Even with all the corrections, sequencing depth or gene coverage is typically still a major aspects of variability. In most studies, we would want to control for that as a technical artifact (exceptions are cell mixtures where subtypes significantly differ in the amount of total mRNA). Below we will control for the gene coverage (estimated as a number of genes with non-zero magnitude per cell) and normalize out that aspect of cell heterogeneity:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "varinfo <- pagoda.subtract.aspect(varinfo, colSums(cd[, rownames(knn)]>0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "head(varinfo$arv)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evaluate overdispersion of pre-defined gene sets\n",
    "\n",
    "In order to detect significant aspects of heterogeneity across the population of single cells, 'pagoda' identifies pathways and gene sets that exhibit statistically significant excess of coordinated variability. Specifically, for each gene set, we tested whether the amount of variance explained by the first principal component significantly exceed the background expectation. \n",
    "\n",
    "__ Using Ensembl's BioMart service to get the GO annotations__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "# Initialize the connection to the Ensembl BioMart Service\n",
    "# Available datasets can be listed with \n",
    "# listDatasets(useMart(\"ensembl\"))\n",
    "# Use mmusculus_gene_ensembl for mouse\n",
    "# For mouse symbols we can use \"mgi_symbol\"\n",
    "ensembl = useMart(\"ensembl\", dataset = \"mmusculus_gene_ensembl\")\n",
    "\n",
    "# Constructs a dataframe with two columns: hgnc_symbol and go_id\n",
    "# If rownames are Ensembl IDs, use ensembl_gene_id as filter value\n",
    "#go = getBM(attributes = c(\"ensembl_gene_id\", \"go_id\"), filters = \"ensembl_gene_id\", values = gsub(\"\\\\.[0-9]+\", \"\", rownames(cd)), mart = ensembl)\n",
    "go = getBM(attributes = c(\"ensembl_gene_id\", \"mgi_symbol\", \"go_id\"), filters = \"ensembl_gene_id\", values = rownames(cd), mart = ensembl)\n",
    "#go = getBM(attributes = c(\"ensembl_gene_id\", \"go_id\"), mart = ensembl)\n",
    "\n",
    "# Use the GO.db library to add a column with the GO-term to the dataframe\n",
    "go$term = Term(go$go_id)\n",
    "\n",
    "# Create a named list of character vectors out of the df\n",
    "s = split(go$ensembl_gene_id, paste(go$go_id,go$term),go$mgi_symbol)\n",
    "\n",
    "# Saves the list as a R environment\n",
    "go.env = list2env(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "# Create a named list of character vectors out of the df\n",
    "s = split(cbind(go$ensembl_gene_id, go$mgi_symbol), paste(go$go_id,go$term))\n",
    "# Saves the list as a R environment\n",
    "go.env = list2env(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "head(ls(go.env))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "# Initialize the connection to the Ensembl BioMart Service\n",
    "# Available datasets can be listed with \n",
    "# listDatasets(useMart(\"ensembl\"))\n",
    "# Use mmusculus_gene_ensembl for mouse\n",
    "# For mouse symbols we can use \"mgi_symbol\"\n",
    "ensembl = useMart(\"ensembl\", dataset = \"mmusculus_gene_ensembl\")\n",
    "\n",
    "# Constructs a dataframe with two columns: hgnc_symbol and go_id\n",
    "# If rownames are Ensembl IDs, use ensembl_gene_id as filter value\n",
    "#go = getBM(attributes = c(\"ensembl_gene_id\", \"go_id\"), filters = \"ensembl_gene_id\", values = gsub(\"\\\\.[0-9]+\", \"\", rownames(cd)), mart = ensembl)\n",
    "goVarinfo = getBM(attributes = c( \"mgi_symbol\", \"go_id\"), filters = \"ensembl_gene_id\", values = rownames(cd), mart = ensembl)\n",
    "#go = getBM(attributes = c(\"ensembl_gene_id\", \"go_id\"), mart = ensembl)\n",
    "\n",
    "# Use the GO.db library to add a column with the GO-term to the dataframe\n",
    "goVarinfo$term = Term(goVarinfo$go_id)\n",
    "\n",
    "# Create a named list of character vectors out of the df\n",
    "s = split(goVarinfo$ensembl_gene_id, paste(goVarinfo$go_id,goVarinfo$term),goVarinfo$mgi_symbol)\n",
    "\n",
    "# Saves the list as a R environment\n",
    "goVarinfo.env = list2env(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "# Use the GO.db library to add a column with the GO-term to the dataframe\n",
    "goVarinfo$term = Term(goVarinfo$go_id)\n",
    "\n",
    "# Create a named list of character vectors out of the df\n",
    "s = split(goVarinfo$mgi_symbol, paste(goVarinfo$go_id,goVarinfo$term))\n",
    "\n",
    "# Saves the list as a R environment\n",
    "goVarinfo.env = list2env(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "# Saving the GO terms with Gene names\n",
    "GOAnnotated <- cbind(go$ensembl_gene_id,go$go_id,go$term)\n",
    "write.csv(GOAnnotated, \"/home/baker/my-scratch-share/James_Bagnall_Single_cell_rna_seq_V3/output/GOAnnotated.csv\", row.names=T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "head(ls(go.env)) # Look at gene set names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "head(get(ls(go.env), go.env)) # Look at one gene set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "ngenes <- unlist(lapply(as.list(go.env),function(x) sum(x %in% rownames(varinfo$mat))))\n",
    "#ngenes <- unlist(lapply(as.list(go.env),function(x) x))\n",
    "head(ngenes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we can calculate weighted first principal component magnitudes for each GO gene set in the provided environment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "pwpca <- pagoda.pathway.wPCA(varinfo, go.env, n.components = 1, n.cores = 2, verbose=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "#pwpca[[1]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now evaluate the statistical significance of the observed overdispersion for each GO gene set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "df <- pagoda.top.aspects(pwpca, return.table = TRUE, plot = TRUE, z.score = 1.96)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "Each point on the plot shows the PC1 variance (lambda1) magnitude (normalized by set size) as a function of set size. The red lines show expected (solid) and 95% upper bound (dashed) magnitudes based on the Tracey-Widom model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "head(df, 20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- The z column gives the Z-score of pathway over-dispersion relative to the genome-wide model (Z-score of 1.96 corresponds to P-value of 5%, etc.).\n",
    "- \"z.adj\" column shows the Z-score adjusted for multiple hypothesis (using Benjamini-Hochberg correction).\n",
    "- \"score\" gives observed/expected variance ratio\n",
    "- \"sh.z\" and \"adj.sh.z\" columns give the raw and adjusted Z-scores of \"pathway cohesion\", which compares the observed PC1 magnitude to the magnitudes obtained when the observations for each gene are randomized with respect to cells. When such Z-score is high (e.g. for GO:0008009) then multiple genes within the pathway contribute to the coordinated pattern.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluate overdispersion of 'de novo' gene sets\n",
    "\n",
    "We can also test 'de novo' gene sets whose expression profiles are well-correlated within the given dataset. The following procedure will determine 'de novo' gene clusters in the data, and build a background model for the expectation of the gene cluster weighted principal component magnitudes. Note the higher trim values for the clusters, as we want to avoid clusters that are formed by outlier cells.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R -w 800 -h 400\n",
    "library(extRemes)\n",
    "library(Lmoments)\n",
    "library(distillery)\n",
    "library(car)\n",
    "\n",
    "clpca <- pagoda.gene.clusters(varinfo, trim = 7.1/ncol(varinfo$mat), n.clusters = 50, n.cores = 2, plot = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The plot above shows background distribution of the first principal component (PC1) variance (lambda1) magnitude. The blue scatterplot on the left shows lambda1 magnitude vs. cluster size for clusters determined based on randomly-generated matrices of the same size. The black circles show top cluster in each simulation. The red lines show expected magnitude and 95% confidence interval based on Tracy-Widom distribution. The right plot shows extreme value distribution fit of residual cluster PC1 variance magnitude relative to the Gumbel (extreme value) distribution.\n",
    "\n",
    "Now the set of top aspects can be recalculated taking these de novo gene clusters into account:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R -w 400 -h 400\n",
    "df <- pagoda.top.aspects(pwpca, clpca, return.table = TRUE, plot = TRUE, z.score = 1.96)\n",
    "write.table(df, \"/home/baker/my-scratch-share/James_Bagnall_Single_cell_rna_seq_V3/output/BASiCS_OUTPUT/TopAspects.txt\", quote = FALSE,\n",
    "            sep = \"\\t\", col.names = NA)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The gene clusters and their corresponding model expected value and 95% upper bound are shown in green."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualize significant aspects of heterogeneity\n",
    "\n",
    "To view top heterogeneity aspects, we will first obtain information on all the significant aspects of transcriptional heterogeneity. We will also determine the overall cell clustering based on this full information:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "# get full info on the top aspects\n",
    "tam <- pagoda.top.aspects(pwpca, clpca, n.cells = NULL, z.score = qnorm(0.01/2, lower.tail = FALSE))\n",
    "# determine overall cell clustering\n",
    "hc <- pagoda.cluster.cells(tam, varinfo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R -w 800 -h 400\n",
    "plot(hc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we will reduce redundant aspects in two steps. First we will combine pathways that are driven by the same sets of genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "tamr <- pagoda.reduce.loading.redundancy(tam, pwpca, clpca)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the second step we will combine aspects that show similar patterns (i.e. separate the same sets of cells). Here we will plot the cells using the overall cell clustering determined above:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R -w 800 -h 400\n",
    "tamr2 <- pagoda.reduce.redundancy(tamr, distance.threshold = 0.9, plot = TRUE, cell.clustering = hc, labRow = NA, labCol = NA, box = TRUE, margins = c(0.5, 0.5), trim = 0, col.cols = rbind(groups = cutree(hc, 4)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the plot above, the columns are cells, rows are different significant aspects, clustered by their similarity pattern.The green-to-orange color scheme shows low-to-high weighted PCA scores (aspect patterns), where generally orange indicates higher expression. Blocks of color on the left margin show which aspects have been combined by the command above. Here the number of resulting aspects is relatively small. \"top\" argument (i.e. top = 10) can be used to limit further analysis to top N aspects.\n",
    "\n",
    "We will view the top aspects, clustering them by pattern similarity (note, to view aspects in the order of increasing `lambda1` magnitude, use `row.clustering = NA`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R -w 800 -h 400\n",
    "pagoda.view.aspects(tamr2, cell.clustering = hc, box = TRUE, labCol = hc, margins = c(0.5, 40), col.cols = rbind(groups = cutree(hc, 4)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "go.env\n",
    "#save.image()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "varinfoTest <- varinfo\n",
    "testVarinfo <- cbind((goVarinfo$ensembl_gene_id),(goVarinfo$mgi_symbol))\n",
    "dim(testVarinfo)\n",
    "#rownames(varinfoTest$mat)<- test[rownames(varinfoTest$mat) == test[,1],2]\n",
    "#test[(rownames(varinfoTest$mat)[1]) == (test[,1]),][1,2]\n",
    "#for(i in 1:dim(varinfoTest$mat)[1]){\n",
    "#    print(test[rownames(varinfoTest$mat)[i] == test[,1],][1,])\n",
    "#    print(i)\n",
    "#}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "print(rownames(varinfoTest$mat)[1148])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R -w 900 -h 400\n",
    "test<- pagoda.show.pathways(c(\"GO:0071346 cellular response to interferon-gamma\",\"GO:0070098 chemokine-mediated signaling pathway\"), varinfo, go.env, cell.clustering = hc, margins = c(1,10), show.cell.dendrogram = TRUE, showRowLabels = TRUE, showPC = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "# compile a browsable app, showing top three clusters with the top color bar\n",
    "app <- make.pagoda.app(tamr2, tam, varinfo, go.env, pwpca, clpca, col.cols = col.cols, cell.clustering = hc, title = \"NPCs\")\n",
    "# show app in the browser (port 1468)\n",
    "show.app(app, \"James_Cluster\", browse = TRUE, port = 1468) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2.7",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "name": "python",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
